---
title: "Ancestry-based prediction"
output: 
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=F)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***
# Introduction
Genotypes vary in frequency between different ancestral groups. As such, it is important to consider an individuals' ancestry in genetic association studies to avoid population stratification and to avoid false positives. However, there are many aspects of health that do vary between ancestral groups, however we are unable to determine whether a variant is causally associated with an outcome, or if it is associated by chance.

In the context of prediction, where the aim is to explain the maximum amount of variance in a phenotype with no consideration for what is causal, difference in peoples ancestry may be a useful predictor. Therefore, here we explore whether principal components of population structure (mainly ancestry), typically used as covariates in association studies, can be used a predictor for a range of health outcomes.

Principal components are typically estimated within the sample of interest to capture sources of population structure which are unique to the sample. However, in this context we want the principal components (PCs), to capture variance that can be generalised across samples, to improve the external validity of our prediction models. One way to achieve this is to derive the PCs in a reference sample, and then project the derived PCs into subsequent target samples, ensuring that PCs correspond to the same variance across samples.

The allele frequency differences between ancestries are not only captured by PCs, but also by other genotype-based scores. Hence why PCs are used as covariates for inference when using polygenic risk scores. Therefore, it is also of interest to determine whether PCs can improve prediction over and above polygenic scores.

<br/>

***

# Aims

1. Test whether reference-projected PCs can significantly predict variance in a range of phenotypes.
2. Test whether reference-projected PCs can significantly improve prediction over polygenic scores alone.
3. Compare the predictive utility of ancestry-specific PCs and across-ancestry PCs.
4. Determine the optimal number of PCs for prediction.

# Methods
## Sample 
- UK Biobank

## Outcomes
- Depression (binary)
- Intelligence (continuous)
- Body mass index (BMI - continuous)
- Height (continuous)

## Genotypic data
HapMap3 variants were extracted from the HRC imputed genetic data, and converted to hard-call PLINK format with no hard-call threshold to maximise overlap with the HapMap3 SNP list. Individuals were excluded if they had extensive missing data, had non-European ancestry, or were closely-related to other individuals in the sample. This was done prior to this project.

## Ancestry scoring
Calculation of the first 100 European-specific and 100 all-ancestry PCs was perfomed using a reference standardised approach (see [here](https://opain.github.io/GenoPred/Pipeline_prep.html#3_ancestry_scoring)). In brief, PCs were derived in the 1000 Genomes reference, then projected onto the UK Biobank sample, and subsequent factor scores were then centered and scaled according to the mean and standard deviation of the PCs in an ancestry matched sample.

## Polygenic scoring
Polygenic scores were derived using PRScs, a Bayesian shrinkage method. Polygenic scores based on a range of global shrinkage parameters were used as this approach was shown to explain the most variance (see [here](https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach.html)). Furthermore, PRScs-based polygenic scores for a range of outcomes were included, as this approach has been shown to increase the variance explained by polygenic scores (see [here](https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach.html)). Maximising the variance explained by polygenic scores will provide more robust evidence that PCs can explain additional variance to polygenic scores, and therefore are a useful predictor.

Polygenic scores were calculated in UK Biobank using a reference-standardised approach (see [here](https://opain.github.io/GenoPred/Pipeline_prep.html#4_polygenic_scoring)). In brief, all scores were derived using HapMap3 SNPs only, modelling LD based on European individuals within the 1000 Genomes reference. Any HapMap3 missing in the target sample are imputed using the allele frequency measured in the European subset of the 1000 Genomes reference. Score were centreed and scaled based on the mean and standard deviation of scores in an ancestry matched samples.

## Estimating predictive ability
All models were derived using elastic-net regularisation to reduce the likelihood of overfitting and account for multicollinearity when modelling highly correlated predictors. 10-fold cross validation was performed using 80% of individuals to identify optimal parameters, with subsequent test-set validation in the remaining 20% of individuals to estimate the predictive utility among individuals not included in the parameter selection process. Permutation testing was used to determine whether the difference in variance explained was significantly greater than 0. Model building and evaluation was performed using an Rscript called Model_builder.R (more information [here]())










