---
title: "Comparison of polygenic scoring methods"
output: 
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

# Introduction
There are two main approaches for calculating polygenic scores (PRSs):

* *p*-value thresholding and clumping
* Shrinkage of GWAS summary statistics
* Best-Linear Unbiased Predictor (BLUP)

We often don't know what the optimal parameters are when deriving a polygenic score, i.e. best p-value threshold or best shrinkage parameters. Methods to identify the best parameter are: 

* Cross-validation (e.g. 10-fold cross validation).
  + Will find optimal parameters, however is time consuming.
* Pseudo-validation - optimal parameters are predicted based on distribution of GWAS summary statistics.
  + Fast but may not find optimal threshold.

Rather than using a single PRS, it may also be advantageous to predict outcome using multiple PRSs based on different parameters. A limitation to this approach is that modelling many predictors is computationally intensive and may increase the risk of overfitting.

Another approach for improving prediction using polygenic scores is to model polygenic scores for multiple outcomes in an elastic net model ([ref](https://www.nature.com/articles/mp2017163),[ref](https://www.nature.com/articles/s41380-019-0394-4)). Here we will aim to replicate these findings with the polygenic score method that this study determines as optimal.

<br/>

***

# Aims
1. Compare PRS derived from p-value and clumping method, shrinkage methods (including lassosum and PRScs), and S-BLUP, using 10-fold cross validation to identify the optimal parameters.
2. Compare the predictive utility of the best PRS identified using 10-fold cross-validated PRS with the PRS selected using pseudo-validation.
3. Compare predictive utility of models including multiple PRSs based on difference parameters, to those using a single PRS selected using 10-fold cross-validation or pseudo-validation.
4. Test whether models containing polygenic scores for multiple outcomes improve prediction over the use of a polygenic score for a single outcome.

<br/>

***

# Methods

## Samples
- UK Biobank
- TEDS

## Outcomes

* UK Biobank
  * Depression (binary)
  * Intelligence (continuous)
  * Body mass index (BMI - continuous)
  * Height (continuous)
  * Coronary Artery Disease (CAD - Binary)
  * Type II Diabetes (T2D - Binary)

* TEDS
  * ADHD traits (continuous)
  * Height (continuous)
  * Body mass index (BMI - continuous)
  * GCSE scores (continuous)


## Genotypic data
HapMap3 variants were extracted from the HRC imputed genetic data, and converted to hard-call PLINK format with no hard-call threshold to maximise overlap with the HapMap3 SNP list. Individuals were excluded if they had extensive missing data, had non-European ancestry, or were closely-related to other individuals in the sample. This was done prior to this project.

## Polygenic scoring
Polygenic scores were derived using three methods: 
1. p-value thresholding and clumping (pT + clump). No pseudo-validation approach for selecting the optimal pT is available). Both sparse (8 pTs) and dense (500 pTs) will be tested.
2. lassosum - Lasso-based shrinkage method.
3. PRScs - Bayesian shrinkage method.
3. SBLUP - Summary statistic-based BLUP estimates

Polygenic scores were derived using a reference standardised pipeline, previously described in detail [here](https://opain.github.io/GenoPred/Pipeline_prep.html#4_polygenic_scoring). In brief, all scores were derived using HapMap3 SNPs only, modelling LD based on European individuals within the 1000 Genomes reference. Any HapMap3 missing in the target sample are imputed using the allele frequency measured in the European subset of the 1000 Genomes reference.

## Estimating predictive ability
All models were derived using elastic-net regularisation to reduce the likelihood of overfitting and account for multicollinearity when modelling highly correlated predictors. 10-fold cross validation was performed using 80% of individuals to identify optimal parameters, with subsequent test-set validation in the remaining 20% of individuals to estimate the predictive utility among individuals not included in the parameter selection process.

Model building and evaluation was performed using an Rscript called Model_builder.R (more information [here](https://github.com/opain/GenoPred/tree/master/Scripts/Model_builder)).

## Code

### UK Biobank
<details><summary>pT + clump PRSs: Sparse</summary>
```{bash, echo=T, eval=F}
##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################

# Make required directories
for pheno_i in $(echo Depression Intelligence BMI Height T2D CAD);do
mkdir -p /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs
done

# Create a file listing the predictors files
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/UKBB/PolygenicScores/EUR/${gwas_i}/UKBB.w_hm3.${gwas_i}.EUR.profiles
EOF

done

# Derive and evaluate models
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d ' ')
  gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
  prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

  qsub -l h_vmem=5G -pe smp 2 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
    --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
    --keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
    --out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs \
    --n_core 2 \
    --compare_predictors T \
    --assoc T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups
done

```

</details>

<details><summary>pT + clump PRSs: Dense</summary>
```{bash, echo=T, eval=F}
##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################

# Create a file listing the predictors files
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs-dense.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/UKBB/PolygenicScores_dense/EUR/${gwas_i}/UKBB.w_hm3.${gwas_i}.EUR.profiles
EOF

done

# Derive and evaluate models
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d ' ')
  gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
  prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

  qsub -l h_vmem=40G -pe smp 3 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
    --pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
    --keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
    --out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs-dense \
    --n_core 3 \
    --assoc T \
    --compare_predictors T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs-dense.predictor_groups
done

```
</details>

<details><summary>lassosum PRSs</summary>
```{bash, echo=T, eval=F}
################################
# Evaluate use of lassosum PRSs when using multiple shrinkage parameters individually and in combination
################################

# Create a file listing the predictors files
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)

for i in $(seq 5 5);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_lassosum.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/UKBB/PolygenicScores_lassosum/EUR/${gwas_i}/UKBB.w_hm3.${gwas_i}.EUR.lassosum_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=20G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
--keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
--out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_lassosum \
--n_core 4 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_lassosum.predictor_groups
done

```
</details>

<details><summary>PRScs PRSs</summary>
```{bash, eval=F, echo=T}
################################
# Evaluate use of multiple PRSs based on different shrinkage parameters, and use of pseudo-validated PRS (PRScs)
################################

# Create a file listing the predictors files
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_PRScs.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/UKBB/PolygenicScores_PRScs/EUR/${gwas_i}/UKBB.w_hm3.${gwas_i}.EUR.PRScs_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=10G -pe smp 2 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
--keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
--out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_PRScs \
--n_core 2 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_PRScs.predictor_groups
done

```
</details>

<details><summary>SBLUP PRSs</summary>
```{bash, eval=F, echo=T}
################################
# Evaluate use of multiple PRSs based on different shrinkage parameters, and use of pseudo-validated PRS (PRScs)
################################

# Create a file listing the predictors files
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_SBLUP.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/UKBB/PolygenicScores_SBLUP/EUR/${gwas_i}/UKBB.w_hm3.${gwas_i}.EUR.SBLUP_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Depression Intelligence BMI Height T2D CAD)
pheno_file=$(echo ever_depressed_pheno_final.UpdateIDs.txt UKBB_Fluid.intelligence.score.UpdateIDs.pheno UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno t2d_only_111119.UpdateIDs.txt cad_only_111119.UpdateIDs.txt)
gwas=$(echo DEPR06 COLL01 BODY03 HEIG03 DIAB05 COAD01)
prev=$(echo 0.15 NA NA NA 0.05 0.03)

for i in $(seq 1 6);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=10G -pe smp 2 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/UKBB/Phenotype/${pheno_i}/${pheno_file_i} \
--keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
--out /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_SBLUP \
--n_core 2 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.w_hm3.${gwas_i}.EUR-PRSs_SBLUP.predictor_groups
done

```
</details>

<br/>

### TEDS
<details><summary>pT + clump PRSs: Sparse</summary>
```{bash, echo=T, eval=F}
##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################

# Make required directories
for pheno_i in $(echo Height21 BMI21 GCSE ADHD);do
mkdir -p /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs
done

# Create a file listing the predictors files
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/TEDS/PolygenicScores/EUR/${gwas_i}/TEDS.w_hm3.${gwas_i}.EUR.profiles
EOF

done

# Derive and evaluate models
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)

for i in $(seq 1 4);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
  prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

  qsub -l h_vmem=5G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
    --pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
	  --keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
    --out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs \
    --n_core 4 \
    --compare_predictors T \
    --assoc T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups
done

```

</details>

<details><summary>pT + clump PRSs: Dense</summary>
```{bash, echo=T, eval=F}
##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################

# Create a file listing the predictors files
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs-dense.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/TEDS/PolygenicScores_dense/EUR/${gwas_i}/TEDS.w_hm3.${gwas_i}.EUR.profiles
EOF

done

# Derive and evaluate models
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)

for i in $(seq 1 4);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
  prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

  qsub -l h_vmem=10G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
    --pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
	  --keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
    --out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs-dense \
    --n_core 4 \
    --assoc T \
    --compare_predictors T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs-dense.predictor_groups
done

```
</details>

<details><summary>lassosum PRSs</summary>
```{bash, echo=T, eval=F}
################################
# Evaluate use of lassosum PRSs when using multiple shrinkage parameters individually and in combination
################################

# Create a file listing the predictors files
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_lassosum.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/TEDS/PolygenicScores_lassosum/EUR/${gwas_i}/TEDS.w_hm3.${gwas_i}.EUR.lassosum_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)


for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=10G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
--keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
--out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_lassosum \
--n_core 4 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_lassosum.predictor_groups
done

```
</details>

<details><summary>PRScs PRSs</summary>
```{bash, eval=F, echo=T}
################################
# Evaluate use of multiple PRSs based on different shrinkage parameters, and use of pseudo-validated PRS (PRScs)
################################

# Create a file listing the predictors files
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_PRScs.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/TEDS/PolygenicScores_PRScs/EUR/${gwas_i}/TEDS.w_hm3.${gwas_i}.EUR.PRScs_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=10G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
--keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
--out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_PRScs \
--n_core 4 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_PRScs.predictor_groups
done

```
</details>

<details><summary>SBLUP PRSs</summary>
```{bash, eval=F, echo=T}
################################
# Evaluate use of multiple PRSs based on different shrinkage parameters, and use of pseudo-validated PRS (PRScs)
################################

# Create a file listing the predictors files
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')

cat > /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_SBLUP.predictor_groups <<EOF
predictors 
/users/k1806347/brc_scratch/Data/TEDS/PolygenicScores_SBLUP/EUR/${gwas_i}/TEDS.w_hm3.${gwas_i}.EUR.SBLUP_profiles
EOF

done

# Derive and evaluate
pheno=$(echo Height21 BMI21 GCSE ADHD)
gwas=$(echo HEIG03 BODY11 EDUC03 ADHD04)
prev=$(echo NA NA NA NA)

for i in $(seq 1 4);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
gwas_i=$(echo ${gwas} | cut -f ${i} -d ' ')
prev_i=$(echo ${prev} | cut -f ${i} -d ' ')

qsub -l h_vmem=5G -pe smp 4 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder.R \
--pheno /users/k1806347/brc_scratch/Data/TEDS/Phenotypic/Derived_outcomes/TEDS_${pheno_i}.txt \
--keep /users/k1806347/brc_scratch/Data/TEDS/Projected_PCs/Ancestry_idenitfier/TEDS.w_hm3.AllAncestry.EUR.keep \
--out /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_SBLUP \
--n_core 4 \
--compare_predictors T \
--assoc T \
--outcome_pop_prev ${prev_i} \
--predictors /users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/${pheno_i}/Association_withPRSs/TEDS.w_hm3.${gwas_i}.EUR-PRSs_SBLUP.predictor_groups
done

```
</details>

<br/>

***

# Results
The pattern of results are consistent between 10-fold cross-validation and test set validation.

Modelling multiple PRSs based on a range of parameters consistenly provides a small improvement over the best single PRS idenitfied by cross-validation.

As previous studies have shown, performing a grid search of shrinkage parameters with both lassosum- and PRScs-based PRSs consistently out performs PRS derived using the traditional pT + clump method. 

The pseudovalidation approach within lassosum performs worse than any other approach tested here. In contrast the PRScs pseduovalidated PRS performs well, providing nearly the best prediction for all outcomes.

<br/>

## UK Biobank

<details><summary>Code for plotting results</summary>

```{bash, echo=T, eval=F}
#####
# Compare results from each approach
#####
module add general/R/3.5.0
R

pheno<-c('Depression','Intelligence','BMI','Height','T2D','CAD')
gwas<-c('DEPR06','COLL01','BODY03','HEIG03','DIAB05','COAD01')

# Read in the results, labelling which parameters where pseudovalidated and validated using 10-fold cross validation
All_res<-NULL
for(i in c(1:6)){
  pT_clump<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/UKBB.w_hm3.',gwas[i],'.EUR-PRSs.pred_eval.txt'), header=T, stringsAsFactors=F)
  pT_clump$Method<-"pT + clump\n(sparse)"
  Param_tmp<-sort(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump$Model)))))
  pT_clump$Param<-as.character(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump$Model)))))
  pT_clump$Param[dim(pT_clump)[1]]<-'Multi-PRS'
  pT_clump<-pT_clump[match(c(Param_tmp,'Multi-PRS'), pT_clump$Param),]
  pT_clump$Model<-NULL
  pT_clump$Group<-NA
  pT_clump$Group[pT_clump$CrossVal_R == max(abs(pT_clump$CrossVal_R[pT_clump$Param != 'Multi-PRS']))]<-'10FCVal'
  pT_clump$Group[pT_clump$Param == 'Multi-PRS']<-'Multi-PRS'
  pT_clump<-pT_clump[!is.na(pT_clump$Group),]

  pT_clump_dense<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/UKBB.w_hm3.',gwas[i],'.EUR-PRSs-dense.pred_eval.txt'), header=T, stringsAsFactors=F)
  pT_clump_dense$Method<-"pT + clump\n(dense)"
  Param_tmp<-sort(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'_'),'',pT_clump_dense$Model)))))
  pT_clump_dense$Param<-as.character(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'_'),'',pT_clump_dense$Model)))))
  pT_clump_dense$Param[dim(pT_clump_dense)[1]]<-'Multi-PRS'
  pT_clump_dense<-pT_clump_dense[match(c(Param_tmp,'Multi-PRS'), pT_clump_dense$Param),]
  pT_clump_dense$Model<-NULL
  pT_clump_dense$Group<-NA
  pT_clump_dense$Group[pT_clump_dense$CrossVal_R == max(abs(pT_clump_dense$CrossVal_R[pT_clump_dense$Param != 'Multi-PRS']))][1]<-'10FCVal'
  pT_clump_dense$Group[pT_clump_dense$Param == 'Multi-PRS']<-'Multi-PRS'
  pT_clump_dense<-pT_clump_dense[!is.na(pT_clump_dense$Group),]
  
  lassosum_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/UKBB.w_hm3.',gwas[i],'.EUR-PRSs_lassosum.pred_eval.txt'), header=T, stringsAsFactors=F)
  lassosum_res$Method<-'lassosum'
  lassosum_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',lassosum_res$Model))
  lassosum_res$Param[dim(lassosum_res)[1]]<-'Multi-PRS'
  lassosum_res$Model<-NULL
  lassosum_val<-read.table(paste0('/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_poylygenic_lassosum/',gwas[i],'/1KGPhase3.w_hm3.',gwas[i],'.log'), sep='&')
  lassosum_val_s<-as.numeric(gsub('s = ','',lassosum_val$V1[grepl('s = ',lassosum_val$V1)]))
  lassosum_val_lambda<-as.numeric(gsub('lambda =','',lassosum_val$V1[grepl('lambda =',lassosum_val$V1)]))
  lassosum_val_param<-paste0('s',lassosum_val_s,'.lambda',lassosum_val_lambda)
  lassosum_val_param<-substr(lassosum_val_param, 1, nchar(lassosum_val_param)-1) 
  lassosum_res$Group<-NA
  lassosum_res$Group[which(lassosum_res$CrossVal_R == max(abs(lassosum_res$CrossVal_R[lassosum_res$Param != 'Multi-PRS'])))[1]]<-'10FCVal'
  lassosum_res$Group[grepl(lassosum_val_param,lassosum_res$Param)]<-'PseudoVal'
  lassosum_res$Group[lassosum_res$Param == 'Multi-PRS']<-'Multi-PRS'
  lassosum_res<-lassosum_res[!is.na(lassosum_res$Group),]
  
  PRScs_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/UKBB.w_hm3.',gwas[i],'.EUR-PRSs_PRScs.pred_eval.txt'), header=T, stringsAsFactors=F)
  PRScs_res$Method<-'PRScs'
  PRScs_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',PRScs_res$Model))
  PRScs_res$Param[dim(PRScs_res)[1]]<-'Multi-PRS'
  PRScs_res$Model<-NULL
  PRScs_res$Group<-NA
  PRScs_res$Group[which(PRScs_res$CrossVal_R == max(abs(PRScs_res$CrossVal_R[PRScs_res$Param != 'Multi-PRS' & PRScs_res$Param != 'phiauto'])))[1]]<-'10FCVal'
  PRScs_res$Group[PRScs_res$Param == 'phiauto']<-'PseudoVal'
  PRScs_res$Group[PRScs_res$Param == 'Multi-PRS']<-'Multi-PRS'
  PRScs_res<-PRScs_res[!is.na(PRScs_res$Group),]

  SBLUP_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/UKBB.w_hm3.',gwas[i],'.EUR-PRSs_SBLUP.pred_eval.txt'), header=T, stringsAsFactors=F)
  SBLUP_res$Method<-'SBLUP'
  SBLUP_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',SBLUP_res$Model))
  SBLUP_res$Model<-NULL
  SBLUP_res$Group<-NA
  SBLUP_res$Group[1]<-'PseudoVal'
  SBLUP_res<-SBLUP_res[!is.na(SBLUP_res$Group),]

#  res<-do.call(rbind, list(pT_clump, pT_clump_dense, lassosum_res, PRScs_res))
  res<-do.call(rbind, list(pT_clump, lassosum_res, PRScs_res, SBLUP_res))
  res$Phenotype<-pheno[i]
  res$Cross_LiabR2<-NULL
  res$Indep_LiabR2<-NULL
  res$CrossVal_pval<-NULL
  res$IndepVal_pval<-NULL

  All_res<-rbind(All_res, res)
}

# Reformat to be a table
All_res<-All_res[c('Phenotype','Method','Group','CrossVal_R','CrossVal_R_SE','IndepVal_R','IndepVal_R_SE')]

write.csv(All_res, '/mnt/lustre/users/k1806347/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_UKBB.csv', row.names=F)

###
# Plot comparison between validated PRS vs multi-PRS when using pT + clump method
###

library(ggplot2)
library(cowplot)

All_res$Method<-factor(All_res$Method, levels=unique(All_res$Method))

plots<-list()
for(i in 1:6){
plots[[pheno[i]]] <- ggplot( All_res[All_res$Phenotype==pheno[i],], aes(x=Method, y=CrossVal_R, fill=Group)) +
                      geom_bar(stat="identity", position=position_dodge(preserve = "single"), width = 0.7, colour='black') +
                      geom_errorbar(aes(ymin=CrossVal_R-CrossVal_R_SE, ymax=CrossVal_R+CrossVal_R_SE), width=.2, position=position_dodge(width = 0.7, preserve = "single")) +
                      labs(y="Correlation (SE)", x='', title=pheno[i]) +
                      coord_cartesian(ylim=c(min(All_res$CrossVal_R[All_res$Phenotype==pheno[i]])-0.02, max(All_res$CrossVal_R[All_res$Phenotype==pheno[i]])+0.025), clip="on") +
            				  theme(axis.text.x = element_text(angle = 45, hjust = 1))
}

png('/mnt/lustre/users/k1806347/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_CrossVal_UKBB.png', units='px', res=300, width=3000, height=3750)
plot_grid(plotlist=plots, ncol = 2)
dev.off()

plots<-list()
for(i in 1:6){
plots[[pheno[i]]] <- ggplot( All_res[All_res$Phenotype==pheno[i],], aes(x=Method, y=IndepVal_R, fill=Group)) +
                      geom_bar(stat="identity", position=position_dodge(preserve = "single"), width = 0.7, colour='black') +
                      geom_errorbar(aes(ymin=IndepVal_R-IndepVal_R_SE, ymax=IndepVal_R+IndepVal_R_SE), width=.2, position=position_dodge(width = 0.7, preserve = "single")) +
                      labs(y="Correlation (SE)", x='', title=pheno[i]) +
                      coord_cartesian(ylim=c(min(All_res$IndepVal_R[All_res$Phenotype==pheno[i]])-0.02, max(All_res$IndepVal_R[All_res$Phenotype==pheno[i]])+0.025), clip="on") +
            				  theme(axis.text.x = element_text(angle = 45, hjust = 1))
}

png('/mnt/lustre/users/k1806347/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_IndepVal_UKBB.png', units='px', res=300, width=3000, height=3750)
plot_grid(plotlist=plots, ncol = 2)
dev.off()

q()
n
```

</details>

```{bash, eval=T, echo=F}
mkdir -p /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach

cp /mnt/lustre/users/k1806347/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_CrossVal_UKBB.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/

cp /mnt/lustre/users/k1806347/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_IndepVal_UKBB.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/
```

<details><summary>Show UKBB cross-validation results</summary>

<center>

![Figure 1: Predictive utility estimated using 10-fold cross-validation](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/PRS_method_comp_CrossVal_UKBB.png)

</details>

<details><summary>Show UKBB test-validation results</summary>

![Figure 2: Predictive utility estimated using test set validation](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/PRS_method_comp_IndepVal_UKBB.png)

\center

</details>

<details><summary>Show UKBB results table</summary>

```{r, echo=F, eval=T, results='asis'}
res<-read.csv("/users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/PRS_method_comp_UKBB.csv")

res[,-1:-3]<-round(res[,-1:-3], 3)
res$CrossVal_R<-paste0(res$CrossVal_R, " (", res$CrossVal_R_SE, ")")
res$IndepVal_R<-paste0(res$IndepVal_R, " (", res$IndepVal_R_SE, ")")
res<-res[c('Phenotype','Method','Group','CrossVal_R','IndepVal_R')]

names(res)<-c('Phenotype','Method','Group',"CrossVal R (SE)","IndepVal R (SE)")

library(DT)
datatable(res, rownames = FALSE, filter="top", options = list(pageLength = 10, scrollX=T),caption='Table 1: Correlation between PRS model predictions and observed values in UK Biobank')
```

</details>

<br/>

## TEDS

<details><summary>Code for plotting results</summary>
```{bash, echo=T, eval=F}
#####
# Compare results from each approach
#####
module add general/R/3.5.0
R

pheno<-c('Height21', 'BMI21', 'GCSE', 'ADHD')
gwas<-c('HEIG03', 'BODY11', 'EDUC03', 'ADHD04')

# Read in the results, labelling which parameters where pseudovalidated and validated using 10-fold cross validation
All_res<-NULL
for(i in c(1:4)){
  pT_clump<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/TEDS.w_hm3.',gwas[i],'.EUR-PRSs.pred_eval.txt'), header=T, stringsAsFactors=F)
  pT_clump$Method<-"pT + clump\n(sparse)"
  Param_tmp<-sort(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump$Model)))))
  pT_clump$Param<-as.character(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump$Model)))))
  pT_clump$Param[dim(pT_clump)[1]]<-'Multi-PRS'
  pT_clump<-pT_clump[match(c(Param_tmp,'Multi-PRS'), pT_clump$Param),]
  pT_clump$Model<-NULL
  pT_clump$Group<-NA
  pT_clump$Group[pT_clump$CrossVal_R == max(abs(pT_clump$CrossVal_R[pT_clump$Param != 'Multi-PRS']))]<-'10FCVal'
  pT_clump$Group[pT_clump$Param == 'Multi-PRS']<-'Multi-PRS'
  pT_clump<-pT_clump[!is.na(pT_clump$Group),]

  pT_clump_dense<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/TEDS.w_hm3.',gwas[i],'.EUR-PRSs-dense.pred_eval.txt'), header=T, stringsAsFactors=F)
  pT_clump_dense$Method<-"pT + clump\n(dense)"
  Param_tmp<-sort(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump_dense$Model)))))
  pT_clump_dense$Param<-as.character(as.numeric(gsub('e.','e-',gsub('_group','',gsub(paste0(gwas[i],'.'),'',pT_clump_dense$Model)))))
  pT_clump_dense$Param[dim(pT_clump_dense)[1]]<-'Multi-PRS'
  pT_clump_dense<-pT_clump_dense[match(c(Param_tmp,'Multi-PRS'), pT_clump_dense$Param),]
  pT_clump_dense$Model<-NULL
  pT_clump_dense$Group<-NA
  pT_clump_dense$Group[pT_clump_dense$CrossVal_R == max(abs(pT_clump_dense$CrossVal_R[pT_clump_dense$Param != 'Multi-PRS']))][1]<-'10FCVal'
  pT_clump_dense$Group[pT_clump_dense$Param == 'Multi-PRS']<-'Multi-PRS'
  pT_clump_dense<-pT_clump_dense[!is.na(pT_clump_dense$Group),]
  
  lassosum_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/TEDS.w_hm3.',gwas[i],'.EUR-PRSs_lassosum.pred_eval.txt'), header=T, stringsAsFactors=F)
  lassosum_res$Method<-'lassosum'
  lassosum_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',lassosum_res$Model))
  lassosum_res$Param[dim(lassosum_res)[1]]<-'Multi-PRS'
  lassosum_res$Model<-NULL
  lassosum_val<-read.table(paste0('/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_poylygenic_lassosum/',gwas[i],'/1KGPhase3.w_hm3.',gwas[i],'.log'), sep='&')
  lassosum_val_s<-as.numeric(gsub('s = ','',lassosum_val$V1[grepl('s = ',lassosum_val$V1)]))
  lassosum_val_lambda<-as.numeric(gsub('lambda =','',lassosum_val$V1[grepl('lambda =',lassosum_val$V1)]))
  lassosum_val_param<-paste0('s',lassosum_val_s,'.lambda',lassosum_val_lambda)
  lassosum_val_param<-substr(lassosum_val_param, 1, nchar(lassosum_val_param)-1) 
  lassosum_res$Group<-NA
  lassosum_res$Group[which(lassosum_res$CrossVal_R == max(abs(lassosum_res$CrossVal_R[lassosum_res$Param != 'Multi-PRS'])))[1]]<-'10FCVal'
  lassosum_res$Group[grepl(lassosum_val_param,lassosum_res$Param)]<-'PseudoVal'
  lassosum_res$Group[lassosum_res$Param == 'Multi-PRS']<-'Multi-PRS'
  lassosum_res<-lassosum_res[!is.na(lassosum_res$Group),]
  
  PRScs_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/TEDS.w_hm3.',gwas[i],'.EUR-PRSs_PRScs.pred_eval.txt'), header=T, stringsAsFactors=F)
  PRScs_res$Method<-'PRScs'
  PRScs_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',PRScs_res$Model))
  PRScs_res$Param[dim(PRScs_res)[1]]<-'Multi-PRS'
  PRScs_res$Model<-NULL
  PRScs_res$Group<-NA
  PRScs_res$Group[which(PRScs_res$CrossVal_R == max(abs(PRScs_res$CrossVal_R[PRScs_res$Param != 'Multi-PRS' & PRScs_res$Param != 'phiauto'])))[1]]<-'10FCVal'
  PRScs_res$Group[PRScs_res$Param == 'phiauto']<-'PseudoVal'
  PRScs_res$Group[PRScs_res$Param == 'Multi-PRS']<-'Multi-PRS'
  PRScs_res<-PRScs_res[!is.na(PRScs_res$Group),]

  SBLUP_res<-read.table(paste0('/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/',pheno[i],'/Association_withPRSs/TEDS.w_hm3.',gwas[i],'.EUR-PRSs_SBLUP.pred_eval.txt'), header=T, stringsAsFactors=F)
  SBLUP_res$Method<-'SBLUP'
  SBLUP_res$Param<-gsub('_group','',gsub(paste0(gwas[i],'.'),'',SBLUP_res$Model))
  SBLUP_res$Model<-NULL
  SBLUP_res$Group<-NA
  SBLUP_res$Group[1]<-'PseudoVal'
  SBLUP_res<-SBLUP_res[!is.na(SBLUP_res$Group),]

  res<-do.call(rbind, list(pT_clump, pT_clump_dense, lassosum_res, PRScs_res, SBLUP_res))
  res$Phenotype<-pheno[i]
  res$Cross_LiabR2<-NULL
  res$Indep_LiabR2<-NULL
  res$CrossVal_pval<-NULL
  res$IndepVal_pval<-NULL

  All_res<-rbind(All_res, res)
}

# Reformat to be a table
All_res<-All_res[c('Phenotype','Method','Group','CrossVal_R','CrossVal_R_SE','IndepVal_R','IndepVal_R_SE')]

write.csv(All_res, '/mnt/lustre/users/k1806347/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_TEDS.csv', row.names=F)

###
# Plot comparison between validated PRS vs multi-PRS when using pT + clump method
###

library(ggplot2)
library(cowplot)

All_res$Method<-factor(All_res$Method, levels=unique(All_res$Method))

plots<-list()
for(i in 1:4){
plots[[pheno[i]]] <- ggplot( All_res[All_res$Phenotype==pheno[i],], aes(x=Method, y=CrossVal_R, fill=Group)) +
                      geom_bar(stat="identity", position=position_dodge(preserve = "single"), width = 0.7, colour='black') +
                      geom_errorbar(aes(ymin=CrossVal_R-CrossVal_R_SE, ymax=CrossVal_R+CrossVal_R_SE), width=.4, position=position_dodge(width = 0.7, preserve = "single")) +
                      labs(y="Correlation (SE)", x='', title=pheno[i]) +
                      coord_cartesian(ylim=c(min(All_res$CrossVal_R[All_res$Phenotype==pheno[i]])-0.02, max(All_res$CrossVal_R[All_res$Phenotype==pheno[i]])+0.025), clip="on") +
            				  theme(axis.text.x = element_text(angle = 45, hjust = 1))
}

png('/mnt/lustre/users/k1806347/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_CrossVal_TEDS.png', units='px', res=300, width=3000, height=2750)
plot_grid(plotlist=plots, ncol = 2)
dev.off()

plots<-list()
for(i in 1:4){
plots[[pheno[i]]] <- ggplot( All_res[All_res$Phenotype==pheno[i],], aes(x=Method, y=IndepVal_R, fill=Group)) +
                      geom_bar(stat="identity", position=position_dodge(preserve = "single"), width = 0.7, colour='black') +
                      geom_errorbar(aes(ymin=IndepVal_R-IndepVal_R_SE, ymax=IndepVal_R+IndepVal_R_SE), width=.2, position=position_dodge(width = 0.7, preserve = "single")) +
                      labs(y="Correlation (SE)", x='', title=pheno[i]) +
                      coord_cartesian(ylim=c(min(All_res$IndepVal_R[All_res$Phenotype==pheno[i]])-0.02, max(All_res$IndepVal_R[All_res$Phenotype==pheno[i]])+0.025), clip="on") +
            				  theme(axis.text.x = element_text(angle = 45, hjust = 1))
}

png('/mnt/lustre/users/k1806347/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_IndepVal_TEDS.png', units='px', res=300, width=3000, height=2750)
plot_grid(plotlist=plots, ncol = 2)
dev.off()

q()
n
```
</details>

```{bash, eval=T, echo=F}
mkdir -p /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach

cp /mnt/lustre/users/k1806347/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_CrossVal_TEDS.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/

cp /mnt/lustre/users/k1806347/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_IndepVal_TEDS.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/
```

<details><summary>Show TEDS cross-validation results</summary>

<center>

![Figure 3: Predictive utility estimated using 10-fold cross-validation](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/PRS_method_comp_CrossVal_TEDS.png)

</details>

<details><summary>Show TEDS test-validation results</summary>

![Figure 4: Predictive utility estimated using test set validation](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Determine_optimal_polygenic_scoring_approach/PRS_method_comp_IndepVal_TEDS.png)

\center

</details>

<details><summary>Show UKBB results table</summary>

```{r, echo=F, eval=T, results='asis'}
res<-read.csv("/users/k1806347/brc_scratch/Analyses/TEDS_outcomes_for_prediction/PRS_method_comp_TEDS.csv")

res[,-1:-3]<-round(res[,-1:-3], 3)
res$CrossVal_R<-paste0(res$CrossVal_R, " (", res$CrossVal_R_SE, ")")
res$IndepVal_R<-paste0(res$IndepVal_R, " (", res$IndepVal_R_SE, ")")
res<-res[c('Phenotype','Method','Group','CrossVal_R','IndepVal_R')]

names(res)<-c('Phenotype','Method','Group',"CrossVal R (SE)","IndepVal R (SE)")

library(DT)
datatable(res, rownames = FALSE, filter="top", options = list(pageLength = 10, scrollX=T),caption='Table 1: Correlation between PRS model predictions and observed values in TEDS')
```

</details>

<br/>

***

# Discussion

This study has replicated previous findings that shrinkage of GWAS summary statistics using lassosum or PRScs can increase the variance explained by subsequent PRSs. Interestingly predicting the optimal shrinkage paramter via pseudovalidation is suboptimal for lassosum, with selection of the shrinkage parameter by formal cross validation providing substantial improvements the variance explained by the PRS. In contrast psudovalidation, or fully bayesian mode, implemented by PRScs performs well, providing PRSs with a predictive utility close to those derived using a shrinkage parameter identified via cross validation.

This study also reports that use of an elastic net to model model PRSs, based on a range of shrinkage paramters provides the optimal variance explained for all methods and outcomes. However, the improvement in prediction is small, and computational burden of generating multiple PRSs, and modelling them in an elastic net is substantial. Furthermore, as more predictors are included in the model, the likelihood of overfitting increases, particularly in smaller samples.

<br/>

***

# Conclusion
Based on our results we think suggest the use of PRScs using only the pseudovalidated, or fully bayesian, global shrinkage parameter, as this provides near optimal prediction whilst reducing computional burden and risk of overfitting.

<br/>

***

