<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Prediction within Ancestral Diversity</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Prediction within Ancestral Diversity</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Genotype-based prediction is more accurate within European populations due to European-based GWAS being larger in sample size, and the relatively low admixture within European populations compared to non-European populations. Apart from phenotypic heterogenity across populations, a key reason why European GWAS do not predict well in non-European populations is due to differences in LD and MAF. Although the underlying causal variant maybe the same across populations, the variant best tagging the causal variant in one population may not wel tag the causal variant in another population.</p>
<p>Prediction in non-European populations by:</p>
<ul>
<li>Performing larger GWAS in non-European populations</li>
<li>Combination of European-GWAS with existing non-European GWAS</li>
<li>Modelling differences in MAF and LD between populations</li>
<li>Using gene-based scores</li>
</ul>
<p>Across the genome some regions will be more comparable across populations than others. Furthermore, within an individual there is often a mixture of haplotypes from different populations (admixture), meaning a European GWAS maybe predictive for an individual for certain regions only. Merely by removing loci that do not generalise across populations well may improve prediction by reducing noise.</p>
<hr />
</div>
<div id="evaluate-prediction-across-a-range-of-ancestries" class="section level1">
<h1><span class="header-section-number">2</span> Evaluate prediction across a range of ancestries</h1>
<p>Global ancestry is typically defined using reference-projected genotype-based principal components. Often individuals are said to be of a population if they are within N SD of the population mean. Alternatively principal component estimates can be used in a machine learning based approach such as k-means clustering. I have used an elastic net model to predict global ancestry.</p>
<p>Estimate the predictive utility of polygenic scores within each super population. UKB contains individuals within each super population.</p>
<hr />
<div id="defining-ancestry-and-qc" class="section level2">
<h2><span class="header-section-number">2.1</span> Defining ancestry and QC</h2>
<p>Further information can be found here: <a href="https://opain.github.io/UKB-GenoPrep/" class="uri">https://opain.github.io/UKB-GenoPrep/</a></p>
<hr />
</div>
<div id="using-external-gwas-sumstats" class="section level2">
<h2><span class="header-section-number">2.2</span> Using external GWAS sumstats</h2>
<hr />
<div id="prepare-phenotype-files" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Prepare phenotype files</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="r"><code>source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config&#39;)
library(data.table)

# Read list of individuals surviving QC in each population
# Read in fam to convert row number to application specific ID
# Row numbers in pheno file cannot be trusted
fam&lt;-fread(&#39;/scratch/groups/ukbiobank/ukb18177_glanville/genotyped/ukb18177_glanville_binary_pre_qc.fam&#39;)
fam$V1&lt;-seq(1:dim(fam)[1])

fam_2&lt;-fread(paste0(UKBB_output,&#39;/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam&#39;))

fam_merged&lt;-merge(fam, fam_2, by=&#39;V2&#39;)

keep&lt;-list()
for(pop in c(&#39;EUR&#39;,&#39;AFR&#39;,&#39;SAS&#39;,&#39;EAS&#39;,&#39;AMR&#39;)){
  keep[[pop]]&lt;-fread(paste0(&#39;/scratch/groups/ukbiobank/usr/ollie_pain/ReQC/PostQC/UKB.postQC.&#39;,pop,&#39;.keep&#39;))
  keep[[pop]]&lt;-merge(keep[[pop]][,1], fam[,1:2], by=&#39;V1&#39;)
  keep[[pop]]$V1&lt;-keep[[pop]]$V2
  
  keep_tmp&lt;-fread(paste0(&#39;/scratch/groups/ukbiobank/usr/ollie_pain/ReQC/PostQC/UKB.postQC.&#39;,pop,&#39;.keep&#39;))
  keep_tmp&lt;-merge(keep_tmp[,1], fam_merged, by.x=&#39;V1&#39;, by.y=&#39;V1.x&#39;)
  keep_tmp&lt;-keep_tmp[,c(&#39;V1.y&#39;,&#39;V2&#39;),with=F]

    # Save keep file with updated IDs
  write.table(keep[[pop]], paste0(UKBB_output,&#39;/Phenotype/&#39;,pop,&#39;.QC.keep&#39;), col.names=F, row.names=F, quote=F)
  write.table(keep_tmp, paste0(UKBB_output,&#39;/Phenotype/&#39;,pop,&#39;.QC.UpdateIDs.keep&#39;), col.names=F, row.names=F, quote=F)
}

# Read in phenotype data
pheno&lt;-c(&#39;Depression&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;,&#39;Height&#39;,&#39;T2D&#39;,&#39;CAD&#39;,&#39;IBD&#39;,&#39;MultiScler&#39;,&#39;RheuArth&#39;)
pheno_file&lt;-c(&#39;ever_depressed_pheno_final.UpdateIDs.txt&#39;,&#39;UKBB_Fluid.intelligence.score.UpdateIDs.pheno&#39;,&#39;UKBB_BMI.score.UpdateIDs.pheno&#39;,&#39;UKBB_Height.score.UpdateIDs.pheno&#39;,&#39;t2d_only_111119.UpdateIDs.txt&#39;,&#39;cad_only_111119.UpdateIDs.txt&#39;,&#39;UKBB.IBD.txt&#39;,&#39;UKBB.MultiScler.txt&#39;,&#39;UKBB.RheuArth.txt&#39;)

pheno_list&lt;-list()
for(i in 1:length(pheno)){
  pheno_list[[pheno[i]]]&lt;-fread(paste0(UKBB_output,&#39;/Phenotype/&#39;,pheno[i],&#39;/&#39;,pheno_file[i]))
}

pheno_all&lt;-Reduce(function(...) merge(..., all=T, by=c(&#39;FID&#39;,&#39;IID&#39;)), pheno_list)
names(pheno_all)&lt;-c(&#39;FID&#39;,&#39;IID&#39;,pheno)

# Count number of non-missing values for phenotype in each population 
pheno_dat_per_pop&lt;-NULL
for(pop in c(&#39;EUR&#39;,&#39;AFR&#39;,&#39;SAS&#39;,&#39;EAS&#39;,&#39;AMR&#39;)){
  pheno_all_pop&lt;-pheno_all[(pheno_all$IID %in% keep[[pop]]$V1),]
  for(i in 1:length(pheno)){
    if(length(na.omit(unique(pheno_all[[pheno[i]]]))) == 2){
          pheno_dat_per_pop&lt;-rbind(pheno_dat_per_pop, data.frame(Pop=pop,
                                                           Phenotype=pheno[i],
                                                           N=sum(!is.na(pheno_all_pop[[pheno[i]]])),
                                                           Ncase=sum(pheno_all_pop[[pheno[i]]] == 1, na.rm=T),
                                                           Ncon=sum(pheno_all_pop[[pheno[i]]] == 0, na.rm=T)))

    } else {
          pheno_dat_per_pop&lt;-rbind(pheno_dat_per_pop, data.frame(Pop=pop,
                                                           Phenotype=pheno[i],
                                                           N=sum(!is.na(pheno_all_pop[[pheno[i]]])),
                                                           Ncase=NA,
                                                           Ncon=NA))

    }
  }
}

# Idenitfy phenotype with at least 1000 indidivuals and 50 cases with data
pheno_dat_per_pop_retain&lt;-pheno_dat_per_pop[which((pheno_dat_per_pop$N &gt;=500 &amp; pheno_dat_per_pop$Ncase &gt;= 50) | (pheno_dat_per_pop$N &gt;=500 &amp; is.na(pheno_dat_per_pop$Ncase))),]

# BMI and Height are the only phenotypes meeting these criteria. This is fine for a preliminary analysis, but other phenotypes that are more available across ancestries should be explored. Intelligence was also available in EUR, AFR, and SAS.

# Save keep file with updated IDs for each phenotype, restricting the sample size to 50K
# Read in fam file with ID to match
system(paste0(&#39;mkdir &#39;,UKBB_output,&#39;/DiverseAncestry/Phenotype_subsets/&#39;))

set.seed(1)
for(pop in c(&#39;EUR&#39;,&#39;AFR&#39;,&#39;SAS&#39;,&#39;EAS&#39;,&#39;AMR&#39;)){
  for(pheno_i in c(&#39;BMI&#39;,&#39;Height&#39;)){
      pheno_all_pop&lt;-pheno_all[(pheno_all$IID %in% keep[[pop]]$V1),c(&#39;FID&#39;,&#39;IID&#39;,pheno_i), with=F]
      pheno_all_pop&lt;-pheno_all_pop[complete.cases(pheno_all_pop),]
      if(dim(pheno_all_pop)[1] &gt; 50000){
        pheno_all_pop&lt;-pheno_all_pop[sample(1:50000),]
      }
      
      write.table(pheno_all_pop[,1:2], paste0(UKBB_output,&#39;/DiverseAncestry/Phenotype_subsets/UKB.&#39;,pheno_i,&#39;.&#39;,pop,&#39;.QC.keep&#39;), col.names=F, row.names=F, quote=F)
  }
}</code></pre>
</details>
<hr />
</div>
<div id="calculate-scores" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Calculate scores</h3>
<hr />
<div id="prs" class="section level4">
<h4><span class="header-section-number">2.2.2.1</span> PRS</h4>
<details>
<p><summary>pT + clump: Sparse</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch --mem 2G -p brc,shared -J pT_clump /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/pt_clump/${gwas_i}/1KGPhase3.w_hm3.${gwas_i} \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/pt_clump/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --plink ${plink1_9} \
      --pheno_name ${gwas_i} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>lassosum</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 5 --mem 5G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_lassosum/Scaled_polygenic_scorer_lassosum.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/lassosum/${gwas_i}/1KGPhase3.w_hm3.${gwas_i} \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/lassosum/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --n_cores 5 \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/lassosum/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>PRScs</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 1 --mem 5G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_PRScs/Scaled_polygenic_scorer_PRScs.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
    --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/PRScs/${gwas_i}/1KGPhase3.w_hm3.${gwas_i} \
    --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/PRScs/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/PRScs/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>SBLUP</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 1 --mem 5G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_SBLUP/Scaled_polygenic_scorer_SBLUP.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
    --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/SBLUP/${gwas_i}/GWAS_sumstats_SBLUP.sblup.cojo \
    --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/SBLUP/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/SBLUP/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>SBayesR</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 1 --mem 5G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_SBayesR/Scaled_polygenic_scorer_SBayesR.R \
 \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/SBayesR/${gwas_i}_GCTB_203_robust/GWAS_sumstats_SBayesR.GW.snpRes \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/SBayesR/${gwas_i}_GCTB_203_robust/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/SBayesR/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
      
  done
done</code></pre>
</details>
<details>
<p><summary>LDPred</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 1 --mem 5G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_LDPred/Scaled_polygenic_scorer_LDPred.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/LDPred/${gwas_i}/1KGPhase3.w_hm3.${gwas_i} \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/LDPred/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/LDPred/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>LDPred2</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch -n 6 --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_LDPred2/Scaled_polygenic_scorer_LDPred2.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/LDPred2/${gwas_i}/1KGPhase3.w_hm3.${gwas_i} \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/LDPred2/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --n_cores 6 \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/LDPred2/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<details>
<p><summary>DBSLMM</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  
    sbatch --mem 10G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_DBSLMM/Scaled_polygenic_scorer_DBSLMM.R \
      --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
      --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.${pheno_i}.${pop}.QC.keep \
      --ref_score ${Geno_1KG_dir}/Score_files_for_polygenic/DBSLMM/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.dbslmm.GW.txt \
      --ref_scale ${Geno_1KG_dir}/Score_files_for_polygenic/DBSLMM/${gwas_i}/1KGPhase3.w_hm3.${gwas_i}.${pop}.scale \
      --ref_freq_chr ${Geno_1KG_dir}/freq_files/${pop}/1KGPhase3.w_hm3.${pop}.chr \
      --pheno_name ${gwas_i} \
      --plink ${plink1_9} \
      --output ${UKBB_output}/DiverseAncestry/1KG_ref/DBSLMM/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}
  done
done
</code></pre>
</details>
<hr />
</div>
<div id="gers" class="section level4">
<h4><span class="header-section-number">2.2.2.2</span> GeRS</h4>
<details>
<p><summary>GeRS</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for pop in $(echo AFR AMR EAS EUR SAS);do
  mkdir -p ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}
  &gt; ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt
  for i in $(seq 1 2);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
    
    for weights in $(cat ~/brc_scratch/Data/TWAS_sumstats/FUSION/snp_weight_list.txt);do
      if [ ! -f ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/${weights}/UKBB.w_hm3.EUR.${weights}.${gwas_i}.fiprofile ]; then
        echo $gwas_i $pheno_i $weights &gt;&gt; ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt
      fi
    done
  done
cat &lt;&lt;EOF &gt;${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo_array.sh
#!/bin/bash
#SBATCH -p shared,brc
#SBATCH --mem 10G
#SBATCH -n 1

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

gwas=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$1}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt)
pheno=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$2}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt)
weights=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$3}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt)

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_functionally_informed_risk_scorer/scaled_functionally_informed_risk_scorer.R \
  --targ_feature_pred ${UKBB_output}/Predicted_expression/FUSION/${pop}/\${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.\${weights}.predictions.gz \
  --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.\${pheno}.${pop}.QC.keep \
  --ref_score ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/\${gwas}/1KGPhase3.w_hm3.EUR.FUSION.\${gwas}.\${weights}.score \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/\${gwas}/1KGPhase3.w_hm3.EUR.FUSION.\${gwas}.\${weights}.scale \
  --pheno_name \${gwas} \
  --n_cores 1 \
  --pigz ${pigz_binary} \
  --output ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/\${weights}/UKBB.w_hm3.EUR.\${weights}.\${gwas}

EOF
done

for pop in $(echo AFR AMR EAS EUR SAS);do
sbatch --array=1-$(wc -l ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo.txt | cut -d &#39; &#39; -f 1)%10 ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/todo_array.sh
done
</code></pre>
</details>
<details>
<p><summary>GeRS (coloc)</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create variable listing phenotypes and corresponding GWAS
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

# Calculate polygenic scores using 1KG reference
for pop in $(echo AFR AMR EAS EUR SAS);do
  mkdir -p ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}
  &gt; ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt
  for i in $(seq 1 2);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
    
    for weights in $(cat ~/brc_scratch/Data/TWAS_sumstats/FUSION/snp_weight_list.txt);do
      if [ ! -f ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/${weights}/UKBB.w_hm3.EUR.${weights}.${gwas_i}.fiprofile ]; then
        echo $gwas_i $pheno_i $weights &gt;&gt; ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt
      fi
    done
  done
cat &lt;&lt;EOF &gt;${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo_array.sh
#!/bin/bash
#SBATCH -p shared,brc
#SBATCH --mem 10G
#SBATCH -n 1

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

gwas=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$1}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt)
pheno=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$2}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt)
weights=\$(awk -v var=\${SLURM_ARRAY_TASK_ID} &#39;NR == var {print \$3}&#39; ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt)

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_functionally_informed_risk_scorer/scaled_functionally_informed_risk_scorer.R \
  --targ_feature_pred ${UKBB_output}/Predicted_expression/FUSION/${pop}/\${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.\${weights}.predictions.gz \
  --target_keep ${UKBB_output}/DiverseAncestry/Phenotype_subsets/UKB.\${pheno}.${pop}.QC.keep \
  --ref_score ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/\${gwas}_COLOC_PP4/1KGPhase3.w_hm3.EUR.FUSION.\${gwas}.\${weights}.score \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/\${gwas}_COLOC_PP4/1KGPhase3.w_hm3.EUR.FUSION.\${gwas}.\${weights}.scale \
  --pheno_name \${gwas} \
  --n_cores 1 \
  --pigz ${pigz_binary} \
  --output ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/\${weights}/UKBB.w_hm3.EUR.\${weights}.\${gwas}

EOF
done

for pop in $(echo EUR AFR SAS EAS AMR);do
sbatch --array=1-$(wc -l ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo.txt | cut -d &#39; &#39; -f 1)%10 ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/todo_array.sh
done
</code></pre>
</details>
<p>Note. Some GeRS are not being calculated. Investigate why.</p>
<hr />
</div>
</div>
<div id="evaluate-scores" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Evaluate scores</h3>
<details>
<p><summary>pT + clump comparison</summary></p>
<pre class="bash"><code>##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Make required directories
for pheno_i in $(echo BMI Height);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs
done

# Create a file listing the predictors files
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

for i in $(seq 1 2);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
for pop in $(echo AFR AMR EAS EUR SAS);do
cat &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups &lt;&lt;EOF
predictors 
${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.profiles
EOF
done
done

# Derive and evaluate models
pheno=$(echo BMI Height)
pheno_file=$(echo UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno)
gwas=$(echo BODY04 HEIG03)
prev=$(echo NA NA)

# pT + clump (sparse)
for i in $(seq 1 2);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d &#39; &#39;)
  gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

  for pop in $(echo AFR AMR EAS EUR SAS);do
  sbatch --mem 10G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
      --pheno ${UKBB_output}/Phenotype/${pheno_i}/${pheno_file_i} \
      --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs \
      --n_core 1 \
      --compare_predictors T \
      --assoc T \
      --outcome_pop_prev ${prev_i} \
      --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs.predictor_groups
  done
done
</code></pre>
</details>
<details>
<p><summary>Comparison of PRS methods</summary></p>
<pre class="r"><code>##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Make required directories
for pheno_i in $(echo BMI Height);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs
done

# Create a file listing the predictors files
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)

for i in $(seq 1 2);do
pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
for pop in $(echo AFR AMR EAS EUR SAS);do
cat &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs.AllMethodComp.predictor_groups &lt;&lt;EOF
predictors group
${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.profiles pt_clump
${UKBB_output}/DiverseAncestry/1KG_ref/lassosum/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.lassosum_profiles lassosum
${UKBB_output}/DiverseAncestry/1KG_ref/PRScs/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.PRScs_profiles PRScs
${UKBB_output}/DiverseAncestry/1KG_ref/SBLUP/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.SBLUP_profiles SBLUP
${UKBB_output}/DiverseAncestry/1KG_ref/SBayesR/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.SBayesR_profiles SBayesR
${UKBB_output}/DiverseAncestry/1KG_ref/LDPred/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.LDPred_profiles LDPred
${UKBB_output}/DiverseAncestry/1KG_ref/LDPred2/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.LDPred_profiles LDPred2
${UKBB_output}/DiverseAncestry/1KG_ref/DBSLMM/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.DBSLMM_profiles DBSLMM
EOF
done
done

# Derive and evaluate models
pheno=$(echo BMI Height)
pheno_file=$(echo UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno)
gwas=$(echo BODY04 HEIG03)
prev=$(echo NA NA)

for i in $(seq 1 2);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d &#39; &#39;)
  gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

  for pop in $(echo AFR AMR EAS EUR SAS);do
  sbatch --mem 20G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
      --pheno ${UKBB_output}/Phenotype/${pheno_i}/${pheno_file_i} \
      --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs.AllMethodComp \
      --n_core 1 \
      --compare_predictors T \
      --assoc T \
      --outcome_pop_prev ${prev_i} \
      --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRSs/UKBB.${pop}.w_hm3.${gwas_i}.EUR-PRSs.AllMethodComp.predictor_groups
  done
done

# Note. The nested cross validation model builder is not ideal when there are multiple parameters for each method. Previously we have identifying the best single parameter in the training sample, and then testing it out of sample. With nested cross validation, we are selecting the best pT based on results out of sample. Probably doesn&#39;t make a big diffrence, but will lead to slight inflation of R2 of single parameters PRS. I like using the full sample though, so I could add an option select best predictor within each group, rather than nested cross validate single predictors which is quite pointless. Leave as is for now.</code></pre>
</details>
<details>
<p><summary>GeRS + PRS</summary></p>
<pre class="bash"><code>##############################
# Evaluating predictive utility of GeRS and PRS individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Make required directories
for pheno_i in $(echo BMI Height);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs
done

# Create a file listing the predictors files
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)
weights=$(cat ${TWAS_rep}/snp_weight_list.txt)

for i in $(seq 1 2);do
  for pop in $(echo AFR AMR EAS EUR SAS);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
    
    echo &quot;predictors group&quot; &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
  
    for weight in ${weights}; do
      if [ -f ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/${weight}/UKBB.w_hm3.EUR.${weight}.${gwas_i}.fiprofile ]; then
        echo ${UKBB_output}/GeRS_for_diversity/1KG_ref/${pop}/${weight}/UKBB.w_hm3.EUR.${weight}.${gwas_i}.fiprofile GeRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
      fi
    done
  
      echo ${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.profiles PRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
  done
done

# Derive and evaluate models
pheno=$(echo BMI Height)
pheno_file=$(echo UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno)
gwas=$(echo BODY04 HEIG03)
prev=$(echo NA NA)

# 1KG reference
for i in $(seq 1 2);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d &#39; &#39;)
  gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

  for pop in $(echo AFR AMR EAS EUR SAS);do
sbatch --mem 10G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
    --pheno ${UKBB_output}/Phenotype/${pheno_i}/${pheno_file_i} \
    --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs.${pop}-PRSs.pt_clump \
    --n_core 1 \
    --compare_predictors T \
    --assoc T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
done
done
</code></pre>
</details>
<details>
<p><summary>GeRS (coloc) + PRS</summary></p>
<pre class="bash"><code>##############################
# Evaluating predictive utility of GeRS and PRS individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Make required directories
for pheno_i in $(echo BMI Height);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs
done

# Create a file listing the predictors files
pheno=$(echo BMI Height)
gwas=$(echo BODY04 HEIG03)
weights=$(cat ${TWAS_rep}/snp_weight_list.txt)

for i in $(seq 1 2);do
  for pop in $(echo EUR EAS AMR SAS AFR);do
    pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
    gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
    
    echo &quot;predictors group&quot; &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs_coloc.${pop}-PRSs.pt_clump.predictor_groups
  
    for weight in ${weights}; do
      if [ -f ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/${weight}/UKBB.w_hm3.EUR.${weight}.${gwas_i}.fiprofile ]; then
        echo ${UKBB_output}/GeRS_for_diversity/1KG_ref_withCOLOC/${pop}/${weight}/UKBB.w_hm3.EUR.${weight}.${gwas_i}.fiprofile GeRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs_coloc.${pop}-PRSs.pt_clump.predictor_groups
      fi
    done
  
      echo ${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${gwas_i}/UKBB.${pop}.w_hm3.${gwas_i}.profiles PRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs_coloc.${pop}-PRSs.pt_clump.predictor_groups
  done
done

# Derive and evaluate models
pheno=$(echo BMI Height)
pheno_file=$(echo UKBB_BMI.score.UpdateIDs.pheno UKBB_Height.score.UpdateIDs.pheno)
gwas=$(echo BODY04 HEIG03)
prev=$(echo NA NA)

# 1KG reference
for i in $(seq 1 2);do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d &#39; &#39;)
  pheno_file_i=$(echo ${pheno_file} | cut -f ${i} -d &#39; &#39;)
  gwas_i=$(echo ${gwas} | cut -f ${i} -d &#39; &#39;)
  prev_i=$(echo ${prev} | cut -f ${i} -d &#39; &#39;)

  for pop in $(echo AFR AMR EAS EUR SAS);do
sbatch --mem 10G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
    --pheno ${UKBB_output}/Phenotype/${pheno_i}/${pheno_file_i} \
    --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs_coloc.${pop}-PRSs.pt_clump \
    --n_core 1 \
    --compare_predictors T \
    --assoc T \
    --outcome_pop_prev ${prev_i} \
    --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno_i}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${gwas_i}.${pop}-GeRSs_coloc.${pop}-PRSs.pt_clump.predictor_groups
done
done
</code></pre>
</details>
<hr />
</div>
<div id="plot-the-results" class="section level3">
<h3><span class="header-section-number">2.2.4</span> Plot the results</h3>
<hr />
<div id="ptclump" class="section level4">
<h4><span class="header-section-number">2.2.4.1</span> pT+clump</h4>
<details>
<p><summary>pT + clump comparison</summary></p>
<pre class="r"><code>pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR&#39;,&#39;SAS&#39;)
pheno&lt;-c(&#39;BMI&#39;,&#39;Height&#39;)
gwas&lt;-c(&#39;BODY04&#39;,&#39;HEIG03&#39;)

library(data.table)
res&lt;-list()
for(i in 1:length(gwas)){
  res_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.pred_eval.txt&#39;))
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_pheno&lt;-rbind(res_pheno, tmp2)
  }

  res_pheno$Model&lt;-gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;_.*&#39;,&#39;&#39;,gsub(paste0(gwas[i],&#39;.&#39;),&#39;&#39;,res_pheno$Model)))
  res_pheno$Model&lt;-factor(res_pheno$Model, levels=unique(res_pheno$Model))
  res[[pheno[i]]]&lt;-res_pheno
}

library(ggplot2)
library(cowplot)

plot_list&lt;-NULL
for(i in 1:length(gwas)){
  plot_list[[pheno[i]]]&lt;-ggplot(res[[pheno[i]]], aes(x=factor(Model), y=R, colour=Population)) +
                            geom_point(stat=&quot;identity&quot;, position=position_dodge( width=.25)) +
                            geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.25)) +
                            theme_half_open() +
                            ylim(0,NA) +
                            theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.4, size=12)) +
                            background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                            labs(x=&#39;pT&#39;,y=&#39;Predicted-Observed Correlation&#39;, title=pheno[i])
}

png(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.pTclump_per_pT.png&#39;, units=&#39;px&#39;, res=300, width=2000, height=3000)
  plot_grid(plotlist=plot_list, ncol = 1)
dev.off()</code></pre>
</details>
<hr />
<div class="figure">
<img src="Images/UKB.Diverse.pTclump_per_pT.png" alt="pT+clump prediction across populations" />
<p class="caption">pT+clump prediction across populations</p>
</div>
<hr />
<p>Results are concordant with previous estimates, show the elastic net method improve prediction over individuals pTs, and that the optimal pT varies substantially across populations.</p>
<hr />
</div>
<div id="prs-method-comparison" class="section level4">
<h4><span class="header-section-number">2.2.4.2</span> PRS method comparison</h4>
<pre class="r"><code>source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config&#39;)

meta_res_eval&lt;-NULL
meta_res_comp&lt;-NULL
meta_res_eval_plots&lt;-list()
meta_res_eval_diff_plots&lt;-list()
meta_res_eval_diff_matrix&lt;-list()

pheno&lt;-c(&#39;Height&#39;,&#39;BMI&#39;)
gwas&lt;-c(&#39;HEIG03&#39;,&#39;BODY04&#39;)
pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR&#39;,&#39;SAS&#39;)

##################
# Organise, make tables and figures for per phenotype results
##################

res_eval&lt;-NULL
res_comp&lt;-NULL
res_eval_plots&lt;-list()
res_eval_diff_plots&lt;-list()
res_eval_diff_matrix&lt;-list()
j&lt;-1

for(k in 1:length(pop)){
  for(i in 1:length(pheno)){
    res_eval_i&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.AllMethodComp.pred_eval.txt&#39;), header=T, stringsAsFactors=F)

    res_comp_i&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,gwas[i],&#39;.EUR-PRSs.AllMethodComp.pred_comp.txt&#39;), header=T, stringsAsFactors=F)
  
    # Idenitfy model names for 10FCVal, multi-PRS and pseudoval
    selected_models&lt;-NULL
    
    ##
    # pT+clump (sparse)
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_pT&lt;-res_eval_i[grepl(&#39;pt.clump.PredFile&#39;, res_eval_i$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;pT+clump.10FCVal&#39;, Label=
    res_eval_i_per_pT[res_eval_i_per_pT$R == max(res_eval_i_per_pT$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;pT+clump.MultiPRS&#39;, Label=&#39;pt.clump&#39;))
  
    ##
    # lassosum
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_lassosum&lt;-res_eval_i[grepl(&#39;lassosum.PredFile&#39;, res_eval_i$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.10FCVal&#39;, Label=
    res_eval_i_per_lassosum[res_eval_i_per_lassosum$R == max(res_eval_i_per_lassosum$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.MultiPRS&#39;, Label=&#39;lassosum&#39;))
  
    # Identify pseudoVal
    lassosum_val&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/lassosum/&#39;,gwas[i],&#39;/1KGPhase3.w_hm3.&#39;,gwas[i],&#39;.log&#39;), sep=&#39;&amp;&#39;)
    lassosum_val_s&lt;-as.numeric(gsub(&#39;s = &#39;,&#39;&#39;,lassosum_val$V1[grepl(&#39;s = &#39;,lassosum_val$V1)]))
    lassosum_val_lambda&lt;-as.numeric(gsub(&#39;lambda =&#39;,&#39;&#39;,lassosum_val$V1[grepl(&#39;lambda =&#39;,lassosum_val$V1)]))
    lassosum_val_param&lt;-paste0(&#39;s&#39;,lassosum_val_s,&#39;.lambda&#39;,lassosum_val_lambda)
    lassosum_val_param&lt;-substr(lassosum_val_param, 1, nchar(lassosum_val_param)-1) 
  
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.PseudoVal&#39;, Label=res_eval_i[grepl(lassosum_val_param, res_eval_i$Model),]$Model[1]))
    
    ##
    # PRScs
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_PRScs&lt;-res_eval_i[grepl(&#39;PRScs.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_PRScs&lt;-res_eval_i_per_PRScs[!grepl(&#39;phiauto&#39;, res_eval_i_per_PRScs$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;PRScs.10FCVal&#39;, Label=res_eval_i_per_PRScs[res_eval_i_per_PRScs$R == max(res_eval_i_per_PRScs$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;PRScs.MultiPRS&#39;, Label=&#39;PRScs&#39;))
  
    # Identify pseudoval
    res_eval_i_per_PRScs&lt;-res_eval_i[grepl(&#39;PRScs.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_PRScs&lt;-res_eval_i_per_PRScs[grepl(&#39;phiauto&#39;, res_eval_i_per_PRScs$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;PRScs.PseudoVal&#39;, Label=res_eval_i_per_PRScs[res_eval_i_per_PRScs$R == max(res_eval_i_per_PRScs$R),]$Model[1]))
  
    ##
    # SBLUP
    ##
    
    # Identify PseudoVal
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;SBLUP.Inf&#39;, Label=&#39;SBLUP&#39;))
    
    ##
    # SBayesR
    ##
    
    # Identify PseudoVal
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;SBayesR.PseudoVal&#39;, Label=&#39;SBayesR&#39;))

    ##
    # LDPred1
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_LDPred&lt;-res_eval_i[grepl(&#39;LDPred.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_LDPred&lt;-res_eval_i_per_LDPred[!grepl(&#39;.LDpred.inf&#39;, res_eval_i_per_LDPred$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred1.10FCVal&#39;, Label=res_eval_i_per_LDPred[res_eval_i_per_LDPred$R == max(res_eval_i_per_LDPred$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred1.MultiPRS&#39;, Label=&#39;LDPred&#39;))
  
    # Identify pseudoval
    res_eval_i_per_LDPred&lt;-res_eval_i[grepl(&#39;LDPred.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_LDPred&lt;-res_eval_i_per_LDPred[grepl(&#39;.LDpred.inf&#39;, res_eval_i_per_LDPred$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred1.Inf&#39;, Label=res_eval_i_per_LDPred[res_eval_i_per_LDPred$R == max(res_eval_i_per_LDPred$R),]$Model[1]))
  
    ##
    # LDPred2
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_LDPred2&lt;-res_eval_i[grepl(&#39;LDPred2.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_LDPred2&lt;-res_eval_i_per_LDPred2[!grepl(&#39;beta.inf|beta.auto&#39;, res_eval_i_per_LDPred2$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred2.10FCVal&#39;, Label=res_eval_i_per_LDPred2[res_eval_i_per_LDPred2$R == max(res_eval_i_per_LDPred2$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred2.MultiPRS&#39;, Label=&#39;LDPred2&#39;))
  
    # Identify pseudoval
    res_eval_i_per_LDPred2&lt;-res_eval_i[grepl(&#39;LDPred2.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_LDPred2&lt;-res_eval_i_per_LDPred2[grepl(&#39;beta.inf&#39;, res_eval_i_per_LDPred2$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred2.Inf&#39;, Label=res_eval_i_per_LDPred2$Model))

    res_eval_i_per_LDPred2&lt;-res_eval_i[grepl(&#39;LDPred2.PredFile&#39;, res_eval_i$Model),]
    res_eval_i_per_LDPred2&lt;-res_eval_i_per_LDPred2[grepl(&#39;beta.auto&#39;, res_eval_i_per_LDPred2$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;LDPred2.PseudoVal&#39;, Label=res_eval_i_per_LDPred2$Model))

    ##
    # DBSLMM
    ##
    
    # Identify PseudoVal
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;DBSLMM.PseudoVal&#39;, Label=&#39;DBSLMM&#39;))
  
    ##
    # All methods
    ##
    
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;All.MultiPRS&#39;, Label=&#39;All&#39;))
  
    ###
    # Subset selected models and format
    ###
    
    # Eval
    selected_models$Label_2&lt;-paste0(gsub(&#39;_group&#39;,&#39;&#39;,selected_models$Label),&#39;_group&#39;)
    res_eval_i_select&lt;-res_eval_i[(res_eval_i$Model %in% selected_models$Label_2),]
    res_eval_i_select&lt;-merge(res_eval_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_2&#39;)], by.x=&#39;Model&#39;,by.y=&#39;Label_2&#39;)
    res_eval_i_select$Phenotype&lt;-pheno[i]
  
    res_eval_i_select&lt;-res_eval_i_select[c(&#39;Phenotype&#39;,&#39;Test&#39;,&quot;R&quot;,&quot;SE&quot;,&#39;P&#39;,&#39;R2o&#39;,&#39;N&#39;)]
                                           
    res_eval&lt;-rbind(res_eval,res_eval_i_select)
    
    # Comp
    selected_models$Label_3&lt;-gsub(&#39;_group&#39;,&#39;&#39;,selected_models$Label)
    res_comp_i_select&lt;-res_comp_i[(res_comp_i$Model_1 %in% selected_models$Label_3) &amp; (res_comp_i$Model_2 %in% selected_models$Label_3),]
    res_comp_i_select&lt;-merge(res_comp_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_3&#39;)], by.x=&#39;Model_1&#39;,by.y=&#39;Label_3&#39;)
    names(res_comp_i_select)[names(res_comp_i_select) == &#39;Test&#39;]&lt;-&#39;Model_1_Test&#39;
    res_comp_i_select&lt;-merge(res_comp_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_3&#39;)], by.x=&#39;Model_2&#39;,by.y=&#39;Label_3&#39;)
    names(res_comp_i_select)[names(res_comp_i_select) == &#39;Test&#39;]&lt;-&#39;Model_2_Test&#39;
    res_comp_i_select$Label&lt;-NULL
    res_comp_i_select$Phenotype&lt;-pheno[i]
    res_comp_i_select&lt;-res_comp_i_select[c(&#39;Phenotype&#39;,&#39;Model_1_Test&#39;,&#39;Model_2_Test&#39;,&#39;Model_1_R&#39;,&#39;Model_2_R&#39;,&#39;R_diff&#39;,&#39;R_diff_pval&#39;)]
  
    res_comp&lt;-rbind(res_comp,res_comp_i_select)
    
    ###
    # Create plot showing performance of each method
    ###
    
    library(ggplot2)
    library(cowplot)
    
    res_eval_i_select$Method&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;,res_eval_i_select$Test)
    res_eval_i_select$Model&lt;-gsub(&#39;.*\\.&#39;,&#39;&#39;,res_eval_i_select$Test)
    
    res_eval_i_select$Method&lt;-factor(res_eval_i_select$Method, levels=c(&#39;pT+clump&#39;,&#39;lassosum&#39;,&#39;PRScs&#39;,&#39;SBLUP&#39;,&#39;SBayesR&#39;,&#39;LDPred1&#39;,&#39;LDPred2&#39;,&#39;DBSLMM&#39;,&#39;All&#39;))
    res_eval_i_select$Model&lt;-factor(res_eval_i_select$Model, levels=c(&#39;10FCVal&#39;,&#39;MultiPRS&#39;,&#39;PseudoVal&#39;,&#39;Inf&#39;))
    
    res_eval_plots[[j]] &lt;- ggplot( res_eval_i_select, aes(x=Method, y=R, fill=Model)) +
                          geom_bar(stat=&quot;identity&quot;, position=position_dodge(preserve = &quot;single&quot;), width = 0.7) +
                          geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(width = 0.7, preserve = &quot;single&quot;)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;)) +
                          coord_cartesian(ylim=c(min(res_eval_i_select$R[res_eval_i_select$Phenotype==pheno[i]])-0.02, max(res_eval_i_select$R[res_eval_i_select$Phenotype==pheno[i]])+0.025), clip=&quot;on&quot;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                          theme(legend.position=&quot;top&quot;, legend.title = element_blank(), legend.box=&quot;vertical&quot;, legend.margin=margin()) +
                          guides(fill=guide_legend(nrow=2))
                                  
    ###
    # Create plot showing performance of each method compared to pT+clump
    ###
  
    res_eval_i_select$R_diff&lt;-res_eval_i_select$R-res_eval_i_select$R[res_eval_i_select$Test == &quot;pT+clump.10FCVal&quot;]
    
    res_eval_diff_plots[[j]] &lt;- ggplot( res_eval_i_select, aes(x=Method, y=R_diff, fill=Model)) +
                          geom_bar(stat=&quot;identity&quot;, position=position_dodge(preserve = &quot;single&quot;), width = 0.7) +
                          geom_errorbar(aes(ymin=R_diff-SE, ymax=R_diff+SE), width=.2, position=position_dodge(width = 0.7, preserve = &quot;single&quot;)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;)) +
                          coord_cartesian(ylim=c(min(res_eval_i_select$R_diff[res_eval_i_select$Phenotype==pheno[i]])-0.02, max(res_eval_i_select$R_diff[res_eval_i_select$Phenotype==pheno[i]])+0.025), clip=&quot;on&quot;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                          theme(legend.position=&quot;top&quot;, legend.title = element_blank(), legend.box=&quot;vertical&quot;, legend.margin=margin()) +
                          guides(fill=guide_legend(nrow=2))
    
  
    ###
    # Create plot showing matrix of differences with significance
    ###
  
    library(reshape2)
    
    res_comp_i_select$R_diff[res_comp_i_select$Model_1_Test == res_comp_i_select$Model_2_Test]&lt;-NA
  
    res_comp_i_select$R_diff_catagory&lt;-&#39;NA&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.002]&lt;-&#39;-0.025 - -0.002&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.025]&lt;-&#39;-0.08 - -0.025&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.08]&lt;-&#39;&lt; -0.08&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; -0.002 &amp; res_comp_i_select$R_diff &lt; 0.002]&lt;-&#39;-0.002 - 0.002&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.002]&lt;-&#39;0.002 - 0.025&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.025]&lt;-&#39;0.025 - 0.08&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.08]&lt;-&#39;&gt; 0.08&#39;
    
    res_comp_i_select$R_diff_catagory&lt;-factor(res_comp_i_select$R_diff_catagory, level=rev(c(&#39;&lt; -0.08&#39;,&#39;-0.08 - -0.025&#39;,&#39;-0.025 - -0.002&#39;,&#39;-0.002 - 0.002&#39;,&#39;0.002 - 0.025&#39;,&#39;0.025 - 0.08&#39;,&#39;&gt; 0.08&#39;)))
  
    res_comp_i_select$star&lt;-&#39; &#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 0.05]&lt;-&#39;*&#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 1e-3]&lt;-&#39;**&#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 1e-6]&lt;-&#39;***&#39;
  
    
    library(RColorBrewer)
    res_eval_diff_matrix[[j]]&lt;-ggplot(data = res_comp_i_select, aes(Model_2_Test, Model_1_Test, fill=R_diff_catagory))+
     geom_tile(color = &quot;white&quot;)+
     labs(y=&#39;Test&#39;, x=&#39;Comparison&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;), fill=&#39;R difference&#39;) +
      theme_minimal()+ 
     theme(axis.text.x = element_text(angle = 45, vjust = 1, 
        size = 8, hjust = 1))+
     coord_fixed() +
    geom_text(data=res_comp_i_select, aes(Model_2_Test, Model_1_Test, label = star), color = &quot;black&quot;, size = 4, angle = 0, vjust=0.8) +
    scale_fill_brewer(breaks = levels(res_comp_i_select$R_diff_catagory), palette = &quot;RdBu&quot;, drop=F, na.value=&#39;grey&#39;)

    j&lt;-j+1
  }
}

write.csv(res_eval, paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp.csv&#39;), row.names=F, quote=F)

write.csv(res_comp, paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_diff.csv&#39;), row.names=F, quote=F)

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1000*ceiling(length(res_eval_plots)/2))
print(plot_grid(plotlist=res_eval_plots, ncol = 2))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_pTDiff.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1000*ceiling(length(res_eval_plots)/2))
print(plot_grid(plotlist=res_eval_diff_plots, ncol = 2))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_diff.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1750*length(res_eval_diff_matrix))
print(plot_grid(plotlist=res_eval_diff_matrix, ncol = 1))
dev.off()

# With only two phenotypes, meta-analysis is not necessary or helpful.</code></pre>
</details>
</div>
<div id="gers-vs.ptclump" class="section level4">
<h4><span class="header-section-number">2.2.4.3</span> GeRS vs. pT+clump</h4>
<details>
<p><summary>GeRS and pT+clump comparison</summary></p>
<pre class="r"><code>pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR&#39;,&#39;SAS&#39;)
pheno&lt;-c(&#39;BMI&#39;,&#39;Height&#39;)
gwas&lt;-c(&#39;BODY04&#39;,&#39;HEIG03&#39;)

library(data.table)
res&lt;-list()
for(i in 1:length(gwas)){
  res_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.&#39;,gwas[i],&#39;.&#39;,pop[k],&#39;-GeRSs.&#39;,pop[k],&#39;-PRSs.pt_clump.pred_eval.txt&#39;))
    tmp&lt;-tmp[!grepl(&#39;PredFile&#39;, tmp$Model),]
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_pheno&lt;-rbind(res_pheno, tmp2)
  }

  res_pheno$Model&lt;-gsub(&#39;_group&#39;,&#39;&#39;,res_pheno$Model)
  res_pheno$Model&lt;-factor(res_pheno$Model, levels=unique(res_pheno$Model))
  res[[pheno[i]]]&lt;-res_pheno
}

library(ggplot2)
library(cowplot)

plot_list&lt;-NULL
for(i in 1:length(gwas)){
  plot_list[[pheno[i]]]&lt;-ggplot(res[[pheno[i]]], aes(x=factor(Model), y=R, colour=Population)) +
                            geom_point(stat=&quot;identity&quot;, position=position_dodge( width=.30)) +
                            geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.30)) +
                            theme_half_open() +
                            theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.4, size=12)) +
                            background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                            labs(x=NULL,y=&quot;Correlation (SE)&quot;, title=pheno[i])
  
  if(i != length(gwas)){
    plot_list[[pheno[i]]]&lt;-plot_list[[pheno[i]]] + theme(legend.position = &quot;none&quot;)
  }
}

bitmap(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.GeRS_vs_pTclump.png&#39;, units=&#39;px&#39;, res=300, width=2250, height=900)
  plot_grid(plotlist=plot_list, nrow = 1, rel_widths = c(1.3/3, 1.7/3))
dev.off()

res_2&lt;-list()
for(i in 1:length(gwas)){
  res_2_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.&#39;,gwas[i],&#39;.&#39;,pop[k],&#39;-GeRSs.&#39;,pop[k],&#39;-PRSs.pt_clump.pred_comp.txt&#39;))
    tmp&lt;-tmp[tmp$Model_1 == &#39;All&#39; &amp; tmp$Model_2 == &#39;PRS&#39;,]
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_2_pheno&lt;-rbind(res_2_pheno, tmp2)
  }

  res_2_pheno$Model_1&lt;-&#39;GeRS + PRS&#39;
  res_2_pheno$Model_2&lt;-&#39;PRS only&#39;
  res_2[[pheno[i]]]&lt;-res_2_pheno
}

res_2_all&lt;-do.call(rbind, res_2)
write.csv(res_2_all, &#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.GeRS_vs_pTclump.csv&#39;, row.names=F, quote=F)</code></pre>
</details>
<hr />
<div class="figure">
<img src="Images/UKB.Diverse.GeRS_vs_pTclump.png" alt="GeRS vs. PRS across populations" />
<p class="caption">GeRS vs. PRS across populations</p>
</div>
<table>
<caption>Difference between GeRS+PRS model and PRS only model</caption>
<thead>
<tr class="header">
<th align="left">Population</th>
<th align="left">Phenotype</th>
<th align="left">Model_1</th>
<th align="left">Model_2</th>
<th align="right">Model_1_R</th>
<th align="right">Model_2_R</th>
<th align="right">R_diff</th>
<th align="left">R_diff_pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AFR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.166</td>
<td align="right">0.164</td>
<td align="right">0.002</td>
<td align="left">6.81e-01</td>
</tr>
<tr class="even">
<td align="left">AMR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.121</td>
<td align="right">0.117</td>
<td align="right">0.004</td>
<td align="left">8.37e-01</td>
</tr>
<tr class="odd">
<td align="left">EAS</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.266</td>
<td align="right">0.198</td>
<td align="right">0.068</td>
<td align="left">1.96e-06</td>
</tr>
<tr class="even">
<td align="left">EUR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.276</td>
<td align="right">0.271</td>
<td align="right">0.005</td>
<td align="left">1.02e-07</td>
</tr>
<tr class="odd">
<td align="left">SAS</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.207</td>
<td align="right">0.206</td>
<td align="right">0.001</td>
<td align="left">6.55e-01</td>
</tr>
<tr class="even">
<td align="left">AFR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.142</td>
<td align="right">0.148</td>
<td align="right">-0.006</td>
<td align="left">2.72e-01</td>
</tr>
<tr class="odd">
<td align="left">AMR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.275</td>
<td align="right">0.205</td>
<td align="right">0.070</td>
<td align="left">4.12e-04</td>
</tr>
<tr class="even">
<td align="left">EAS</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.241</td>
<td align="right">0.211</td>
<td align="right">0.031</td>
<td align="left">1.39e-02</td>
</tr>
<tr class="odd">
<td align="left">EUR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.328</td>
<td align="right">0.322</td>
<td align="right">0.006</td>
<td align="left">1.47e-12</td>
</tr>
<tr class="even">
<td align="left">SAS</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.257</td>
<td align="right">0.255</td>
<td align="right">0.002</td>
<td align="left">5.43e-01</td>
</tr>
</tbody>
</table>
<hr />
<p>GeRS significantly improves cross ancestry prediction for BMI in EAS, and Height in EAS and AMR. Relative impovement between 10-20%.</p>
<p>Investigate individual associations to determine what is driving the improved prediction.</p>
<hr />
<pre class="r"><code>pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR&#39;,&#39;SAS&#39;)
pheno&lt;-c(&#39;BMI&#39;,&#39;Height&#39;)
gwas&lt;-c(&#39;BODY04&#39;,&#39;HEIG03&#39;)

library(data.table)
res&lt;-list()
for(i in 1:length(gwas)){
  res_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.&#39;,gwas[i],&#39;.&#39;,pop[k],&#39;-GeRSs.&#39;,pop[k],&#39;-PRSs.pt_clump.assoc.txt&#39;))
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_pheno&lt;-rbind(res_pheno, tmp2)
  }

  res[[pheno[i]]]&lt;-res_pheno
}

res_bmi_eas&lt;-res[[&#39;BMI&#39;]][res[[&#39;BMI&#39;]]$Population == &#39;EAS&#39;,]
res_bmi_eas[order(res_bmi_eas$P),]
# GeRS explain more variance than PRS. PredFile 34 (CMC.BRAIN.RNASEQ) and 23 (METSIM ADIPOSE).

pred_group&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/BMI/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.BODY04.EAS-GeRSs.EAS-PRSs.pt_clump.predictor_groups&#39;)
pred_group[c(23,34),]
# The GeRS correpond to CMC.BRAIN.RNASEQ and METSIM ADIPOSE. These are two of the best powered eqtl dataset but they are also relevent to the phenotype.

res_height_amr&lt;-res[[&#39;Height&#39;]][res[[&#39;Height&#39;]]$Population == &#39;AMR&#39;,]
res_height_amr[order(res_height_amr$P),]
# GeRS explain more variance than PRS. PredFile 4, 6, 36

pred_group&lt;-fread(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/Height/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.HEIG03.AMR-GeRSs.AMR-PRSs.pt_clump.predictor_groups&#39;)
pred_group[c(4,6,36),]
# The GeRS correpond to Artery_Aorta, Artery_Tibial, Muscle_Skeletal.</code></pre>
</div>
<div id="gers-coloc-vs.ptclump" class="section level4">
<h4><span class="header-section-number">2.2.4.4</span> GeRS (coloc) vs. pT+clump</h4>
<details>
<p><summary>GeRS (coloc) and pT+clump comparison</summary></p>
<pre class="r"><code>pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR&#39;,&#39;SAS&#39;)
pheno&lt;-c(&#39;BMI&#39;,&#39;Height&#39;)
gwas&lt;-c(&#39;BODY04&#39;,&#39;HEIG03&#39;)

library(data.table)
res&lt;-list()
for(i in 1:length(gwas)){
  res_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.&#39;,gwas[i],&#39;.&#39;,pop[k],&#39;-GeRSs_coloc.&#39;,pop[k],&#39;-PRSs.pt_clump.pred_eval.txt&#39;))
    tmp&lt;-tmp[!grepl(&#39;PredFile&#39;, tmp$Model),]
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_pheno&lt;-rbind(res_pheno, tmp2)
  }

  res_pheno$Model&lt;-gsub(&#39;_group&#39;,&#39;&#39;,res_pheno$Model)
  res_pheno$Model&lt;-factor(res_pheno$Model, levels=unique(res_pheno$Model))
  res[[pheno[i]]]&lt;-res_pheno
}

library(ggplot2)
library(cowplot)

plot_list&lt;-NULL
for(i in 1:length(gwas)){
  plot_list[[pheno[i]]]&lt;-ggplot(res[[pheno[i]]], aes(x=factor(Model), y=R, colour=Population)) +
                            geom_point(stat=&quot;identity&quot;, position=position_dodge( width=.30)) +
                            geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.30)) +
                            theme_half_open() +
                            theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.4, size=12)) +
                            background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                            labs(x=&#39;pT&#39;,y=&quot;Correlation (SE)&quot;, title=pheno[i])
  
  if(i != length(gwas)){
    plot_list[[pheno[i]]]&lt;-plot_list[[pheno[i]]] + theme(legend.position = &quot;none&quot;)
  }
}

png(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.GeRS_coloc_vs_pTclump.png&#39;, units=&#39;px&#39;, res=300, width=1750, height=600)
  plot_grid(plotlist=plot_list, nrow = 1, rel_widths = c(1.2/3, 1.8/3))
dev.off()

res_2&lt;-list()
for(i in 1:length(gwas)){
  res_2_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.&#39;,gwas[i],&#39;.&#39;,pop[k],&#39;-GeRSs_coloc.&#39;,pop[k],&#39;-PRSs.pt_clump.pred_comp.txt&#39;))
    tmp&lt;-tmp[tmp$Model_1 == &#39;All&#39; &amp; tmp$Model_2 == &#39;PRS&#39;,]
    tmp2&lt;-data.frame( Population=pop[k],
                      Phenotype=pheno[i],
                      tmp)
    
    res_2_pheno&lt;-rbind(res_2_pheno, tmp2)
  }

  res_2_pheno$Model_1&lt;-&#39;GeRS + PRS&#39;
  res_2_pheno$Model_2&lt;-&#39;PRS only&#39;
  res_2[[pheno[i]]]&lt;-res_2_pheno
}

res_2_all&lt;-do.call(rbind, res_2)
write.csv(res_2_all, &#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.GeRS_coloc_vs_pTclump.csv&#39;, row.names=F, quote=F)</code></pre>
</details>
<hr />
<div class="figure">
<img src="Images/UKB.Diverse.GeRS_coloc_vs_pTclump.png" alt="GeRS (coloc) vs. PRS across populations" />
<p class="caption">GeRS (coloc) vs. PRS across populations</p>
</div>
<table>
<caption>Difference between GeRS (coloc) + PRS model and PRS only model</caption>
<thead>
<tr class="header">
<th align="left">Population</th>
<th align="left">Phenotype</th>
<th align="left">Model_1</th>
<th align="left">Model_2</th>
<th align="right">Model_1_R</th>
<th align="right">Model_2_R</th>
<th align="right">R_diff</th>
<th align="left">R_diff_pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">AFR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.177</td>
<td align="right">0.179</td>
<td align="right">-0.003</td>
<td align="left">1.40e-01</td>
</tr>
<tr class="even">
<td align="left">AMR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.141</td>
<td align="right">0.138</td>
<td align="right">0.003</td>
<td align="left">7.71e-01</td>
</tr>
<tr class="odd">
<td align="left">EAS</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.277</td>
<td align="right">0.217</td>
<td align="right">0.061</td>
<td align="left">6.48e-05</td>
</tr>
<tr class="even">
<td align="left">EUR</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.277</td>
<td align="right">0.277</td>
<td align="right">0.000</td>
<td align="left">7.58e-01</td>
</tr>
<tr class="odd">
<td align="left">SAS</td>
<td align="left">BMI</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.224</td>
<td align="right">0.229</td>
<td align="right">-0.005</td>
<td align="left">1.16e-02</td>
</tr>
<tr class="even">
<td align="left">AFR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.140</td>
<td align="right">0.143</td>
<td align="right">-0.003</td>
<td align="left">5.79e-01</td>
</tr>
<tr class="odd">
<td align="left">AMR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.252</td>
<td align="right">0.218</td>
<td align="right">0.034</td>
<td align="left">6.69e-02</td>
</tr>
<tr class="even">
<td align="left">EAS</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.223</td>
<td align="right">0.216</td>
<td align="right">0.007</td>
<td align="left">5.56e-01</td>
</tr>
<tr class="odd">
<td align="left">EUR</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.329</td>
<td align="right">0.323</td>
<td align="right">0.006</td>
<td align="left">3.83e-07</td>
</tr>
<tr class="even">
<td align="left">SAS</td>
<td align="left">Height</td>
<td align="left">GeRS + PRS</td>
<td align="left">PRS only</td>
<td align="right">0.257</td>
<td align="right">0.255</td>
<td align="right">0.002</td>
<td align="left">2.51e-01</td>
</tr>
</tbody>
</table>
<hr />
<p>Colocalisation does not improve portability of GeRS across populations.</p>
<hr />
</div>
</div>
</div>
<div id="using-internal-gwas-sumstats" class="section level2">
<h2><span class="header-section-number">2.3</span> Using internal GWAS sumstats</h2>
<hr />
<p>I have performed GWAS for 17 traits in a training subset of EUR individuals in UKB. Perform polygenic scoring for each of these GWAS, and compares portability.</p>
<div id="run-twas" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Run TWAS</h3>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>mkdir /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS

&gt; /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/todo.txt

for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
if [ ! -f /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/${pheno}_withCOLOC/${pheno}_res_GW.txt ]; then
echo $pheno /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/GWAS/${pheno}/UKBB.w_hm3.QCd.AllSNP.${pheno}.GW.qassoc.clean.QC.sumstats.gz &gt;&gt; /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/todo.txt
fi
done

sbatch -p brc,shared --mem 10G -n 4 --array 1-$(wc -l /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/todo.txt | cut -d&#39; &#39; -f1)%3 /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Run_TWAS/Run_TWAS_withCOLOC.sh \
  --gwas_list /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/todo.txt \
  --pos /users/k1806347/brc_scratch/Data/TWAS_sumstats/FUSION/All_tissues.pos \
  --outdir /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS \
  --fusion_dir ${FUSION_dir} \
  --ncores 4
</code></pre>
</details>
</div>
<div id="create-score-files" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Create score files</h3>
<details>
<p><summary>GeRS</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

######
# Calculate for Largest_TWAS_selected_75comp.txt
######
mkdir -p /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS
&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/todo.txt
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
for weights in $(cat ${TWAS_rep}/snp_weight_list.txt);do
if [ ! -f /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/${pheno}/1KGPhase3.w_hm3.EUR.FUSION.${pheno}.${weights}.scale ]; then
echo $pheno $weights &gt;&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/todo.txt
fi
done
done

cat &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p brc,shared
#SBATCH -J GeRS
#SBATCH --mem 5G

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/todo.txt)

weights=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/todo.txt)

echo ${pheno}
echo ${weights}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/ref_funtionally_informed_risk_scorer/ref_funtionally_informed_risk_scorer.R \
  --twas_results /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/TWAS/${pheno}_withCOLOC/${pheno}_res_GW.txt \
  --ref_feature_pred ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.predictions.gz \
  --output /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/${pheno}/1KGPhase3.w_hm3.EUR.FUSION.${pheno}.${weights} \
  --ref_keep ${Geno_1KG_dir}/keep_files/EUR_samples.keep \
  --ref_scale ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.EUR.scale \
  --panel ${weights}

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/todo.txt | cut -d&#39; &#39; -f1)%40 /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>pT + clump: Sparse</summary></p>
<pre class="bash"><code>pop1=EUR
&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_todo.txt
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do

if [ ! -f /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/${pheno}/1KGPhase3.w_hm3.${pop1}.pt_clump.${pheno}.EUR.scale ]; then
${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_todo.txt
fi
done

cat &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH -J pt_clump

pop1=EUR
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_todo.txt)

echo ${pop1}
echo ${pheno}

  /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
    --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
    --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop1}_samples.keep \
    --sumstats /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/GWAS/${pheno}/UKBB.w_hm3.QCd.AllSNP.${pheno}.GW.qassoc.clean.QC.gz \
    --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
    --output /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/${pheno}/1KGPhase3.w_hm3.${pop1}.pt_clump.${pheno} \
    --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_todo.txt | cut -d&#39; &#39; -f1)%20 /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>lassosum</summary></p>
<pre class="bash"><code>pop1=EUR
mkdir -p /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum
&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_todo.txt
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
if [ ! -f /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/${pheno}/1KGPhase3.w_hm3.${pop1}.lassosum.${pheno}.EUR.scale ]; then
echo ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_todo.txt
fi
done

cat &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH --mem 10G
#SBATCH -J lassosum

pop1=EUR
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_todo.txt)

echo ${pop1}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_lassosum/polygenic_score_file_creator_lassosum.R \
  --ref_plink_gw /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop1}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/GWAS/${pheno}/UKBB.w_hm3.QCd.AllSNP.${pheno}.GW.qassoc.clean.QC.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --output /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/${pheno}/1KGPhase3.w_hm3.${pop1}.lassosum.${pheno} \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_todo.txt | cut -d&#39; &#39; -f1)%10 /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>PRScs</summary></p>
<pre class="bash"><code>pop1=EUR
mkdir -p /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs
&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_todo.txt
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
if [ ! -f /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/${pheno}/1KGPhase3.w_hm3.${pop1}.PRScs.${pheno}.EUR.scale ]; then
echo ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_todo.txt
fi
done

cat &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH --mem 110G
#SBATCH -n 11
#SBATCH -t 48:00:00
#SBATCH -J PRScs

pop1=EUR
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_todo.txt)

echo ${pop1}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_PRScs/polygenic_score_file_creator_PRScs.R \
  --ref_plink_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.chr \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop1}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/GWAS/${pheno}/UKBB.w_hm3.QCd.AllSNP.${pheno}.GW.qassoc.clean.QC.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --output /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/${pheno}/1KGPhase3.w_hm3.${pop1}.PRScs.${pheno} \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list \
  --PRScs_path /users/k1806347/brc_scratch/Software/PRScs.sh \
  --PRScs_ref_path /users/k1806347/brc_scratch/Software/PRScs/ldblk_1kg_eur \
  --n_cores 11 \
  --phi_param 1e-6,1e-4,1e-2,1,auto

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_todo.txt | cut -d&#39; &#39; -f1)%4 /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/PRScs/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>SBayesR</summary></p>
<pre class="bash"><code>pop1=EUR
mkdir -p /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR
&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_todo.txt
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
if [ ! -f /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/${pheno}/1KGPhase3.w_hm3.${pop1}.SBayesR.${pheno}.EUR.scale ]; then
echo ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_todo.txt
fi
done

cat &gt; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p shared,brc
#SBATCH --mem 50G
#SBATCH -n 6
#SBATCH -J SBayesR

pop1=EUR
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_todo.txt)

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

echo ${pop1}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_SBayesR/polygenic_score_file_creator_SBayesR.R \
  --ref_plink /users/k1806347/brc_scratch/Data/1KG/Phase3/1KGPhase3.w_hm3.GW \
  --ref_keep /users/k1806347/brc_scratch/Data/1KG/Phase3/keep_files/${pop1}_samples.keep \
  --sumstats /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/GWAS/${pheno}/UKBB.w_hm3.QCd.AllSNP.${pheno}.GW.qassoc.clean.QC.gz \
  --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
  --gctb ${gctb_203} \
  --ld_matrix_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/LD_matrix/GCTB_shared/ukbEURu_hm3_shrunk_sparse/ukbEURu_hm3_v3_50k_chr \
  --output /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/${pheno}/1KGPhase3.w_hm3.${pop1}.SBayesR.${pheno} \
  --ref_pop_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/super_pop_keep.list \
  --n_cores 6 \
  --memory 50000

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_todo.txt | cut -d&#39; &#39; -f1)%4 /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/score_sbatch.sh
</code></pre>
</details>
</div>
<div id="calculate-scores-1" class="section level3">
<h3><span class="header-section-number">2.3.3</span> Calculate scores</h3>
<details>
<p><summary>GeRS</summary></p>
<pre class="bash"><code>mkdir -p /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/

pop1=EUR
&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt
for pop2 in $(echo EUR_test EAS AMR AFR SAS); do
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do
for weights in $(cat ~/brc_scratch/Data/TWAS_sumstats/FUSION/snp_weight_list.txt);do
if [ ! -f /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/${pheno}/UKBB.w_hm3.${pop2}.${pheno}.${weights}.fiprofile ]; then
echo ${pop2} ${pheno} ${weights} &gt;&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt
fi
done
done
done

cat &gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p brc,shared
#SBATCH -J GeRS
#SBATCH --mem 15G

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

pop1=EUR
pop2=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt)
weights=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $3}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt)

pop2_short=${pop2}
if echo &quot;$pop2&quot; | grep -q &quot;EUR_test&quot;; then
    pop2_short=EUR
fi  

echo ${pop1}
echo ${pop2}
echo ${pop2_short}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_functionally_informed_risk_scorer/scaled_functionally_informed_risk_scorer.R \
  --targ_feature_pred ${UKBB_output}/Predicted_expression/FUSION/${pop2}/${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.${weights}.predictions.gz \
  --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop2}/${pheno}_resid.pheno \
  --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/${pheno}/1KGPhase3.w_hm3.EUR.FUSION.${pheno}.${weights}.score \
  --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/GeRS/${pheno}/1KGPhase3.w_hm3.EUR.FUSION.${pheno}.${weights}.scale \
  --pheno_name ${pheno} \
  --n_cores 1 \
  --pigz ${pigz_binary} \
  --output /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/${pheno}/UKBB.w_hm3.${pop2}.${pheno}.${weights}

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_todo.txt | cut -d&#39; &#39; -f1)%20 /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/score_sbatch.sh

# Note, standardise all predictors when running model_builder as GeRS are scaled to EUR population. But gene expression is standardised.</code></pre>
</details>
<details>
<p><summary>pT + clump: Sparse</summary></p>
<pre class="bash"><code>pop1=EUR
&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_todo.txt
for pop2 in $(echo EUR_test EAS AMR AFR SAS); do
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do

if [ ! -f /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}.profiles ]; then
echo ${pop2} ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_todo.txt
fi
done
done

cat &gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p brc,shared
#SBATCH -J pt_clump

pop1=EUR
pop2=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_todo.txt)

pop2_short=${pop2}
if echo &quot;$pop2&quot; | grep -q &quot;EUR_test&quot;; then
    pop2_short=EUR
fi  

echo ${pop1}
echo ${pop2}
echo ${pop2_short}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R \
            --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
            --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop2}/${pheno}_resid.pheno \
            --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/${pheno}/1KGPhase3.w_hm3.${pop1}.pt_clump.${pheno} \
            --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/pt_clump/${pheno}/1KGPhase3.w_hm3.${pop1}.pt_clump.${pheno}.${pop2_short}.scale \
            --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/${pop2_short}/1KGPhase3.w_hm3.${pop2_short}.chr \
            --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
            --output /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_todo.txt | cut -d&#39; &#39; -f1)%20 /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/pt_clump/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>lassosum</summary></p>
<pre class="bash"><code>pop1=EUR
&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_todo.txt
for pop2 in $(echo EUR_test EAS AMR AFR SAS); do
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do

if [ ! -f /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}.lassosum_profiles ]; then
echo ${pop2} ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_todo.txt
fi
done
done

cat &gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p brc,shared
#SBATCH -J lassosum
#SBATCH -n 5
#SBATCH --mem 5G

pop1=EUR
pop2=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_todo.txt)

pop2_short=${pop2}
if echo &quot;$pop2&quot; | grep -q &quot;EUR_test&quot;; then
    pop2_short=EUR
fi  

echo ${pop1}
echo ${pop2}
echo ${pop2_short}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_lassosum/Scaled_polygenic_scorer_lassosum.R \
            --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
            --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop2}/${pheno}_resid.pheno \
            --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/${pheno}/1KGPhase3.w_hm3.${pop1}.lassosum.${pheno} \
            --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/lassosum/${pheno}/1KGPhase3.w_hm3.${pop1}.lassosum.${pheno}.${pop2_short}.scale \
            --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/${pop2_short}/1KGPhase3.w_hm3.${pop2_short}.chr \
            --n_cores 5 \
            --pheno_name ${pheno} \
            --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
            --output /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_todo.txt | cut -d&#39; &#39; -f1)%6 /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/lassosum/score_sbatch.sh
</code></pre>
</details>
<details>
<p><summary>SBayesR</summary></p>
<pre class="bash"><code>pop1=EUR
mkdir /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR
&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_todo.txt
for pop2 in $(echo EUR_test EAS AMR AFR SAS); do
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt); do

if [ ! -f /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}.SBayesR_profiles ]; then
echo ${pop2} ${pheno} &gt;&gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_todo.txt
fi
done
done

cat &gt; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_sbatch.sh &lt;&lt; &#39;EOF&#39;
#!/bin/sh

#SBATCH -p brc,shared
#SBATCH -J SBayesR
#SBATCH --mem 10G

pop1=EUR
pop2=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $1}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_todo.txt)
pheno=$(awk -v var=&quot;$SLURM_ARRAY_TASK_ID&quot; &#39;NR == var {print $2}&#39; /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_todo.txt)

pop2_short=${pop2}
if echo &quot;$pop2&quot; | grep -q &quot;EUR_test&quot;; then
    pop2_short=EUR
fi  

echo ${pop1}
echo ${pop2}
echo ${pop2_short}
echo ${pheno}

/users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_SBayesR/Scaled_polygenic_scorer_SBayesR.R \
            --target_plink_chr /users/k1806347/brc_scratch/Data/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
            --target_keep /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop2}/${pheno}_resid.pheno \
            --ref_score /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/${pheno}/GWAS_sumstats_SBayesR.GW.snpRes \
            --ref_scale /users/k1806347/brc_scratch/Data/1KG/Phase3/DiverseAncestry/score_files/SBayesR/${pheno}/1KGPhase3.w_hm3.${pop1}.SBayesR.${pheno}.${pop2_short}.scale \
            --ref_freq_chr /users/k1806347/brc_scratch/Data/1KG/Phase3/freq_files/${pop2_short}/1KGPhase3.w_hm3.${pop2_short}.chr \
            --pheno_name ${pheno} \
            --plink /users/k1806347/brc_scratch/Software/plink1.9.sh \
            --output /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/${pop2}/${pheno}/UKBB.${pop2}.w_hm3.${pheno}

EOF

sbatch --array 1-$(wc -l /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_todo.txt | cut -d&#39; &#39; -f1)%6 /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/SBayesR/score_sbatch.sh
</code></pre>
</details>
</div>
<div id="evaluate-scores-1" class="section level3">
<h3><span class="header-section-number">2.3.4</span> Evaluate scores</h3>
<details>
<p><summary>pT + clump comparison</summary></p>
<pre class="bash"><code>##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs
for pop in $(echo AFR AMR EAS EUR_test SAS);do
cat &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs.predictor_groups &lt;&lt;EOF
predictors 
${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${pheno}/UKBB.${pop}.w_hm3.${pheno}.profiles
EOF
done
done

# pT + clump (sparse)
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
  for pop in $(echo AFR AMR EAS EUR_test SAS);do
  sbatch --mem 10G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
      --pheno /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop}/${pheno}_resid.pheno \
      --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs \
      --n_core 1 \
      --compare_predictors T \
      --assoc T \
      --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs.predictor_groups
  done
done
</code></pre>
</details>
<details>
<p><summary>Comparison of PRS methods</summary></p>
<pre class="r"><code>##############################
# Evaluating predictive utility of pT + clump PRSs across multiple pTs individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
  for pop in $(echo AFR AMR EAS EUR_test SAS);do
cat &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs.AllMethodComp.predictor_groups &lt;&lt;EOF
predictors group
${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${pheno}/UKBB.${pop}.w_hm3.${pheno}.profiles pt_clump
${UKBB_output}/DiverseAncestry/1KG_ref/lassosum/${pop}/${pheno}/UKBB.${pop}.w_hm3.${pheno}.lassosum_profiles lassosum
${UKBB_output}/DiverseAncestry/1KG_ref/SBayesR/${pop}/${pheno}/UKBB.${pop}.w_hm3.${pheno}.SBayesR_profiles SBayesR
EOF
done
done

# Derive and evaluate models
for pheno in $(head -n 5 /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
  for pop in $(echo AFR AMR EAS EUR_test SAS);do
  sbatch --mem 20G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
      --pheno /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop}/${pheno}_resid.pheno \
      --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs.AllMethodComp \
      --n_core 1 \
      --compare_predictors T \
      --assoc T \
      --outcome_pop_prev ${prev_i} \
      --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRSs/UKBB.${pop}.w_hm3.${pheno}.EUR-PRSs.AllMethodComp.predictor_groups
  done
done</code></pre>
</details>
<details>
<p><summary>GeRS + PRS</summary></p>
<pre class="bash"><code>##############################
# Evaluating predictive utility of GeRS and PRS individually and in combination
##############################
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Make required directories
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
mkdir -p /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs
done

# Create a file listing the predictors files
weights=$(cat ${TWAS_rep}/snp_weight_list.txt)

for pheno in $(head -n 5 /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
  for pop in $(echo AFR AMR EAS EUR_test SAS);do
    
    echo &quot;predictors group&quot; &gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${pheno}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
  
    for weight in ${weights}; do
      if [ -f /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/${pheno}/UKBB.w_hm3.${pop}.${pheno}.${weight}.fiprofile ]; then
        echo /users/k1806347/brc_scratch/Data/UKBB/DiverseAncestry/1KG_ref/GeRS/${pheno}/UKBB.w_hm3.${pop}.${pheno}.${weight}.fiprofile GeRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${pheno}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
      fi
    done
  
      echo ${UKBB_output}/DiverseAncestry/1KG_ref/pt_clump/${pop}/${pheno}/UKBB.${pop}.w_hm3.${pheno}.profiles PRS &gt;&gt; /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${pheno}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
  done
done

# Derive and evaluate models
for pheno in $(cat /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt);do
  for pop in $(echo AFR AMR EAS EUR_test SAS);do
sbatch --mem 10G -n 1 -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Model_builder/Model_builder_V2_nested.R \
      --pheno /users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/Residuals/${pop}/${pheno}_resid.pheno \
    --out /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${pheno}.${pop}-GeRSs.${pop}-PRSs.pt_clump \
    --n_core 1 \
    --compare_predictors F \
    --assoc T \
    --predictors /users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/${pheno}/Association_withPRS_and_GeRSs/UKBB.w_hm3.AllTissue.${pheno}.${pop}-GeRSs.${pop}-PRSs.pt_clump.predictor_groups
done
done
</code></pre>
</details>
</div>
<div id="plot-the-results-1" class="section level3">
<h3><span class="header-section-number">2.3.5</span> Plot the results</h3>
<hr />
<div id="ptclump-1" class="section level4">
<h4><span class="header-section-number">2.3.5.1</span> pT+clump</h4>
<details>
<p><summary>pT + clump comparison</summary></p>
<pre class="r"><code>library(data.table)
pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR_test&#39;,&#39;SAS&#39;)
pheno&lt;-fread(&#39;/users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt&#39;, header=F)$V1

res&lt;-list()
for(i in 1:length(pheno)){
  res_pheno&lt;-NULL
  for(k in 1:length(pop)){
    tmp&lt;-fread(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,pheno[i],&#39;.EUR-PRSs.pred_eval.txt&#39;))
    tmp2&lt;-data.frame( Population=gsub(&#39;_test&#39;,&#39;&#39;,pop[k]),
                      Phenotype=pheno[i],
                      tmp)
    
    res_pheno&lt;-rbind(res_pheno, tmp2)
  }

  res_pheno$Model&lt;-gsub(&#39;e.&#39;,&#39;e-&#39;,gsub(&#39;_.*&#39;,&#39;&#39;,gsub(&#39;SCORE.&#39;,&#39;&#39;,res_pheno$Model)))
  res_pheno$Model&lt;-factor(res_pheno$Model, levels=unique(res_pheno$Model))
  res[[pheno[i]]]&lt;-res_pheno
}

library(ggplot2)
library(cowplot)

plot_list&lt;-NULL
for(i in 1:length(pheno)){
  plot_list[[pheno[i]]]&lt;-ggplot(res[[pheno[i]]], aes(x=factor(Model), y=R, colour=Population)) +
                            geom_point(stat=&quot;identity&quot;, position=position_dodge( width=.25)) +
                            geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(.25)) +
                            theme_half_open() +
                            theme(axis.text.x = element_text(angle = 55, vjust = 1, hjust=1), plot.title = element_text(hjust = 0.4, size=12)) +
                            background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                            labs(x=&#39;pT&#39;,y=&quot;Correlation (SE)&quot;, title=pheno[i])
}

png(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/UKB.Diverse.internal.pTclump_per_pT.png&#39;, units=&#39;px&#39;, res=300, width=3000, height=10000)
  plot_grid(plotlist=plot_list, ncol = 2)
dev.off()</code></pre>
</details>
<hr />
</div>
<div id="prs-method-comparison-1" class="section level4">
<h4><span class="header-section-number">2.3.5.2</span> PRS method comparison</h4>
<pre class="r"><code>source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config&#39;)

meta_res_eval&lt;-NULL
meta_res_comp&lt;-NULL
meta_res_eval_plots&lt;-list()
meta_res_eval_diff_plots&lt;-list()
meta_res_eval_diff_matrix&lt;-list()

pheno&lt;-fread(&#39;/users/k1806347/brc_scratch/Misc/Toshiki/Phenotypes/liang_pheno_names.txt&#39;, header=F)$V1
pop&lt;-c(&#39;AFR&#39;,&#39;AMR&#39;,&#39;EAS&#39;,&#39;EUR_test&#39;,&#39;SAS&#39;)

##################
# Organise, make tables and figures for per phenotype results
##################

res_eval&lt;-NULL
res_comp&lt;-NULL
res_eval_plots&lt;-list()
res_eval_diff_plots&lt;-list()
res_eval_diff_matrix&lt;-list()
j&lt;-1

for(k in 1:length(pop)){
  for(i in 1:length(pheno)){
    res_eval_i&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,pheno[i],&#39;.EUR-PRSs.AllMethodComp.pred_eval.txt&#39;), header=T, stringsAsFactors=F)

    res_comp_i&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/&#39;,pheno[i],&#39;/Association_withPRSs/UKBB.&#39;,pop[k],&#39;.w_hm3.&#39;,pheno[i],&#39;.EUR-PRSs.AllMethodComp.pred_comp.txt&#39;), header=T, stringsAsFactors=F)
  
    # Idenitfy model names for 10FCVal, multi-PRS and pseudoval
    selected_models&lt;-NULL
    
    ##
    # pT+clump (sparse)
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_pT&lt;-res_eval_i[grepl(&#39;pt.clump.PredFile&#39;, res_eval_i$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;pT+clump.10FCVal&#39;, Label=
    res_eval_i_per_pT[res_eval_i_per_pT$R == max(res_eval_i_per_pT$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;pT+clump.MultiPRS&#39;, Label=&#39;pt.clump&#39;))
  
    ##
    # lassosum
    ##
    
    # Identify best pT in 10FCVal
    res_eval_i_per_lassosum&lt;-res_eval_i[grepl(&#39;lassosum.PredFile&#39;, res_eval_i$Model),]
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.10FCVal&#39;, Label=
    res_eval_i_per_lassosum[res_eval_i_per_lassosum$R == max(res_eval_i_per_lassosum$R),]$Model[1]))
  
    # Identify multi pT
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.MultiPRS&#39;, Label=&#39;lassosum&#39;))
  
    # Identify pseudoVal
    lassosum_val&lt;-read.table(paste0(&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_polygenic/lassosum/&#39;,pheno[i],&#39;/1KGPhase3.w_hm3.&#39;,pheno[i],&#39;.log&#39;), sep=&#39;&amp;&#39;)
    lassosum_val_s&lt;-as.numeric(gsub(&#39;s = &#39;,&#39;&#39;,lassosum_val$V1[grepl(&#39;s = &#39;,lassosum_val$V1)]))
    lassosum_val_lambda&lt;-as.numeric(gsub(&#39;lambda =&#39;,&#39;&#39;,lassosum_val$V1[grepl(&#39;lambda =&#39;,lassosum_val$V1)]))
    lassosum_val_param&lt;-paste0(&#39;s&#39;,lassosum_val_s,&#39;.lambda&#39;,lassosum_val_lambda)
    lassosum_val_param&lt;-substr(lassosum_val_param, 1, nchar(lassosum_val_param)-1) 
  
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;lassosum.PseudoVal&#39;, Label=res_eval_i[grepl(lassosum_val_param, res_eval_i$Model),]$Model[1]))
    
    ##
    # SBayesR
    ##
    
    # Identify PseudoVal
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;SBayesR.PseudoVal&#39;, Label=&#39;SBayesR&#39;))

    ##
    # All methods
    ##
    
    selected_models&lt;-rbind(selected_models,data.frame(Test=&#39;All.MultiPRS&#39;, Label=&#39;All&#39;))
  
    ###
    # Subset selected models and format
    ###
    
    # Eval
    selected_models$Label_2&lt;-paste0(gsub(&#39;_group&#39;,&#39;&#39;,selected_models$Label),&#39;_group&#39;)
    res_eval_i_select&lt;-res_eval_i[(res_eval_i$Model %in% selected_models$Label_2),]
    res_eval_i_select&lt;-merge(res_eval_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_2&#39;)], by.x=&#39;Model&#39;,by.y=&#39;Label_2&#39;)
    res_eval_i_select$Phenotype&lt;-pheno[i]
  
    res_eval_i_select&lt;-res_eval_i_select[c(&#39;Phenotype&#39;,&#39;Test&#39;,&quot;R&quot;,&quot;SE&quot;,&#39;P&#39;,&#39;R2o&#39;,&#39;N&#39;)]
                                           
    res_eval&lt;-rbind(res_eval,res_eval_i_select)
    
    # Comp
    selected_models$Label_3&lt;-gsub(&#39;_group&#39;,&#39;&#39;,selected_models$Label)
    res_comp_i_select&lt;-res_comp_i[(res_comp_i$Model_1 %in% selected_models$Label_3) &amp; (res_comp_i$Model_2 %in% selected_models$Label_3),]
    res_comp_i_select&lt;-merge(res_comp_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_3&#39;)], by.x=&#39;Model_1&#39;,by.y=&#39;Label_3&#39;)
    names(res_comp_i_select)[names(res_comp_i_select) == &#39;Test&#39;]&lt;-&#39;Model_1_Test&#39;
    res_comp_i_select&lt;-merge(res_comp_i_select, selected_models[c(&#39;Test&#39;,&#39;Label_3&#39;)], by.x=&#39;Model_2&#39;,by.y=&#39;Label_3&#39;)
    names(res_comp_i_select)[names(res_comp_i_select) == &#39;Test&#39;]&lt;-&#39;Model_2_Test&#39;
    res_comp_i_select$Label&lt;-NULL
    res_comp_i_select$Phenotype&lt;-pheno[i]
    res_comp_i_select&lt;-res_comp_i_select[c(&#39;Phenotype&#39;,&#39;Model_1_Test&#39;,&#39;Model_2_Test&#39;,&#39;Model_1_R&#39;,&#39;Model_2_R&#39;,&#39;R_diff&#39;,&#39;R_diff_pval&#39;)]
  
    res_comp&lt;-rbind(res_comp,res_comp_i_select)
    
    ###
    # Create plot showing performance of each method
    ###
    
    library(ggplot2)
    library(cowplot)
    
    res_eval_i_select$Method&lt;-gsub(&#39;\\..*&#39;,&#39;&#39;,res_eval_i_select$Test)
    res_eval_i_select$Model&lt;-gsub(&#39;.*\\.&#39;,&#39;&#39;,res_eval_i_select$Test)
    
    res_eval_i_select$Method&lt;-factor(res_eval_i_select$Method, levels=c(&#39;pT+clump&#39;,&#39;lassosum&#39;,&#39;SbayesR&#39;,&#39;All&#39;))
    res_eval_i_select$Model&lt;-factor(res_eval_i_select$Model, levels=c(&#39;10FCVal&#39;,&#39;MultiPRS&#39;,&#39;PseudoVal&#39;,&#39;Inf&#39;))
    
    res_eval_plots[[j]] &lt;- ggplot( res_eval_i_select, aes(x=Method, y=R, fill=Model)) +
                          geom_bar(stat=&quot;identity&quot;, position=position_dodge(preserve = &quot;single&quot;), width = 0.7) +
                          geom_errorbar(aes(ymin=R-SE, ymax=R+SE), width=.2, position=position_dodge(width = 0.7, preserve = &quot;single&quot;)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;)) +
                          coord_cartesian(ylim=c(min(res_eval_i_select$R[res_eval_i_select$Phenotype==pheno[i]])-0.02, max(res_eval_i_select$R[res_eval_i_select$Phenotype==pheno[i]])+0.025), clip=&quot;on&quot;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                          theme(legend.position=&quot;top&quot;, legend.title = element_blank(), legend.box=&quot;vertical&quot;, legend.margin=margin()) +
                          guides(fill=guide_legend(nrow=2))
                                  
    ###
    # Create plot showing performance of each method compared to pT+clump
    ###
  
    res_eval_i_select$R_diff&lt;-res_eval_i_select$R-res_eval_i_select$R[res_eval_i_select$Test == &quot;pT+clump.10FCVal&quot;]
    
    res_eval_diff_plots[[j]] &lt;- ggplot( res_eval_i_select, aes(x=Method, y=R_diff, fill=Model)) +
                          geom_bar(stat=&quot;identity&quot;, position=position_dodge(preserve = &quot;single&quot;), width = 0.7) +
                          geom_errorbar(aes(ymin=R_diff-SE, ymax=R_diff+SE), width=.2, position=position_dodge(width = 0.7, preserve = &quot;single&quot;)) +
                          labs(y=&quot;Correlation (SE)&quot;, x=&#39;&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;)) +
                          coord_cartesian(ylim=c(min(res_eval_i_select$R_diff[res_eval_i_select$Phenotype==pheno[i]])-0.02, max(res_eval_i_select$R_diff[res_eval_i_select$Phenotype==pheno[i]])+0.025), clip=&quot;on&quot;) +
                          theme_half_open() +
                                  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
                          background_grid(major = &#39;y&#39;, minor = &#39;y&#39;) +
                          theme(legend.position=&quot;top&quot;, legend.title = element_blank(), legend.box=&quot;vertical&quot;, legend.margin=margin()) +
                          guides(fill=guide_legend(nrow=2))
    
  
    ###
    # Create plot showing matrix of differences with significance
    ###
  
    library(reshape2)
    
    res_comp_i_select$R_diff[res_comp_i_select$Model_1_Test == res_comp_i_select$Model_2_Test]&lt;-NA
  
    res_comp_i_select$R_diff_catagory&lt;-&#39;NA&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.002]&lt;-&#39;-0.025 - -0.002&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.025]&lt;-&#39;-0.08 - -0.025&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &lt; -0.08]&lt;-&#39;&lt; -0.08&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; -0.002 &amp; res_comp_i_select$R_diff &lt; 0.002]&lt;-&#39;-0.002 - 0.002&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.002]&lt;-&#39;0.002 - 0.025&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.025]&lt;-&#39;0.025 - 0.08&#39;
    res_comp_i_select$R_diff_catagory[res_comp_i_select$R_diff &gt; 0.08]&lt;-&#39;&gt; 0.08&#39;
    
    res_comp_i_select$R_diff_catagory&lt;-factor(res_comp_i_select$R_diff_catagory, level=rev(c(&#39;&lt; -0.08&#39;,&#39;-0.08 - -0.025&#39;,&#39;-0.025 - -0.002&#39;,&#39;-0.002 - 0.002&#39;,&#39;0.002 - 0.025&#39;,&#39;0.025 - 0.08&#39;,&#39;&gt; 0.08&#39;)))
  
    res_comp_i_select$star&lt;-&#39; &#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 0.05]&lt;-&#39;*&#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 1e-3]&lt;-&#39;**&#39;
    res_comp_i_select$star[res_comp_i_select$R_diff_pval &lt; 1e-6]&lt;-&#39;***&#39;
  
    
    library(RColorBrewer)
    res_eval_diff_matrix[[j]]&lt;-ggplot(data = res_comp_i_select, aes(Model_2_Test, Model_1_Test, fill=R_diff_catagory))+
     geom_tile(color = &quot;white&quot;)+
     labs(y=&#39;Test&#39;, x=&#39;Comparison&#39;, title=paste0(pheno[i],&quot;(&quot;,pop[k],&quot;)&quot;), fill=&#39;R difference&#39;) +
      theme_minimal()+ 
     theme(axis.text.x = element_text(angle = 45, vjust = 1, 
        size = 8, hjust = 1))+
     coord_fixed() +
    geom_text(data=res_comp_i_select, aes(Model_2_Test, Model_1_Test, label = star), color = &quot;black&quot;, size = 4, angle = 0, vjust=0.8) +
    scale_fill_brewer(breaks = levels(res_comp_i_select$R_diff_catagory), palette = &quot;RdBu&quot;, drop=F, na.value=&#39;grey&#39;)

    j&lt;-j+1
  }
}

write.csv(res_eval, paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp.csv&#39;), row.names=F, quote=F)

write.csv(res_comp, paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_diff.csv&#39;), row.names=F, quote=F)

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1000*ceiling(length(res_eval_plots)/2))
print(plot_grid(plotlist=res_eval_plots, ncol = 2))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_pTDiff.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1000*ceiling(length(res_eval_plots)/2))
print(plot_grid(plotlist=res_eval_diff_plots, ncol = 2))
dev.off()

png(paste0(&#39;/users/k1806347/brc_scratch/Analyses/DiverseAncestry/UKBB_outcomes_for_prediction/PRS_method_comp_diff.png&#39;), units=&#39;px&#39;, res=300, width=2450, height=1750*length(res_eval_diff_matrix))
print(plot_grid(plotlist=res_eval_diff_matrix, ncol = 1))
dev.off()

# With only two phenotypes, meta-analysis is not necessary or helpful.</code></pre>
</details>
<hr />
</div>
</div>
</div>
</div>
<div id="future-directions" class="section level1">
<h1><span class="header-section-number">3</span> Future directions</h1>
<p>To do:</p>
<ul>
<li>Run with more phenotypes</li>
<li>Run for other PRS methods</li>
<li>Compare methods that improve portability of EUR GWAS</li>
</ul>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
