---
title: "Genotype-based scoring in target samples"
output: 
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

This page describes calculating reference-standardised genotype-based scores in a target sample. This process requires a number of reference files, which can be made using instructions found [here](https://opain.github.io/GenoPred/Pipeline_prep.html). Here, we will calculate ancestry scores, polygenic scores, and functionally informed polygenic scores. Each of these scores can be used for a number of tasks, such as inferrence of ancestry, estimation of risk, test for genetic covariance between two outcomes, and stratification. 

Before these scores can be calculated, the target samples genetic data must be harmonised with the reference. This process varies depending on the source of the data. Here, I show how the UK biobank and Twins Early Development Study (TEDS) genetic data were harmonised with the reference. Note, both of these samples had been imputed prior to receiving the data.

# Harmonisation with the reference

## UK-Biobank
Here I take each chromosome of the UK-Biobank imputed data (.bgen), extract SNPs in the HapMap3 SNP-list, convert the data to hard-call PLINK binaries (.bed/.bim/.fam), check whether any SNPs need to be strand flipped, and insert HapMap3 SNPs that are not present into the UK-Biobank data as missing. 

I used HapMap3 SNPs only as these are typically available after imputation, enabling comparison of scores from diverse sources. I convert SNPs to hard call with a hard-call threshold of 0, meaning it will take the most likely genotype regardless of the certainty, which speeds subsequent stages of the genotype-based scoring and is unlikely to introduce many errors as these SNPs are typically reliable. I insert HapMap3 SNPs as missing to enable allele frequency imputation while scoring, which ensures scores are comparable between samples with different SNPs available.

The code below applies a script called 'Harmonisation_of_UKBB.R' to each chromosome of the UK-Biobank data. See [here](https://github.com/opain/GenoPred/tree/master/Scripts/Harmonisation_of_UKBB) for more information on this script.

<details><summary>See code</summary>
```{bash, echo=T, eval=F}
# Set required variables
UKBB_original=$(sed -n '/UKBB_original/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)
UKBB_fam=$(sed -n '/UKBB_fam/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)
UKBB_harmonised=$(sed -n '/UKBB_harmonised/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)
Geno_1KG_dir=$(sed -n '/Geno_1KG_dir/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)
qctool2=$(sed -n '/qctool2/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)
plink1_9=$(sed -n '/plink1_9/p' /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Variables_used/Target_scoring.var | cut -d ' ' -f 2)

# Process each chromosome of the UK-Biobank data using the Harmonisation_of_UKBB.R script
for chr in $(seq 1 22); do
qsub -l h_vmem=10G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Harmonisation_of_UKBB/Harmonisation_of_UKBB.R \
  --chr $chr \
  --input_dir ${UKBB_original} \
  --target_fam ${UKBB_fam} \
  --output_dir ${UKBB_harmonised} \
  --reference_dir ${Geno_1KG_dir} \
  --qctool2 ${qctool2} \
  --plink ${plink1_9}
done
```
</details>

## TEDS
The code below applies a script called 'Harmonisation_of_TEDS.R' to each chromosome of the UK-Biobank data. See [here](https://github.com/opain/GenoPred/tree/master/Scripts/Harmonisation_of_TEDS) for more information on this script.

<details><summary>See code</summary>
```{bash, echo=T, eval=F}
# Process each chromosome of the UK-Biobank data using the Harmonisation_of_UKBB.R script
for chr in $(seq 1 22); do
  qsub -l h_vmem=10G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Harmonisation_of_UKBB/Harmonisation_of_TEDS.R --chr $chr
done
```
</details>








