<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Genotype-based scoring in target samples</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Genotype-based scoring in target samples</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<p>This page describes calculating reference-standardised genotype-based scores in a target sample. This process requires a number of reference files, which can be made using instructions found <a href="https://opain.github.io/GenoPred/Pipeline_prep.html">here</a>. Here, we will calculate ancestry scores, polygenic scores, and functionally informed polygenic scores. Each of these scores can be used for a number of tasks, such as inferrence of ancestry, estimation of risk, test for genetic covariance between two outcomes, and stratification.</p>
<p><br/></p>
<hr />
<div id="harmonisation-with-the-reference" class="section level1">
<h1><span class="header-section-number">1</span> Harmonisation with the reference</h1>
<p>Before these scores can be calculated, the target samples genetic data must be harmonised with the reference. This process varies depending on the source of the data. Here, I show how the UK biobank and Twins Early Development Study (TEDS) genetic data were harmonised with the reference. Note, both of these samples had been imputed prior to receiving the data.</p>
<p>In all scenarios, I extract SNPs in the HapMap3 SNP-list, convert the data to hard-call PLINK binaries (.bed/.bim/.fam), check whether any SNPs need to be strand flipped, and insert HapMap3 SNPs that are not present into the target data as missing.</p>
<p>I use HapMap3 SNPs only as these are typically available after imputation, enabling comparison of scores from diverse sources. I convert SNPs to hard call with a hard-call threshold of 0, meaning it will take the most likely genotype regardless of the certainty, which speeds subsequent stages of the genotype-based scoring and is unlikely to introduce many errors as these SNPs are typically reliable. I insert HapMap3 SNPs as missing to enable allele frequency imputation while scoring, which ensures scores are comparable between samples with different SNPs available.</p>
<div id="uk-biobank" class="section level2">
<h2><span class="header-section-number">1.1</span> UK Biobank</h2>
<p>The code below applies a script called ‘Harmonisation_of_UKBB.R’ to each chromosome of the UK-Biobank data. See <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Harmonisation_of_UKBB">here</a> for more information on this script.</p>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Process each chromosome of the UK-Biobank data using the Harmonisation_of_UKBB.R script
for chr in $(seq 1 22); do
qsub -l h_vmem=10G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Harmonisation_of_UKBB/Harmonisation_of_UKBB.R \
  --chr $chr \
  --input_dir ${UKBB_original} \
  --target_fam ${UKBB_fam} \
  --output_dir ${UKBB_output}/Genotype/Harmonised \
  --reference_dir ${Geno_1KG_dir} \
  --qctool2 ${qctool2} \
  --plink ${plink1_9}
done</code></pre>
</details>
<p><br/></p>
</div>
<div id="teds" class="section level2">
<h2><span class="header-section-number">1.2</span> TEDS</h2>
<p>The TEDS data has already undergone pre-imputation QC, HRC imputation, and post-imputation QC. Genotyping was performed using two chips, Affy and OEE. Each Chip was QC’d and imputed seperatly, and then merged retaining the non-overlapping SNPs to maximise the SNP data available for analysis. The data is in plink format. Here we will harmonise the data with the HapMap3 reference.</p>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code>
qsub -l h_vmem=20G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Harmonisation_of_TEDS/Harmonisation_of_TEDS.R
</code></pre>
</details>
<p><br/></p>
</div>
<div id="andme" class="section level2">
<h2><span class="header-section-number">1.3</span> 23andMe</h2>
<div id="format-and-impute" class="section level3">
<h3><span class="header-section-number">1.3.1</span> Format and impute</h3>
<p>Here we will prepare genotype data for Oliver_Pain (Me), using the script 23andMe_imputer.R. This code is very similar to that used by the Impute.Me website created by Lasse Folkersen.</p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for individual in $(echo Rain_Rimfeld); do
geno_file=$(ls ${geno_23andMe}/${individual})

# Run for one user
qsub -l h_vmem=15G -pe smp 6 /users/k1806347/brc_scratch/Software/Rscript.sh  /mnt/lustre/users/k1806347/Software/MyGit/GenoPred/Scripts/23andMe_imputer/23andMe_imputer.R \
--FID ${individual} \
--IID ${individual} \
--geno ${geno_23andMe}/${individual}/${geno_file} \
--plink ${plink1_9} \
--out ${geno_23andMe}/${individual}/${individual} \
--ref ${Impute2_1KG_dir} \
--shapeit ${ShapeIt} \
--impute2 ${Impute2} \
--n_core 6 \
--qctool ${qctool2}

done

#######
# Extract HapMap3 SNPs and insert missing genotypes for subsequent MAF imputation.
#######
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for individual in $(echo Rain_Rimfeld); do
qsub -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh  /mnt/lustre/users/k1806347/Software/MyGit/GenoPred/Scripts/hm3_harmoniser/hm3_harmoniser.R \
--target ${geno_23andMe}/${individual}/${individual}.1KGphase3.chr \
--ref ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
--plink ${plink1_9} \
--out ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr
done
</code></pre>
<hr />
</div>
</div>
</div>
<div id="ancestry-scoring" class="section level1">
<h1><span class="header-section-number">2</span> Ancestry scoring</h1>
<p>In genetic analyses it is important to consider an individual’s ancestry as the frequency of alleles vary between populations making comparison across ancestries often invalid. Differences in ancestry can lead to population stratification in genome-wide assocaition studies because the frequency of an outcome may vary between individuals of different ancestries, which will then lead to an association between the outcome and all genetic variants that differ in frequency between ancestral groups. Furthermore, differences in LD between ancestries lead to difference patterns of association which influences the predictive utility of polygenic scores across diverse populations.</p>
<p>In this genotype-based scoring pipeline, we estimate factor scores for each individual on the first 100 European specific and 100 all-ancestry principal components (PCs) of populations structure by projecting previously identified PCs in the 1000 Genomes reference (see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#3_ancestry_scoring">here</a>). Using the 100 all-ancestry PCs, we estimate which super population individuals belong to. This allows subsquent genotype-based scores to be scaled based on the mean and SD of scores within their ancestral group, improving their interpretability. We also use their inferred ancestry to stratify individuals when building predictive models, to provide population specific models of risk. Furthermore, given that the frequency of outcomes do often vary between ancestral groups, we also derive these estimates of ancestry to be used as predictors of risk. This is discussed more elsewhere (see here).</p>
<p>This section uses a script called ‘scaled_ancestry_scorer.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/scaled_ancestry_scorer">here</a> for more information on this script.</p>
<div id="uk-biobank-1" class="section level2">
<h2><span class="header-section-number">2.1</span> UK Biobank</h2>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

#################
# Calculate all ancestry PCs of population structure and predict ancestry
#################

qsub -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/AllAncestry/1KGPhase3.w_hm3.AllAncestry.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec \
    --plink2 ${plink2} \
    --output ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small \
    --pop_model ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.pop_enet_model.rds \
    --pop_model_scale ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scale \
    --pop_scale_for_keep ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scalefiles.txt

#################
# Calculate European ancestry PCs of population structure
#################

# Use the EUR.keep file from the AllAncestry run above to select people of EUR ancestry.
qsub -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
    --target_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec \
    --ref_scale ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.scale \
    --plink2 ${plink2} \
    --output ${UKBB_output}/Projected_PCs/EUR/UKBB.w_hm3 \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small
</code></pre>
</details>
<p><br/></p>
</div>
<div id="teds-1" class="section level2">
<h2><span class="header-section-number">2.2</span> TEDS</h2>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

#################
# Calculate all ancestry PCs of population structure and predict ancestry
#################

qsub -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${TEDS_output_dir}/Genotype/Harmonised/TEDS.w_hm3.QCd.AllSNP.chr \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/AllAncestry/1KGPhase3.w_hm3.AllAncestry.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec \
    --plink2 ${plink2} \
    --output ${TEDS_output_dir}/Projected_PCs/AllAncestry/TEDS.w_hm3.AllAncestry \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small \
    --pop_model ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.pop_enet_model.rds \
    --pop_model_scale ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scale \
    --pop_scale_for_keep ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scalefiles.txt

#################
# Calculate European ancestry PCs of population structure
#################

# Use the EUR.keep file from the AllAncestry run above to select people of EUR ancestry.
qsub -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${TEDS_output_dir}/Genotype/Harmonised/TEDS.w_hm3.QCd.AllSNP.chr \
    --target_keep ${TEDS_output_dir}/Projected_PCs/AllAncestry/TEDS.w_hm3.AllAncestry.EUR.keep \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec \
    --ref_scale ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.scale \
    --plink2 ${plink2} \
    --output ${TEDS_output_dir}/Projected_PCs/EUR/TEDS.w_hm3 \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small
</code></pre>
</details>
<p><br/></p>
</div>
<div id="andme-1" class="section level2">
<h2><span class="header-section-number">2.3</span> 23andMe</h2>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config


#################
# Calculate all ancestry PCs of population structure and predict ancestry
#################

for individual in $(echo Rain_Rimfeld); do
qsub /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/AllAncestry/1KGPhase3.w_hm3.AllAncestry.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.eigenvec \
    --plink2 ${plink2} \
    --output ${geno_23andMe}/${individual}/Projected_PCs/AllAncestry/${individual}.w_hm3.AllAncestry \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small \
    --pop_model ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.pop_enet_model.rds \
    --pop_model_scale ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scale \
    --pop_scale_for_keep ${Geno_1KG_dir}/Score_files_for_ancestry/AllAncestry/1KGPhase3.w_hm3.AllAncestry.scalefiles.txt
done

#################
# Calculate European ancestry PCs of population structure
#################

# Use the EUR.keep file from the AllAncestry run above to select people of EUR ancestry.
for individual in $(echo Rain_Rimfeld); do
qsub /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_ancestry_scorer/scaled_ancestry_scorer.R \
    --target_plink_chr ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr \
    --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
    --ref_score ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec.var \
    --ref_eigenvec ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.eigenvec \
    --ref_scale ${Geno_1KG_dir}/Score_files_for_ancestry/EUR/1KGPhase3.w_hm3.EUR.scale \
    --plink2 ${plink2} \
    --output ${geno_23andMe}/${individual}/Projected_PCs/EUR/${individual}.w_hm3 \
    --pop_data ${Geno_1KG_dir}/integrated_call_samples_v3.20130502.ALL.panel_small
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="polygenic-scoring" class="section level1">
<h1><span class="header-section-number">3</span> Polygenic scoring</h1>
<p>Here we calculate poylgenic scores for a range of outcomes in the target samples based on previously prepared reference scoring files, and scale the polygenic scores based on scores in a reference sample with matched ancestry. For more information on preparing reference files for polygenic scoring, see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#4_polygenic_scoring">here</a>.</p>
<p>One aim of the GenoPred project is to compare methods for calculating genotype-based scores. Here we compare three approaches for polygenic scoring:</p>
<ul>
<li>p-value thresholding and LD-based clumping (pT + clump)</li>
<li>lassosum</li>
<li>PRScs</li>
<li>SBLUP</li>
</ul>
<div id="pt-clump" class="section level2">
<h2><span class="header-section-number">3.1</span> pT + clump</h2>
<p>This section uses a script called ‘Scaled_polygenic_scorer.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Scaled_polygenic_scorer">here</a> for more information on this script.</p>
<div id="uk-biobank-2" class="section level3">
<h3><span class="header-section-number">3.1.1</span> UK Biobank</h3>
<div id="sparse-p-value-thresholding" class="section level4">
<h4><span class="header-section-number">3.1.1.1</span> Sparse p-value thresholding</h4>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create a list of gwas that haven&#39;t got corresponding PRS files in UKBB
&gt; ${UKBB_output}/PolygenicScores/EUR/todo.txt
for gwas in $(cut -f 1 -d &#39; &#39; ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt);do
if [ ! -f ${UKBB_output}/PolygenicScores/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.profiles ]; then
echo $gwas &gt;&gt; ${UKBB_output}/PolygenicScores/EUR/todo.txt
fi
done

# Calculate polygenic scores for for Europeans only for now.
n=0
for gwas in $(cat ${UKBB_output}/PolygenicScores/EUR/todo.txt);do
qsub -N $(echo PRS_job$(($n+10))) -hold_jid $(echo PRS_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R \
  --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --target_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic/${gwas}/1KGPhase3.w_hm3.${gwas} \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${UKBB_output}/PolygenicScores/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR
n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
</div>
<div id="dense-p-value-thresholding" class="section level4">
<h4><span class="header-section-number">3.1.1.2</span> Dense p-value thresholding</h4>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

###
# Split the EUR UKBB individuals into 10K batches
###
# Temporary files would require a large amount of storage for all UKBB individuals
# Splitting the data into batches avoids this issue
mkdir /users/k1806347/brc_scratch/Data/UKBB/PolygenicScores_dense
split -l 10000 -d ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep ${UKBB_output}/PolygenicScores_dense/UKBB.w_hm3.AllAncestry.EUR.keep.Batch_

# Calculate polygenic scores for for Europeans only for now.
n=0
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
for batch in $(ls ${UKBB_output}/PolygenicScores_dense/UKBB.w_hm3.AllAncestry.EUR.keep.Batch_* | cut -d. -f6-);do
qsub -N $(echo PRS_job$(($n+50))) -hold_jid $(echo PRS_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R \
  --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --target_keep ${UKBB_output}/PolygenicScores_dense/UKBB.w_hm3.AllAncestry.EUR.keep.${batch} \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic_dense/${gwas}/1KGPhase3.w_hm3.${gwas} \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_dense/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/${batch}/UKBB.w_hm3.${gwas}.EUR.${batch}
n=$(($n+1))
done
done

# Check all the batches finished correctly There should be 46 files
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
ls ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/*/UKBB.w_hm3.${gwas}.EUR.*.profiles | wc -l
done

# Combine the batches into a single file
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
head -1 ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/Batch_00/UKBB.w_hm3.${gwas}.EUR.Batch_00.profiles &gt; ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.profiles
tail -n +2 -q ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/*/UKBB.w_hm3.${gwas}.EUR.*.profiles &gt;&gt; ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.profiles
done

# Delete per batch files and move log files into a single folder
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
mkdir ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/log_files
mv ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/*/UKBB.w_hm3.${gwas}.EUR.*.log ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/log_files/
rm -r ${UKBB_output}/PolygenicScores_dense/EUR/${gwas}/Batch_*
done
</code></pre>
</details>
<p><br/></p>
</div>
</div>
<div id="andme-2" class="section level3">
<h3><span class="header-section-number">3.1.2</span> 23andMe</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for individual in $(echo Rain_Rimfeld); do

mkdir -p ${geno_23andMe}/${individual}/PolygenicScores/EUR

# Create a list of gwas that haven&#39;t got corresponding PRS files
&gt; ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo.txt
for gwas in $(ls ${Geno_1KG_dir}/Score_files_for_poylygenic/ | sed &#39;/todo.txt/d&#39;);do
if [ ! -f ${geno_23andMe}/${individual}/PolygenicScores/EUR/${gwas}/${individual}.w_hm3.${gwas}.EUR.profiles ]; then
echo $gwas &gt;&gt; ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo.txt
fi
done

# Calculate polygenic scores for for Europeans only for now.
n=0

for gwas in $(cat ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo.txt);do
qsub -N $(echo PRS_job$(($n+10))) -hold_jid $(echo PRS_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer/Scaled_polygenic_scorer.R \
  --target_plink_chr ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic/${gwas}/1KGPhase3.w_hm3.${gwas} \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${geno_23andMe}/${individual}/PolygenicScores/EUR/${gwas}/${individual}.w_hm3.${gwas}.EUR
n=$(($n+1))
done

done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="lassosum" class="section level2">
<h2><span class="header-section-number">3.2</span> lassosum</h2>
<p>This section uses a script called ‘Scaled_polygenic_scorer_lassosum.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Scaled_polygenic_scorer_lassosum">here</a> for more information on this script.</p>
<div id="uk-biobank-3" class="section level3">
<h3><span class="header-section-number">3.2.1</span> UK Biobank</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Here we only calculate scores relevent to the phenotyps that we will use for comparing methods
n=0
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
qsub -N $(echo PRS_job$(($n+10))) -hold_jid $(echo PRS_job$(($n+0))) -pe smp 6 -l h_vmem=10G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_lassosum/Scaled_polygenic_scorer_lassosum.R \
  --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --target_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
  --ref_model ${Geno_1KG_dir}/Score_files_for_poylygenic_lassosum/${gwas}/1KGPhase3.w_hm3.${gwas}.unvalidated.model.RDS \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_lassosum/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --pheno_name ${gwas} \
  --n_cores 6 \
  --output ${UKBB_output}/PolygenicScores_lassosum/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR
n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="andme-3" class="section level3">
<h3><span class="header-section-number">3.2.2</span> 23andMe</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for individual in $(echo Kaili_Rimfeld); do

# Here we only calculate scores relevent to the phenotyps that we will use for comparing methods
n=0
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03);do
qsub -N $(echo PRS_job$(($n+10))) -hold_jid $(echo PRS_job$(($n+0))) -pe smp 1 -l h_vmem=5G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_lassosum/Scaled_polygenic_scorer_lassosum.R \
  --target_plink_chr ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr \
  --target_keep ${geno_23andMe}/${individual}/Projected_PCs/AllAncestry/${individual}.w_hm3.AllAncestry.model_pred.EUR.keep \
  --ref_model ${Geno_1KG_dir}/Score_files_for_poylygenic_lassosum/${gwas}/1KGPhase3.w_hm3.${gwas}.unvalidated.model.RDS \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_lassosum/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --pheno_name ${gwas} \
  --n_cores 1 \
  --output ${geno_23andMe}/${individual}/PolygenicScores_lassosum/EUR/${gwas}/${individual}.w_hm3.${gwas}.EUR
n=$(($n+1))
done

done

module add general/R/3.5.0
module add compilers/gcc/8.1.0
module add general/python/2.7.10
R

opt$target_plink_chr&lt;-&#39;/users/k1806347/brc_scratch/Data/23andMe/Kaili_Rimfeld/Kaili_Rimfeld.1KGphase3.hm3.chr&#39;
opt$target_keep&lt;-&#39;/users/k1806347/brc_scratch/Data/23andMe/Kaili_Rimfeld/Projected_PCs/AllAncestry/Kaili_Rimfeld.w_hm3.AllAncestry.model_pred.EUR.keep&#39;
opt$ref_model&lt;-&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_poylygenic_lassosum/DEPR06/1KGPhase3.w_hm3.DEPR06.unvalidated.model.RDS&#39;
opt$ref_scale&lt;-&#39;/users/k1806347/brc_scratch/Data/1KG/Phase3/Score_files_for_poylygenic_lassosum/DEPR06/1KGPhase3.w_hm3.DEPR06.EUR.scale&#39;
opt$pheno_name&lt;-&#39;DEPR06&#39;
opt$n_cores&lt;-1
opt$output&lt;-&#39;/users/k1806347/brc_scratch/Data/23andMe/Kaili_Rimfeld/PolygenicScores_lassosum/EUR/DEPR06/Kaili_Rimfeld.w_hm3.DEPR06.EUR&#39;


# Note. lassosum scoring failed with one individual.</code></pre>
</details>
<p><br/></p>
</div>
</div>
<div id="prscs" class="section level2">
<h2><span class="header-section-number">3.3</span> PRScs</h2>
<p>This section uses a script called ‘Scaled_polygenic_scorer_PRScs.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Scaled_polygenic_scorer_PRScs">here</a> for more information on this script.</p>
<div id="uk-biobank-4" class="section level3">
<h3><span class="header-section-number">3.3.1</span> UK Biobank</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create a list of gwas that haven&#39;t got corresponding PRS files in UKBB
&gt; ${UKBB_output}/PolygenicScores/EUR/todo_PRScs.txt
#for gwas in $(cut -f 1 -d &#39; &#39; ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt);do
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
if [ ! -f ${UKBB_output}/PolygenicScores_PRScs/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.PRScs_profiles ]; then
echo $gwas &gt;&gt; ${UKBB_output}/PolygenicScores/EUR/todo_PRScs.txt
fi
done

n=0
for gwas in $(cat ${UKBB_output}/PolygenicScores/EUR/todo_PRScs.txt);do
qsub -N $(echo PRScs_job$(($n+25))) -hold_jid $(echo PRScs_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_PRScs/Scaled_polygenic_scorer_PRScs.R \
  --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --target_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic_PRScs/${gwas}/1KGPhase3.w_hm3.${gwas} \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_PRScs/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${UKBB_output}/PolygenicScores_PRScs/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR
n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
</div>
<div id="andme-4" class="section level3">
<h3><span class="header-section-number">3.3.2</span> 23andMe</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

for individual in $(echo Oliver_Pain Kaili_Rimfeld Robert_Plomin); do

mkdir -p ${geno_23andMe}/${individual}/PolygenicScores_PRScs/EUR

# Create a list of gwas that haven&#39;t got corresponding PRS files in UKBB
&gt; ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo_PRScs.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03);do
if [ ! -f ${geno_23andMe}/${individual}/PolygenicScores_PRScs/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.PRScs_profile ]; then
echo $gwas &gt;&gt; ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo_PRScs.txt
fi
done

n=0
for gwas in $(cat ${geno_23andMe}/${individual}/PolygenicScores/EUR/todo_PRScs.txt);do
qsub -N $(echo PRScs_job$(($n+25))) -hold_jid $(echo PRScs_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_PRScs/Scaled_polygenic_scorer_PRScs.R \
  --target_plink_chr ${geno_23andMe}/${individual}/${individual}.1KGphase3.hm3.chr \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic_PRScs/${gwas}/1KGPhase3.w_hm3.${gwas} \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_PRScs/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${geno_23andMe}/${individual}/PolygenicScores_PRScs/EUR/${gwas}/${individual}.w_hm3.${gwas}.EUR
n=$(($n+1))
done

done</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="sblup" class="section level2">
<h2><span class="header-section-number">3.4</span> SBLUP</h2>
<p>This section uses a script called ‘Scaled_polygenic_scorer_PRScs.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Scaled_polygenic_scorer_PRScs">here</a> for more information on this script.</p>
<div id="uk-biobank-5" class="section level3">
<h3><span class="header-section-number">3.4.1</span> UK Biobank</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

mkdir -p ${UKBB_output}/PolygenicScores_SBLUP/EUR

# Create a list of gwas that haven&#39;t got corresponding PRS files in UKBB
&gt; ${UKBB_output}/PolygenicScores_SBLUP/EUR/todo_SBLUP.txt
#for gwas in $(cut -f 1 -d &#39; &#39; ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt);do
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01);do
if [ ! -f ${UKBB_output}/PolygenicScores_SBLUP/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR.SBLUP_profiles ]; then
echo $gwas &gt;&gt; ${UKBB_output}/PolygenicScores_SBLUP/EUR/todo_SBLUP.txt
fi
done

n=0
for gwas in $(cat ${UKBB_output}/PolygenicScores_SBLUP/EUR/todo_SBLUP.txt);do
qsub -N $(echo SBLUP_job$(($n+25))) -hold_jid $(echo SBLUP_job$(($n+0))) /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Scaled_polygenic_scorer_SBLUP/Scaled_polygenic_scorer_SBLUP.R \
  --target_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --target_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
  --ref_score ${Geno_1KG_dir}/Score_files_for_poylygenic_SBLUP/${gwas}/GWAS_sumstats_SBLUP.sblup.cojo \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_poylygenic_SBLUP/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr \
  --plink ${plink1_9} \
  --pheno_name ${gwas} \
  --output ${UKBB_output}/PolygenicScores_SBLUP/EUR/${gwas}/UKBB.w_hm3.${gwas}.EUR
n=$(($n+1))
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
</div>
<div id="functionally-informed-polygenic-scoring" class="section level1">
<h1><span class="header-section-number">4</span> Functionally-informed polygenic scoring</h1>
<p>Here we calculate functionally-informed poylgenic scores for a range of outcomes in the target samples based on previously prepared reference scoring files, and scale the polygenic scores based on scores in a reference sample with matched ancestry. For more information on preparing reference files for polygenic scoring, see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#5_functionally-informed_polygenic_scoring">here</a>.</p>
<p>First we need to predict the functional genomic features in the target sample, then we calculate the functionally-informed polygenic scores by combining this information with previously prepared score files.</p>
<div id="predicting-functional-genomic-features" class="section level2">
<h2><span class="header-section-number">4.1</span> Predicting functional genomic features</h2>
<p>Here we predict functional genomic features into the target samples. Currently the pipeline only includes features representing gene expression levels. The multiple SNP-predictors of gene expression have been downloaded from the FUSION website, and converted into the correct format for predicting functional features in a target samples. For more information see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#5_functionally-informed_polygenic_scoring">here</a>.</p>
<p>This section uses a script called ‘FUSION_targ_scorer.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/FUSION_targ_scorer">here</a> for more information on this script.</p>
<div id="uk-biobank-6" class="section level3">
<h3><span class="header-section-number">4.1.1</span> UK Biobank</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create file listing weights that haven&#39;t been predicted in the reference.
&gt; ${UKBB_output}/Predicted_expression/FUSION/EUR/todo.txt
for weights in $(cat ~/brc_scratch/Data/TWAS_sumstats/FUSION/snp_weight_list.txt);do
if [ ! -f ${UKBB_output}/Predicted_expression/FUSION/EUR/${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.${weights}.predictions.gz ]; then
echo $weights &gt;&gt; ${UKBB_output}/Predicted_expression/FUSION/EUR/todo.txt
fi
done

n=0
for weights in $(cat ${UKBB_output}/Predicted_expression/FUSION/EUR/todo.txt);do
qsub -N $(echo Predjob_$(($n+1))) -hold_jid $(echo Predjob_$(($n+0))) -pe smp 5 -l h_vmem=15G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/FUSION_targ_scorer/FUSION_targ_scorer.R \
  --targ_plink_chr ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr \
  --targ_keep ${UKBB_output}/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep \
  --weights ${FUSION_dir}/SNP-weights/${weights}/${weights}.pos \
  --plink ${plink1_9} \
  --n_cores 5 \
  --memory 40000 \
  --score_files /users/k1806347/brc_scratch/Data/FUSION/SCORE_FILES \
  --ref_scale ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.EUR.scale \
  --pigz ${pigz_binary} \
  --output ${UKBB_output}/Predicted_expression/FUSION/EUR/${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.${weights} \
  --ref_freq_chr ${Geno_1KG_dir}/freq_files/EUR/1KGPhase3.w_hm3.EUR.chr
n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="calculating-functionally-informed-polygenic-scores" class="section level2">
<h2><span class="header-section-number">4.2</span> Calculating functionally-informed polygenic scores</h2>
<p>Here we use the predicted functional genomic features and previously prepared reference score files to calculate functionally informed polygenic scores in the target samples. For more information see <a href="https://opain.github.io/GenoPred/Pipeline_prep.html#5_functionally-informed_polygenic_scoring">here</a>.</p>
<p>This section uses a script called ‘Scaled_polygenic_scorer.R’. See more information on the utility of this script <a href="https://github.com/opain/GenoPred/tree/master/Scripts/scaled_functionally_informed_risk_scorer">here</a> for more information on this script.</p>
<div id="uk-biobank-7" class="section level3">
<h3><span class="header-section-number">4.2.1</span> UK Biobank</h3>
<details>
<p><summary>See code</summary></p>
<pre class="bash"><code># Set required variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create a list of gwas-tissue combinations that haven&#39;t been calculated.
&gt; ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/todo.txt
for gwas in $(cat ${TWAS_rep}/Largest_TWAS_over15K_withoutUKBB_75comp.txt);do
for weights in $(cat ~/brc_scratch/Data/TWAS_sumstats/FUSION/snp_weight_list.txt);do
if [ ! -f ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/${weights}/UKBB.w_hm3.EUR.${weights}.${gwas}.fiprofile ]; then
echo $gwas $weights &gt;&gt; ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/todo.txt
fi
done
done

n=0
for i in $(seq 1 $(wc -l ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/todo.txt | cut -d &#39; &#39; -f 1));do
gwas=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $1}&#39; ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/todo.txt)
weights=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $2}&#39; ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/todo.txt)

qsub -N $(echo GeRS_job$(($n+10))) -hold_jid $(echo GeRS_job$(($n+0))) -l h_vmem=4G -pe smp 5 /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/scaled_functionally_informed_risk_scorer/scaled_functionally_informed_risk_scorer.R \
  --targ_feature_pred ${UKBB_output}/Predicted_expression/FUSION/EUR/${weights}/UKBB.w_hm3.QCd.AllSNP.FUSION.${weights}.predictions.gz \
  --ref_score ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights}.score \
  --ref_scale ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights}.scale \
  --pheno_name ${gwas} \
  --n_cores 5 \
  --pigz ${pigz_binary} \
  --output ${UKBB_output}/FunctionallyInformedPolygenicScores/EUR/${weights}/UKBB.w_hm3.EUR.${weights}.${gwas}

n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
