---
title: "Genotype-based Prediction (GenoPred)"
output: 
  html_document:
    toc: true
    theme: united
---

***

Welcome!

This website will provide details of a project evaluating methods for genotype-based prediction. The website will be updated continuously as the project develops.

***

#### Contents:

* Data Preparation
  * Genetic
    * [Reference genetic: European 503 subset of 1KG](https://opain.github.io/GenoPred/Pipeline_prep.html)
    * [Reference genetic: European 10K subset of UK Biobank](https://opain.github.io/GenoPred/Pipeline_prep_withUKBB_ref.html)
    * [Target genetic](https://opain.github.io/GenoPred/Genotype-based_scoring_in_target_samples.html)
  * Phenotypic
    * [Target phenotypic](https://opain.github.io/GenoPred/Target_sample_phenotype_prep.html)
    
* Main Research Questions
  * [Comparison of polygenic scoring methods](https://opain.github.io/GenoPred/Determine_optimal_polygenic_scoring_approach.html)
  * [Evaluating gene expression risk scores](https://opain.github.io/GenoPred/Functionally_informed_prediction.html) (under development)

* Side Projects 
  * [Estimating significance of model comparisons](https://opain.github.io/GenoPred/Z-test_vs_ANOVA_V2.html)
  * [Converting PRS-based relative risk to absolute risk](https://opain.github.io/GenoPred/PRS_relative_risk_experiment.html)
  
* Future developments
  * Evaluation of ancestry information for prediction
  * Stratification using pathway/tissue-specific genetic scores
  * Conversion of relative risk to absolute risk
  * Genetic prediction visualisation tools
  * Ancestry sensitive genetic prediction framework
  * Transferability of European GWAS to non-European populations

***

Please contact Oliver Pain (oliver.pain@kcl.ac.uk) to if you have any questions or comments.

***

I would like to thank the following people who have contributed to this project so far:

* **Cathryn Lewis**
* Lasse Folkersen
* Kylie Glanville
* Saskia Hagenaars
* Saskia Selzam
* Anna F&uuml;rtjes
* Helena Gaspar
* Jonathon Coleman
* Gerome Breen
* Robert Plomin

***

#### Funding:

This work is funded by the MRC (MR/N015746/1), and the National Institute for Health Research (NIHR) Biomedical Research Centre at South London and Maudsley NHS Foundation Trust and King’s College London. The authors acknowledge use of the research computing facility at King’s College London, Rosalind (https://rosalind.kcl.ac.uk), which is delivered in partnership with the NIHR Maudsley BRC, and part-funded by capital equipment grants from the Maudsley Charity (award 980) and Guy’s & St. Thomas’ Charity (TR130505). The views expressed are those of the authors and not necessarily those of the NHS, the NIHR or the Department of Health and Social Care. 

***

#### Change log:

**24/06/2020**:

* Change log started
* Updated Ancestry_identifier.R to use proper grid search in enet, allow other modelling methods to be specified, and improve model performance reporting.
