<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Preparing UKBB reference files for genotype-based scoring</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/united.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">GenoPred</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Preparing UKBB reference files for genotype-based scoring</h1>

</div>


<style>
p.caption {
  font-size: 1.5em;
}
</style>
<style type="text/css">
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>
<hr />
<p>Reference files for genotype-based scoring having already been prepared based on the 1000 Genomes Phase 3 data. However, the sample size is relatively small which can reduce the accuracy of LD and MAF estimates, and has reduced genetic diversity. Here we will prepare reference files based on UK Biobank sample.</p>
<p>The imputed UK Biobank data has already been harmonised with 1KG Phase 3 reference. First we need to create a subset (N=10,000) of European indidivuals. Then create reference files based on these individuals.</p>
<p><br/></p>
<hr />
<div id="genotypic-data" class="section level1">
<h1><span class="header-section-number">1</span> Genotypic data</h1>
<p>In this section we will download the required reference genotype data, and format it for use.</p>
<details>
<p><summary>Set required variables for command line</summary></p>
<pre class="bash"><code># Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config</code></pre>
</details>
<p><br/></p>
<hr />
<div id="create-keep-file-for-10k-european-individuals" class="section level2">
<h2><span class="header-section-number">1.1</span> Create keep file for 10K European individuals</h2>
<p>We have already calculated ancestry PCs for UK biobank and identified European ancestry based on 1KG Phase participants. Now create a random subset of 10K European individuals, which aren’t in the phenotype files used for prediction modelling.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

mkdir -p ${UKBB_output}/UKBB_ref/keep_files

module add general/R/3.5.0
R

library(data.table)

# Read in environmental variables
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config&#39;)

# Read in list of EUR UKBB individuals
EUR_UKBB&lt;-fread(paste0(UKBB_output, &#39;/Projected_PCs/Ancestry_idenitfier/UKBB.w_hm3.AllAncestry.EUR.keep&#39;))
names(EUR_UKBB)&lt;-c(&#39;FID&#39;,&#39;IID&#39;)

# Read in list of individuals used for one of the phenotypes
pheno&lt;-c(&#39;Depression&#39;,&#39;Intelligence&#39;,&#39;BMI&#39;,&#39;Height&#39;,&#39;T2D&#39;,&#39;CAD&#39;,&#39;IBD&#39;,&#39;MultiScler&#39;,&#39;RheuArth&#39;)

pheno_all&lt;-NULL
for(i in pheno){
  pheno_tmp&lt;-fread(paste0(UKBB_output, &#39;/Phenotype/PRS_comp_subset/UKBB.&#39;,i,&#39;.txt&#39;))
  pheno_all&lt;-rbind(pheno_all,pheno_tmp[,1:2,with=F])
}

pheno_all&lt;-pheno_all[!duplicated(paste(pheno_all$FID, pheno_all$IID, sep=&#39;-&#39;)),]

# List Eureopean individuals that are not in the phenotype files
pheno_all$UID&lt;-paste(pheno_all$FID, pheno_all$IID, sep=&#39;-&#39;)
EUR_UKBB$UID&lt;-paste(EUR_UKBB$FID, EUR_UKBB$IID, sep=&#39;-&#39;)
EUR_UKBB_nopheno&lt;-EUR_UKBB[!(EUR_UKBB$UID %in% pheno_all$UID),]

# Extract a random 10K individual
set.seed(1)
EUR_UKBB_nopheno_10K&lt;-EUR_UKBB_nopheno[sample(dim(EUR_UKBB_nopheno)[1],10000),]

EUR_UKBB_nopheno_10K$UID&lt;-NULL
write.table(EUR_UKBB_nopheno_10K, paste0(UKBB_output, &#39;/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep&#39;), col.names=F, row.names=F, quote=F)

q()
n
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="make-plink-file-containing-10k-eur-subset" class="section level1">
<h1><span class="header-section-number">2</span> Make plink file containing 10K EUR subset</h1>
<p>To speed up subsequent analyses, we will create plink files for the 10K EUR subset of UKBB. We will also apply a SNP missingness threshold of 0.02 to retain SNPs that are available in nearly all indiviuals. We will also output .freq files</p>
<details>
<p><summary>Create plink dataset</summary></p>
<pre class="bash"><code># Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

mkdir ${UKBB_output}/UKBB_ref/genotype

# Create plink dataset
for chr in $(seq 1 22);do
  qsub ${plink1_9} \
    --bfile ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr${chr} \
    --make-bed \
    --geno 0.02 \
    --keep ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep \
    --out ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr${chr}
done

# Create frq files for the plink dataset
for chr in $(seq 1 22);do
  qsub ${plink1_9} \
    --bfile ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr${chr} \
    --freq \
    --out ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr${chr}
done

# Create genome-wide version of dataset for lassosum
ls ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr*.bed | sed -e &#39;s/\.bed//g&#39; &gt; ${UKBB_output}/UKBB_ref/genotype/merge_list.txt

qsub ${plink1_9} \
  --merge-list ${UKBB_output}/UKBB_ref/genotype/merge_list.txt \
  --make-bed \
  --out ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.GW

rm ${UKBB_output}/UKBB_ref/genotype/merge_list.txt

# Create a file listing the keep file
echo EUR ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep &gt; ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="polygenic-scoring" class="section level1">
<h1><span class="header-section-number">3</span> Polygenic scoring</h1>
<p>In this section we will create files that can be used for polygenic scoring (score files), estimate the mean and SD of polygenic scores within different ancestries for scaling target samples. This will be performed using MAF and LD in the UKBB 10K European subset.</p>
<p>Polygenic scores are derived using multiple methods to allow comparison. After the best method has been idenitified, it will ony be necessary to calculate scores using the one method.</p>
<details>
<p><summary>Set required variables</summary></p>
<pre class="bash"><code># Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config</code></pre>
</details>
<p><br/></p>
<hr />
<div id="prepare-score-files-and-scaling-files-for-polygenic-scoring-pt-clump" class="section level2">
<h2><span class="header-section-number">3.1</span> Prepare score files and scaling files for polygenic scoring (pT + clump)</h2>
<p>Here I prepare reference files for typical polygenic scores derived using the p-value thresholding and LD-based clumping procedure.</p>
<div id="sparse-thresholding-nested" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Sparse thresholding (nested)</h3>
<p>Here we will only use 8 p-value thresholds.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

###
# Run for all in ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt
###
# Create file listing GWAS that haven&#39;t been processed.
mkdir -p ${UKBB_output}/UKBB_ref/Score_files_for_polygenic
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic/todo.txt
fi
done

n=0
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic/todo.txt);do
qsub -N $(echo job$(($n+20))) -hold_jid $(echo job$(($n+0))) -l h_vmem=6G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
  --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
  --plink ${plink1_9} \
  --memory 3000 \
  --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
n=$(($n+1))
done
</code></pre>
</details>
<p><br/></p>
</div>
<div id="sparse-thresholding-not-nested" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Sparse thresholding (not nested)</h3>
<p>Here we will only use 8 p-value thresholds.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

###
# Run for all in ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt
###
# Create file listing GWAS that haven&#39;t been processed.
mkdir ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested/todo.txt
fi
done

n=0
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested/todo.txt);do
qsub -N $(echo job$(($n+20))) -hold_jid $(echo job$(($n+0))) -l h_vmem=6G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator/polygenic_score_file_creator.R \
  --ref_plink_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
  --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
  --plink ${plink1_9} \
  --memory 3000 \
  --nested F \
  --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_non_nested/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
n=$(($n+1))
done
</code></pre>
</details>
<p><br/></p>
</div>
<div id="dense-thresholding" class="section level3">
<h3><span class="header-section-number">3.1.3</span> Dense thresholding</h3>
<p>Here we will only use dense p-value thresholds, mimicking the PRSice approach.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_PRSice">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

###
# Run for best GWAS for target traits in UKBB
###

# Create file listing GWAS that haven&#39;t been processed.
mkdir ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense/todo.txt
fi
done

for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense/todo.txt);do
qsub -l h_vmem=6G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_dense/polygenic_score_file_creator_dense.R \
  --ref_plink_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
  --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
  --plink ${plink1_9} \
  --prsice_path ${prsice_path} \
  --rscript ${rscript} \
  --memory 3000 \
  --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_dense/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="prepare-score-and-scale-files-for-polygenic-scoring-using-lassosum" class="section level2">
<h2><span class="header-section-number">3.2</span> Prepare score and scale files for polygenic scoring using lassosum</h2>
<p>Here we create reference files for polygenic scores calculated by lassosum, a method for performing lasso-based shrinkage to GWAS sumstats to account for LD and winners curse. More information on lassosum can be found <a href="https://github.com/tshmak/lassosum">here</a>. You will need to install the lassosum R package in advance.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator_lassosum.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_lassosum">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

###
# Run for best GWAS for target traits in UKBB
###

mkdir ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum/todo.txt
fi
done

for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum/todo.txt);do
sbatch --mem 10G -p shared,brc /users/k1806347/brc_scratch/Software/Rscript_for_lassosum.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_lassosum/polygenic_score_file_creator_lassosum.R \
    --ref_plink_gw ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.GW \
    --ref_keep ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep \
    --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
    --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_lassosum/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
    --plink ${plink1_9} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="prepare-score-and-scale-files-for-polygenic-scoring-using-prscs" class="section level2">
<h2><span class="header-section-number">3.3</span> Prepare score and scale files for polygenic scoring using PRScs</h2>
<p>Here we create reference files for polygenic scores calculated by PRScs, a method for performing Bayesian continuous shrinkage to GWAS sumstats to account for LD and winners curse. More information on PRScs can be found <a href="https://github.com/getian107/PRScs">here</a>. You will need to download the PRScs software and reference data in advance.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator_PRScs.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_PRScs">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

# Create directory for the output
mkdir ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs

#######
# Run for best GWAS for target traits in UKBB
#######
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs/todo.txt
fi
done

# Run script for selected GWAS
n=0
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs/todo.txt);do
qsub -N $(echo job$(($n+3))) -hold_jid $(echo job$(($n+0))) -pe smp 6 -l h_vmem=6G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_PRScs/polygenic_score_file_creator_PRScs.R \
  --ref_plink_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
  --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
  --plink ${plink1_9} \
  --memory 5000 \
  --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_PRScs/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list \
  --PRScs_path ${PRScs_dir}/PRScs.py \
  --PRScs_ref_path ${PRScs_dir}/ldblk_1kg_eur \
  --n_cores 6 \
  --phi_param 1e-6,1e-4,1e-2,1,auto

n=$(($n+1))
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="prepare-score-and-scale-files-for-polygenic-scoring-using-s-blup" class="section level2">
<h2><span class="header-section-number">3.4</span> Prepare score and scale files for polygenic scoring using S-BLUP</h2>
<p>Here we create reference files for polygenic scores calculated by SBLUP, a method for performing genomic BLUP analysis with summary data and an LD-reference. More information on SBLUP can be found <a href="https://cnsgenomics.com/software/gcta/#Genomicriskprediction">here</a>. You will need to download the GCTA software in advance.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator_SBLUP.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_SBLUP">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

#######
# Run for best GWAS for target traits in UKBB
#######

&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBLUP/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBLUP/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBLUP/todo.txt
fi
done

# Run script for selected GWAS
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBLUP/todo.txt);do
qsub -pe smp 3 -l h_vmem=12G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_SBLUP/polygenic_score_file_creator_SBLUP.R \
  --ref_plink ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.GW \
    --ref_keep ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep \
  --ref_freq_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
  --sumstats ${gwas_rep}/${gwas}.sumstats.gz \
  --plink ${plink1_9} \
  --gcta ${gcta} \
  --ldsc ${ldsc} \
  --ldsc_ref ${ldsc_ref} \
  --memory 25000 \
  --n_cores 3 \
  --output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBLUP/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
  --ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="prepare-score-and-scale-files-for-polygenic-scoring-using-sbayesr" class="section level2">
<h2><span class="header-section-number">3.5</span> Prepare score and scale files for polygenic scoring using SBayesR</h2>
<p>Here we create reference files for polygenic scores calculated by SBayesR, a bayesian shrinkage method for GWAS summary data and an LD-reference. More information on SBayesR can be found <a href="http://cnsgenomics.com/software/gctb/#SummaryBayesianAlphabet">here</a>. You will need to download the GCTB software in advance.</p>
<p>First we need to create a specially formatted LD matrix for SBayesR. The GCTB authors compare the performance of SBayesR when using different datasets for LD matrix estimation. They show that using EUR 1KG data (N=378) leads to poorer prediction accuracy from S-BayesR. They then use LD matrices based on 50,000 random European individuals from UK Biobank to provide optimal prediction, however using 5000 indivduals gave similar results. I think we should be comparing PRS methods based on the same reference data, so we should estimate LD matrices based on the EUR 1KG reference. A subsequent study (by me) can test these PRS methods based on LD estimates from 5000 individuals.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

# When retaining only EUR, I am getting an error saying fixed SNP. Create list of variants with MAF &gt; 0.001.
# Download the required genetic maps
cd ${genetic_map}/CEU
for chr in $(seq 1 22); do
  wget https://github.com/joepickrell/1000-genomes-genetic-maps/raw/master/interpolated_OMNI/chr${chr}.OMNI.interpolated_genetic_map.gz
done

gunzip *.gz

mkdir -p ${UKBB_output}/UKBB_ref/LD_matrix/EUR

module add apps/R/3.6.0
R

# Read in environmental variables
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config&#39;)
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config&#39;)

# Create list of SNPs that have MAF above 0.001 in the EUR 1KG sample
for(i in 1:22){
  frq&lt;-read.table(paste0(UKBB_output, &#39;/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr&#39;,i,&#39;.frq&#39;), header=T)
  frq&lt;-frq[frq$MAF &gt; 0.001,]
  write.table(frq$SNP,paste0(UKBB_output, &#39;/UKBB_ref/LD_matrix/EUR/SNP_maf0.001_EUR_chr&#39;,i,&#39;.txt&#39;), col.names=F, row.names=F, quote=F)
}

# Stop scientific notation
options(scipen=999)

# Create shrunk LD matrix in 5000 SNP pieces
for(i in 1:22){
  nsnp&lt;-system(paste0(&#39;wc -l &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/SNP_maf0.001_EUR_chr&#39;,i,&#39;.txt&#39;), intern=T)
  nsnp&lt;-as.numeric(unlist(strsplit(nsnp, &#39; &#39;))[1])
  nsnp_chunk&lt;-ceiling(nsnp/5000)
  for(j in 1:nsnp_chunk){
    start&lt;-(5000*(j-1))+1
    end&lt;-5000*j
    print(start)
    print(end)

    system(paste0(&#39;sbatch -p brc,shared --mem 5G &#39;,gctb,&#39; --bfile &#39;,UKBB_output,&#39;/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr&#39;,i,&#39; --make-shrunk-ldm --extract &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/SNP_maf0.001_EUR_chr&#39;,i,&#39;.txt --gen-map &#39;,genetic_map,&#39;/CEU/chr&#39;,i,&#39;.OMNI.interpolated_genetic_map --snp &#39;,start,&#39;-&#39;,end,&#39; --out &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr&#39;,i))

  }
}

# Merge the chunks into per chromosome LD matrices
for(i in 1:22){
  nsnp&lt;-system(paste0(&#39;wc -l &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/SNP_maf0.001_EUR_chr&#39;,i,&#39;.txt&#39;), intern=T)
  nsnp&lt;-as.numeric(unlist(strsplit(nsnp, &#39; &#39;))[1])
  nsnp_chunk&lt;-ceiling(nsnp/5000)
  files&lt;-list.files(path=paste0(UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/&#39;), pattern=paste0(&#39;UKBB.noPheno.EUR.10K.chr&#39;,i,&#39;.snp&#39;))
  files&lt;-files[grepl(&#39;.bin&#39;,files)]
  files&lt;-paste0(UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/&#39;,files)
  if(length(files) == nsnp_chunk){
    files&lt;-gsub(&#39;.bin&#39;, &#39;&#39;, files)
    file_num&lt;-gsub(&#39;.*.snp&#39;, &#39;&#39;, files)
    file_num&lt;-as.numeric(gsub(&#39;-.*&#39;, &#39;&#39;, file_num))
    files&lt;-files[order(file_num)]
    # Sort files in order of genomic location otherwise SBayeR does not converge!
    write.table(files,paste0(UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/shrunk_ld_chr&#39;,i,&#39;merge_list&#39;), col.names=F, row.names=F, quote=F)
    system(paste0(&#39;sbatch -p brc,shared --mem &#39;,round(2.5*nsnp_chunk),&#39;G &#39;,gctb,&#39; --mldm &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/shrunk_ld_chr&#39;,i,&#39;merge_list --make-shrunk-ldm --out &#39;,UKBB_output,&#39;/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr&#39;,i))
  } else {
    print(&#39;Not all chunks finished&#39;)
    print(files)
  }
}

q()
n

# Delete temporary files
for chr in $(seq 1 22);do
rm ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr${chr}.snp*
rm ${UKBB_output}/UKBB_ref/LD_matrix/EUR/SNP_maf0.001_EUR_chr${chr}.txt
rm ${UKBB_output}/UKBB_ref/LD_matrix/EUR/shrunk_ld_chr${chr}merge_list
done

# Make the LD matrices sparse
for chr in $(seq 1 22);do 
sbatch -p brc,shared --mem 50G ${gctb} --ldm ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr${chr}.ldm.shrunk --chisq 0 --make-sparse-ldm --out ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr${chr}
done

# Delete temporary files
for chr in $(seq 1 22);do
rm ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr${chr}.ldm.shrunk.*
done

# Make a list of shrunk sparse matrices
ls ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr*.ldm.sparse.bin | sed &quot;s/.bin//&quot; &gt; ${Geno_1KG_dir}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.sparse.ldm.list</code></pre>
</details>
<p>Now the LD reference is ready, we can calculate SBayesR shrunk GWAS summary statistics. This section uses an R script called ‘polygenic_score_file_creator_SBayes.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_SBLUP">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

# Create directory for the output
mkdir ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR

# Create list of gwas that PRScs scores should calculated for
&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR/todo.txt
fi
done

# Run script for selected GWAS
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR/todo.txt);do
sbatch -J SBayesR -n 8 --mem 50G -p brc,shared /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_SBayesR/polygenic_score_file_creator_SBayesR.R \
--ref_plink ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.GW \
--ref_keep ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep \
--ref_freq_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
--sumstats ${gwas_rep}/${gwas}.sumstats.gz \
--plink ${plink1_9} \
--gctb ${gctb} \
--ld_matrix_chr ${UKBB_output}/UKBB_ref/LD_matrix/EUR/UKBB.noPheno.EUR.10K.chr \
--memory 40000 \
--n_cores 8 \
--output ${UKBB_output}/UKBB_ref/Score_files_for_polygenic_SBayesR/${gwas}/UKBB.noPheno.EUR.10K.w_hm3.${gwas} \
--ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
n=$(($n+1))
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="prepare-score-and-scale-files-for-polygenic-scoring-using-ldpred" class="section level2">
<h2><span class="header-section-number">3.6</span> Prepare score and scale files for polygenic scoring using LDPred</h2>
<p>Here we create reference files for polygenic scores calculated by LDPred, a method for performing bayesian shrinkage analysis with summary data and an LD-reference. More information on LDPred can be found <a href="https://github.com/bvilhjal/ldpred">here</a>.</p>
<p>This section uses an R script called ‘polygenic_score_file_creator_LDPred.R’. Further information the usage of this script can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/polygenic_score_file_creator_LDPred">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Pipeline_prep.config

# Create directory for the output
mkdir ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred

#######
# Run for best GWAS for target traits
#######

&gt; ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred/todo.txt
for gwas in $(echo DEPR06 COLL01 HEIG03 BODY03 DIAB05 COAD01 CROH01 SCLE02 RHEU01 EDUC03 ADHD04 BODY11);do
if [ ! -f ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred/${gwas}/1KGPhase3.w_hm3.${gwas}.EUR.scale ]; then
echo $gwas &gt;&gt; ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred/todo.txt
fi
done

# Run script for selected GWAS
for gwas in $(cat ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred/todo.txt | tail -n +6);do
sbatch -p brc,shared --mem 50G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/polygenic_score_file_creator_LDPred/polygenic_score_file_creator_LDPred.R \
--ref_plink ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.GW \
--ref_keep ${UKBB_output}/UKBB_ref/keep_files/UKBB_noPheno_EUR_10K.keep \
--ref_freq_chr ${UKBB_output}/UKBB_ref/genotype/UKBB.noPheno.EUR.10K.chr \
--sumstats ${gwas_rep}/${gwas}.sumstats.gz \
--plink ${plink1_9} \
--memory 30000 \
--n_cores 1 \
--ldpred ${ldpred} \
--output ${UKBB_output}/UKBB_ref/Score_files_for_poylygenic_LDPred/${gwas}/1KGPhase3.w_hm3.${gwas} \
--ref_pop_scale ${UKBB_output}/UKBB_ref/keep_files/keep_file.list
done
</code></pre>
</details>
<p><br/></p>
<hr />
</div>
</div>
<div id="functionally-informed-polygenic-scoring" class="section level1">
<h1><span class="header-section-number">4</span> Functionally-informed polygenic scoring</h1>
<p>In this section we will create polygenic scores based on transcriptome-wide association study (TWAS) results, and predicted expression. In brief TWAS and predicted expression is performed by integrating information on how SNPs affect gene expression with GWAS sumstats or genotypic data. More information about TWAS can be found <a href="http://gusevlab.org/projects/fusion/">here</a>. It is also possible to integrate information other gene expression, and so these scores can be more generally referred to as functionally-informed polygenic scores.</p>
<p>Unlike polygenic scores, functionally informed polygenic scores can be tissue- and developmental stage-specific, providing an oppurtunity to stratify individuals into different components of risk. Functional inference is carried out using SNP-weights, jointly estimated SNP-effects on a functional change to a gene. The joint estimation of SNP-effects on a gene in a functionally aware manner may explain variance not captured by a standard polygenic score. However, functionally informed polygenic scores typically only considers variation proximal to genes and will therefore typically explain less than a genome-wide polygenic score.</p>
<details>
<p><summary>Set required variables</summary></p>
<pre class="bash"><code># Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config</code></pre>
</details>
<p><br/></p>
<hr />
<div id="perform-twas-using-all-tissue-data-and-gwas" class="section level2">
<h2><span class="header-section-number">4.1</span> Perform TWAS using all tissue data and GWAS</h2>
<p>Here we perform a TWAS using FUSION software, FUSION SNP-weights for multiple tissues, and all GWAS previously selected. You will need to download the FUSION software and SNP-weights in advance. Here I use a FUSION repository that I created on the Rosalind server and King’s College London</p>
<p>This section uses a shell script called ‘Run_TWAS.sh’ to run an R script called ‘Run_TWAS.R’ for multiple GWAS as an array. Further information these scripts can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/Run_TWAS">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

mkdir -p ${TWAS_rep}

module add general/R/3.5.0
R

# Read in environmental variables
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Target_scoring.config&#39;)

library(data.table)

# Read in full list of SNP-weights
weights&lt;-list.files(path=paste0(FUSION_dir,&#39;/SNP-weights/.&#39;), recursive=F)

# Write a file listing all SNP-weight sets.
write.table(weights, paste0(TWAS_rep, &#39;/snp_weight_list.txt&#39;, col.names=F, row.names=F, quote=F)

# Create a .pos file containing SNP-weights for all tissues.
pos&lt;-NULL
for(i in weights){
  pos_tmp&lt;-data.frame(fread(paste0(&#39;${FUSION_dir}/SNP-weights/&#39;,i,&#39;/&#39;,i,&#39;.pos&#39;)))
  pos_tmp$PANEL&lt;-i
  pos_tmp$WGT&lt;-paste0(i,&#39;/&#39;,pos_tmp$WGT)
  pos&lt;-rbind(pos,pos_tmp[c(&#39;PANEL&#39;,&#39;WGT&#39;,&#39;ID&#39;,&#39;CHR&#39;,&#39;P0&#39;,&#39;P1&#39;)])
}

write.table(pos, paste0(TWAS_rep,&#39;/All_tissues.pos&#39;, col.names=T, row.names=F, quote=F)

q()
n

# Perform TWAS for all selected GWAS and all tissues
# Create file listing GWAS that haven&#39;t been converted to TWAS
###
# Run for all in ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt
###
&gt; ${TWAS_rep}/todo.txt
for gwas in $(cut -f 1 -d &#39; &#39; ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt);do
if [ ! -f ${TWAS_rep}/${gwas}/${gwas}_res_GW.txt ]; then
grep $gwas ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt &gt;&gt; ${TWAS_rep}/todo.txt
fi
done

qsub -l h_vmem=6G -pe smp 8 -tc 3 -t 1-$(wc -l ${TWAS_rep}/todo.txt | cut -d&#39; &#39; -f1) /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/Run_TWAS/Run_TWAS.sh \
  --gwas_list ${TWAS_rep}/todo.txt \
  --pos ${TWAS_rep}/All_tissues.pos \
  --outdir ${TWAS_rep} \
  --fusion_dir ${FUSION_dir} \
  --ncores 8

# Create a list of TWAS that have less than 25% missing TWAS.Z
module add general/R/3.5.0
R

# Read in environmental variables
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Target_scoring.config&#39;)

library(data.table)
gwas&lt;-fread(&#39;~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutUKBB.txt&#39;, header=F)$V1
twas_comp&lt;-NULL
twas_incomp&lt;-NULL
for(gwas_name in gwas){
    twas&lt;-fread(paste0(TWAS_rep,&#39;/&#39;,gwas_name,&#39;/&#39;,gwas_name,&#39;_res_GW.txt&#39;))
    if((sum(is.na(twas$TWAS.Z)) / dim(twas)[1]) &lt; 0.25){
        twas_comp&lt;-c(twas_comp,gwas_name)
    } else {
        twas_incomp&lt;-c(twas_incomp,gwas_name)
    }
}

write.table(twas_comp, paste0(TWAS_rep,&#39;/Largest_TWAS_over15K_withoutUKBB_75comp.txt&#39;), col.names=F, row.names=F, quote=F)
q()
n

###
# Run for all in ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutTEDS.txt
###
&gt; ${TWAS_rep}/todo.txt
for gwas in $(cut -f 1 -d &#39; &#39; ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutTEDS.txt);do
if [ ! -f ${TWAS_rep}/${gwas}/${gwas}_res_GW.txt ]; then
grep $gwas ~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutTEDS.txt &gt;&gt; ${TWAS_rep}/todo.txt
fi
done

qsub -l h_vmem=6G -pe smp 8 -tc 3 -t 1-$(wc -l ${TWAS_rep}/todo.txt | cut -d&#39; &#39; -f1) /users/k1806347/brc_scratch/Software/Scripts/GenoPred_Pipeline/Run_TWAS_new.sh \
  --gwas_list ${TWAS_rep}/todo.txt \
  --pos ${TWAS_rep}/All_tissues.pos \
  --outdir ${TWAS_rep} \
  --ncores 8
  
# Create a list of TWAS that have less than 25% missing TWAS.Z
module add general/R/3.5.0
R

# Read in environmental variables
source(&#39;/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Target_scoring.config&#39;)

library(data.table)
gwas&lt;-fread(&#39;~/brc_scratch/Data/GWAS_sumstats/Largest_GWAS_over15K_withoutTEDS.txt&#39;, header=F)$V1
twas_comp&lt;-NULL
twas_incomp&lt;-NULL
for(gwas_name in gwas){
    twas&lt;-fread(paste0(TWAS_rep,&#39;/&#39;,gwas_name,&#39;/&#39;,gwas_name,&#39;_res_GW.txt&#39;))
    if((sum(is.na(twas$TWAS.Z)) / dim(twas)[1]) &lt; 0.25){
        twas_comp&lt;-c(twas_comp,gwas_name)
    } else {
        twas_incomp&lt;-c(twas_incomp,gwas_name)
    }
}

write.table(twas_comp, paste0(TWAS_rep,&#39;/Largest_TWAS_over15K_withoutTEDS_75comp.txt&#39;), col.names=F, row.names=F, quote=F)
q()
n</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="predict-functional-genomic-features" class="section level2">
<h2><span class="header-section-number">4.2</span> Predict functional genomic features</h2>
<p>In order to calculate functionally informed polygenic scores, we need the Z scores for each feature (i.e. gene expression), and we need predictions of each feature in the target sample. The functionally informed polygenic scores are then sum(Z*Pred).</p>
<p>Here we calculate predictions for each genomic feature in the reference sample, to estimate the mean and SD of each feature in the different ancestries for scaling predictions in the target samples, and for calculating the correlation between features for clumping and oher analyses that take the correlation between features into account (e.g. <a href="https://github.com/opain/TWAS-GSEA">TWAS-GSEA</a>).</p>
<p>First FUSION SNP-weights must be converted into SCORE files for use in PLINK. This is done using the script called ‘FUSION_score_file_creator.R’ (info <a href="https://github.com/opain/GenoPred/tree/master/Scripts/FUSION_score_file_creator">here</a>)</p>
<p>Then the score files are used to predict features in the 1000 Genomes reference and report the mean and SD. This is done using the script called ‘FUSION_ref_scorer.R’ (info <a href="https://github.com/opain/GenoPred/tree/master/Scripts/FUSION_ref_scorer">here</a>)</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

mkdir -p ~/brc_scratch/Data/FUSION/SCORE_FILES

# Create SCORE files and expression reference.
qsub -pe smp 15 -l h_vmem=2G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/FUSION_score_file_creator/FUSION_score_file_creator.R \
  --weights ${TWAS_rep}/All_tissues.pos \
  --weights_dir ${FUSION_dir}/SNP-weights \
  --output ~/brc_scratch/Data/FUSION/SCORE_FILES \
  --n_cores 15

# Calculate features based on each SNP-weight set. To not swamp the cluster, submit only three weights at a time.
# Create file listing weights that haven&#39;t been predicted in the reference.
&gt; ${Geno_1KG_dir}/Predicted_expression/FUSION/todo.txt
for weights in $(cat ${TWAS_rep}/snp_weight_list.txt);do
if [ ! -f ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.predictions.gz ]; then
echo $weights &gt;&gt; ${Geno_1KG_dir}/Predicted_expression/FUSION/todo.txt
fi
done

n=0
for weights in $(cat ${Geno_1KG_dir}/Predicted_expression/FUSION/todo.txt);do
qsub -N $(echo job$(($n+2))) -hold_jid $(echo job$(($n+0))) -pe smp 5 -l h_vmem=8G /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/FUSION_ref_scorer/FUSION_ref_scorer.R \
  --ref_plink_chr ${Geno_1KG_dir}/1KGPhase3.w_hm3.chr \
  --weights ${FUSION_dir}/SNP-weights/${weights}/${weights}.pos \
  --output ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights} \
  --plink ${plink1_9} \
  --n_cores 5 \
  --pigz ${pigz_binary} \
  --score_files ~/brc_scratch/Data/FUSION/SCORE_FILES \
  --ref_pop_scale ${Geno_1KG_dir}/super_pop_keep.list
n=$(($n+1))
done</code></pre>
</details>
<p><br/></p>
<hr />
</div>
<div id="prepare-score-files-and-scaling-files-for-functionally-informed-polygenic-scores" class="section level2">
<h2><span class="header-section-number">4.3</span> Prepare score files and scaling files for functionally informed polygenic scores</h2>
<p>Now we will create reference files for functionally informed polygenic scores, including score files after clumping, and the mean and SD of scores across ancestries for scaling the target sample.</p>
<p>This section uses an R script called ‘ref_funtionally_informed_risk_scorer.R’ for multiple GWAS as an array. Further information on this scripts can be found <a href="https://github.com/opain/GenoPred/tree/master/Scripts/ref_funtionally_informed_risk_scorer">here</a>.</p>
<details>
<p><summary>Show code</summary></p>
<pre class="bash"><code>. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Target_scoring.config

######
# Calculate for /users/k1806347/brc_scratch/Data/TWAS_sumstats/FUSION/Largest_TWAS_over15K_withoutUKBB_75comp.txt
######
&gt; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo.txt
for gwas in $(cat ${TWAS_rep}/Largest_TWAS_over15K_withoutUKBB_75comp.txt);do
for weights in $(cat ${TWAS_rep}/snp_weight_list.txt);do
if [ ! -f ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights}.scale ]; then
echo $gwas $weights &gt;&gt; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo.txt
fi
done
done

for i in $(seq 1 $(wc -l ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo.txt | cut -d &#39; &#39; -f 1));do
gwas=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $1}&#39; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo.txt)
weights=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $2}&#39; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo.txt)

qsub /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Scripts/ref_funtionally_informed_risk_scorer/ref_funtionally_informed_risk_scorer.R \
  --twas_results ${TWAS_rep}/${gwas}/${gwas}_res_GW.txt \
  --ref_feature_pred ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.predictions.gz \
  --output ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights} \
  --ref_keep ${Geno_1KG_dir}/keep_files/EUR_samples.keep \
  --ref_scale ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.EUR.scale \
  --panel ${weights}

sleep 1
done


#####
# Calculate GeRSs weighting genes by tissue specificity and TWAS.Z
#####

# Create files containing tissue specific weights for each gene
module add general/R/3.5.0
R
library(data.table)

Tissue_Z&lt;-fread(&#39;/users/k1806347/brc_scratch/Data/GTeX/GTEx_v7_gene_median_tpm.tmp1.windsorised50.pseudo1.log2.Z.tsv&#39;)

Tissue_Z$gene_id&lt;-NULL
names(Tissue_Z)[1]&lt;-&#39;ID&#39;

# Scale the Z scores to be between 0 and 1 for each tissue
Tissue_Z_tmp&lt;-Tissue_Z[,-1]
Tissue_Z_tmp&lt;-pnorm(as.matrix(Tissue_Z_tmp)
Tissue_Z_01&lt;-data.frame(Tissue_Z[,1:2], Tissue_Z_tmp)

# Write file out seperately for each tissue
for(i in names(Tissue_Z)[-1]){
  write.table(Tissue_Z[,c(&#39;ID&#39;,i),with=F], paste0(&#39;/users/k1806347/brc_scratch/Data/GTeX/GTex_v7_TissueSpecificity_Weight.&#39;,i,&#39;.txt&#39;), col.names=T, row.names=F, quote=F)
}

# Extract genes in top 95% confidence interval

# Create score files
&gt; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo_TissueWeighted.txt
for gwas in $(cat /users/k1806347/brc_scratch/Data/TWAS_sumstats/FUSION/Largest_TWAS_over15K_withoutUKBB_75comp.txt);do
for weights in $(cat ${TWAS_rep}/snp_weight_list.txt);do
if [ ! -f ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights}.TissueWeighted.scale ]; then
echo $gwas $weights &gt;&gt; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo_TissueWeighted.txt
fi
done
done

for i in $(seq 1 $(wc -l ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo_TissueWeighted.txt | cut -d &#39; &#39; -f 1));do
gwas=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $1}&#39; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo_TissueWeighted.txt)
weights=$(awk -v var=&quot;$i&quot; &#39;NR == var {print $2}&#39; ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/todo_TissueWeighted.txt)

qsub /users/k1806347/brc_scratch/Software/Rscript.sh /users/k1806347/brc_scratch/Software/Scripts/GenoPred_Pipeline/ref_funtionally_informed_risk_scorer.R \
  --twas_results /users/k1806347/brc_scratch/Data/TWAS_sumstats/FUSION/${gwas}/${gwas}_res_GW.txt \
  --ref_feature_pred ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.predictions.gz \
  --output ${Geno_1KG_dir}/Score_files_for_functionally_informed_risk_scores/${gwas}/1KGPhase3.w_hm3.EUR.FUSION.${gwas}.${weights}.TissueWeighted \
  --ref_keep ${Geno_1KG_dir}/keep_files/EUR_samples.keep \
  --ref_scale ${Geno_1KG_dir}/Predicted_expression/FUSION/${weights}/1KGPhase3.w_hm3.FUSION.${weights}.EUR.scale \
  --panel ${weights} \
  --r2_weighted T

sleep 5
done
</code></pre>
</details>
<p><br/></p>
<hr />
<p>Session info:</p>
<details>
<p><summary>R session</summary></p>
<pre><code>## Error in library(&quot;doMC&quot;): there is no package called &#39;doMC&#39;</code></pre>
<pre><code>## Error in library(&quot;optparse&quot;): there is no package called &#39;optparse&#39;</code></pre>
<pre><code>## Error in library(&quot;foreach&quot;): there is no package called &#39;foreach&#39;</code></pre>
<pre><code>## Error in library(&quot;caret&quot;): there is no package called &#39;caret&#39;</code></pre>
<pre><code>## R version 3.6.2 (2019-12-12)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.3.5.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] cowplot_1.0.0     ggplot2_3.2.1     data.table_1.12.8 knitr_1.28       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3       magrittr_1.5     munsell_0.5.0    colorspace_1.4-1
##  [5] R6_2.4.1         rlang_0.4.4      stringr_1.4.0    highr_0.8       
##  [9] tools_3.6.2      grid_3.6.2       gtable_0.3.0     xfun_0.12       
## [13] withr_2.1.2      htmltools_0.4.0  yaml_2.2.1       lazyeval_0.2.2  
## [17] digest_0.6.25    tibble_2.1.3     lifecycle_0.1.0  crayon_1.3.4    
## [21] evaluate_0.14    rmarkdown_2.1    stringi_1.4.6    compiler_3.6.2  
## [25] pillar_1.4.3     scales_1.1.0     pkgconfig_2.0.3</code></pre>
</details>
<details>
<p><summary>Software versions</summary></p>
<ul>
<li>PLINK v1.90b3.31 64-bit (3 Feb 2016)</li>
<li>PLINK v2.00a2LM 64-bit Intel (9 Mar 2019)</li>
<li>PRScs (downloaded from GitHub 5 Jul 2019)</li>
<li>FUSION Software and SNP-weights (downloaded from FUSION website and GitHub 30th November 2018)</li>
<li>GCTA Version 1.26.0</li>
<li>LDSC Version 1.0.0 (downloaded from GitHub 5 Nov 2018)</li>
</ul>
</details>
<p><br/></p>
<hr />
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
