---
title: "Training Sample Phenotype Preparation"
output: 
  html_document:
    toc: true
    theme: united
    toc_depth: 3
    number_sections: true
    toc_float: true
    fig_caption: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error=TRUE)
```

<style>
p.caption {
  font-size: 1.5em;
}
</style>

```{css, echo=F}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: auto !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
```

***

To evaluate genotype-based prediction approaches, I am using a range of phenotypes measured in the UK Biobank and TEDS samples.

# UKBB

Phenotype:

* Depression
* Intelligence
* BMI
* Height
* CAD (Coronary artery disease)
* T2D (Type II diabetes)
* Inflammatory Bowel Disorder (IBD)
* Rheumatoid arthritis (RheumArth)
* Multiple Slerosis (MultiScler)

Below I provide a description of how each phenotype was derived and some descriptive statistics.

```{bash, echo=F, eval=F}
######
# Update file listing IDs after QC to match genetic data
######

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)
keep<-fread(qc_indiv_list)
fam<-fread(paste0(UKBB_output,'/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam'))

keep_fam<-merge(keep, fam, by=c('V2'))
keep_new<-keep_fam[,c('V1.y','V2'), with=F]

fwrite(keep_new, paste0(Results_dir,'/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam'), sep=' ', na='NA', col.names=F)

q()
n
```

<br/>

***

## Depression
The depression phenotype was shared with me by Kylie Glanville.

<details><summary>Preparation of Depression phenotype</summary>
```{bash, echo=T, eval=F}
######
# Modify depression phenotype file so IDs match genotypic data
######

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)
pheno<-fread(paste0(UKBB_output, '/Phenotype/Depression/ever_depressed_pheno_final.txt'))
fam<-fread(paste0(UKBB_output,'/UKBB/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam')
names(fam)[1:2]<-c('FID','IID')

pheno_fam<-merge(pheno, fam, by=c('IID'))
pheno_new<-pheno_fam[,c('FID.y','IID','Depressed.Ever.Clean'), with=F]
names(pheno_new)[1]<-'FID'

fwrite(pheno_new, paste0(UKBB_output,'/Phenotype/Depression/ever_depressed_pheno_final.UpdateIDs.txt'), sep=' ', na='NA')

q()
n
```
</details>

<br/>

***

## Intelligence
The Intelligence phenotype is derived from the Fluid.intelligence.score.0.0 (f.20191.0.0) variable.

<details><summary>Preparation of Intelligence phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Extract intelligence variable, update IDs and calculate descriptives
#####
module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

# Read in the field finder to find he code for the phenotype of interest.
field_finder<-read.table(paste0(ukbb_pheno_file,'/ukb18177_glanville_field_finder.txt'), stringsAsFactors=F)
field_finder[grepl('Fluid.intelligence', field_finder$V3),]
# The Fluid intelligence variable has a code of f.20191.0.0 and is called Fluid.intelligence.score.0.0

# Identify which column in the phenotype file contains the relevant code
library(data.table)
pheno_header<-names(data.frame(fread(cmd=paste0('head -n 1 ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt'))))
rel_cols<-c(1,which(grepl('f.20191.0.0',pheno_header)))

# Read in the IID and fluid intelligence columns
pheno<-data.frame(fread(cmd=paste0('cut -f ',paste(rel_cols,collapse=','),' ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt')))
names(pheno)<-c('IID','Fluid.intelligence.score')

# Remove individuals with missing phenotype data
pheno<-pheno[complete.cases(pheno),]

# Calculate descriptive statistics for the phenotype
library(e1071)
descriptives<-data.frame(Phenotype='Fluid.intelligence.score',
                        Code='f.20127.0.0',
                        Mean=mean(pheno$Fluid.intelligence.score),
                        SD=sd(pheno$Fluid.intelligence.score),
                        Min=min(pheno$Fluid.intelligence.score),
                        Max=max(pheno$Fluid.intelligence.score),
                        Skewness=skewness(pheno$Fluid.intelligence.score),
                        N=dim(pheno)[1])

system(paste0('mkdir ',UKBB_output,'/Phenotype/Intelligence'))
write.table(descriptives,paste0(UKBB_output,'/Phenotype/Intelligence/Descriptives.txt'), row.names=F, quote=F)

# Plot the distribution
png(paste0(UKBB_output,'/Phenotype/Intelligence/Histogram.png'))
hist(pheno$Fluid.intelligence.score)
dev.off()

# Write out the phenotype in PLINK format.
pheno_plink<-data.frame(FID=pheno$IID,
                        IID=pheno$IID,
                        Fluid.intelligence.score=pheno$Fluid.intelligence.score)
                        
fwrite(pheno_plink, paste0(UKBB_output,'/Phenotype/Intelligence/UKBB_Fluid.intelligence.score.pheno'), sep='\t')

q()
n
```
</details>

```{bash, echo=F, eval=T}
# Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config

mkdir -p /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Intelligence

cp ${UKBB_output}/Phenotype/Intelligence/Histogram.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Intelligence/
```

<details><summary>Descriptives of Intelligence phenotype</summary>

<center>

![Figure 1: Distribution of Intelligence in UK Biobank](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Intelligence/Histogram.png)

\center

<br/>

***

```{r, echo=F, eval=T, results='asis'}
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
desc<-read.table(paste0(UKBB_output,'/Phenotype/Intelligence/Descriptives.txt'), header=T)
desc$Mean<-round(desc$Mean,3)
desc$SD<-round(desc$SD,3)
desc$Min<-round(desc$Min,3)
desc$Max<-round(desc$Max,3)
desc$Skewness<-round(desc$Skewness,3)
desc$N<-round(desc$N,3)

library(knitr)
kable(desc, row.names = FALSE, caption='Table 1: Descriptives of Intelligence in UK Biobank')
```

</details>

<br/>

***

## Height
The Height phenotype is derived from the Standing.height.0.0 (f.50.0.0) variable.

<details><summary>Preparation of Height phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Extract Height variable, update IDs and calculate descriptives
#####
module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

# Read in the field finder to find he code for the phenotype of interest.
field_finder<-read.table(paste0(ukbb_pheno_file,'/ukb18177_glanville_field_finder.txt'), stringsAsFactors=F)
field_finder[grepl('height', field_finder$V3),]
# The best height variable has a code of f.50.0.0 and is called Standing.height.0.0

# Identify which column in the phenotype file contains the relevant code
library(data.table)
pheno_header<-names(data.frame(fread(cmd=paste0('head -n 1 ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt'))))
rel_cols<-c(1,which(grepl('f.50.0.0',pheno_header)))

# Read in the IID and fluid intelligence columns
pheno<-data.frame(fread(cmd=paste0('cut -f ',paste(rel_cols,collapse=','),' ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt')))
names(pheno)<-c('IID','Height')

# Remove individuals with missing phenotype data
pheno<-pheno[complete.cases(pheno),]

# Calculate descriptive statistics for the phenotype
library(e1071)
descriptives<-data.frame(Phenotype='Standing.height.0.0',
                        Code='f.50.0.0',
                        Mean=mean(pheno$Height),
                        SD=sd(pheno$Height),
                        Min=min(pheno$Height),
                        Max=max(pheno$Height),
                        Skewness=skewness(pheno$Height),
                        N=dim(pheno)[1])

system(paste0('mkdir ',UKBB_output,'/Phenotype/Height'))
write.table(descriptives,paste0(UKBB_output,'/Phenotype/Height/Descriptives.txt'), row.names=F, quote=F)

# Plot the distribution
png(paste0(UKBB_output,'/Phenotype/Height/Histogram.png'))
hist(pheno$Height)
dev.off()

# Remove outliers
pheno_no_outlier<-pheno[(pheno$Height < mean(pheno$Height)+(3*sd(pheno$Height))) & (pheno$Height > mean(pheno$Height)-(3*sd(pheno$Height))),]

# Plot the distribution
png(paste0(UKBB_output,'/Phenotype/Height/Histogram_noOutlier.png'))
hist(pheno_no_outlier$Height)
dev.off()

# Write out the phenotype in PLINK format.
pheno_plink<-data.frame(FID=pheno_no_outlier$IID,
                        IID=pheno_no_outlier$IID,
                        Height=pheno_no_outlier$Height)
                        
fwrite(pheno_plink, paste0(UKBB_output,'/Phenotype/Height/UKBB_Height.pheno'), sep='\t')

q()
n
```
</details>

```{bash, echo=F, eval=T}
# Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config

mkdir -p /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Height

cp ${UKBB_output}/Phenotype/Height/Histogram.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Height/
cp ${UKBB_output}/Phenotype/Height/Histogram_noOutlier.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Height/
```

<details><summary>Descriptives of Height phenotype</summary>

<center>

![Figure 2: Distribution of Height in UK Biobank](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Height/Histogram.png)

![Figure 3: Distribution of Height in UK Biobank after removal of outliers](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/Height/Histogram_noOutlier.png)

\center

<br/>

***

```{r, echo=F, eval=T, results='asis'}
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
desc<-read.table(paste0(UKBB_output,'/Phenotype/Height/Descriptives.txt'), header=T)
desc$Mean<-round(desc$Mean,3)
desc$SD<-round(desc$SD,3)
desc$Min<-round(desc$Min,3)
desc$Max<-round(desc$Max,3)
desc$Skewness<-round(desc$Skewness,3)
desc$N<-round(desc$N,3)

library(knitr)
kable(desc, row.names = FALSE, caption='Table 1: Descriptives of Height in UK Biobank')
```

</details>

<br/>

***

## BMI
The BMI phenotype is derived from the Body.mass.index.BMI.0.0 (f.21001.0.0) variable.

<details><summary>Preparation of BMI phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Extract BMI variable, update IDs and calculate descriptives
#####

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

# Read in the field finder to find he code for the phenotype of interest.
field_finder<-read.table(paste0(ukbb_pheno_file,'/ukb18177_glanville_field_finder.txt'), stringsAsFactors=F)
field_finder[grepl('bmi', field_finder$V3),]
# The best height variable has a code of f.21001.0.0 and is called Body.mass.index.BMI.0.0

# Identify which column in the phenotype file contains the relevant code
library(data.table)
pheno_header<-names(data.frame(fread(cmd=paste0('head -n 1 ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt'))))
rel_cols<-c(1,which(grepl('f.21001.0.0',pheno_header)))

# Read in the IID and fluid intelligence columns
pheno<-data.frame(fread(cmd=paste0('cut -f ',paste(rel_cols,collapse=','),' ',ukbb_pheno_file,'/ukb18177_glanville_phenotypes.txt')))
names(pheno)<-c('IID','BMI')

# Remove individuals with missing phenotype data
pheno<-pheno[complete.cases(pheno),]

# Calculate descriptive statistics for the phenotype
library(e1071)
descriptives<-data.frame(Phenotype='Body.mass.index.BMI.0.0',
                        Code='f.50.0.0',
                        Mean=mean(pheno$BMI),
                        SD=sd(pheno$BMI),
                        Min=min(pheno$BMI),
                        Max=max(pheno$BMI),
                        Skewness=skewness(pheno$BMI),
                        N=dim(pheno)[1])

system(paste0('mkdir ',UKBB_output,'/Phenotype/BMI'))
write.table(descriptives,paste0(UKBB_output,'/Phenotype/BMI/Descriptives.txt'), row.names=F, quote=F)

# Plot the distribution
png(paste0(UKBB_output,'/Phenotype/BMI/Histogram.png'))
hist(pheno$BMI)
dev.off()

# Remove outliers
pheno_no_outlier<-pheno[(pheno$BMI < mean(pheno$BMI)+(3*sd(pheno$BMI))) & (pheno$BMI > mean(pheno$BMI)-(3*sd(pheno$BMI))),]

# Plot the distribution
png(paste0(UKBB_output,'/Phenotype/BMI/Histogram_noOutlier.png'))
hist(pheno_no_outlier$BMI)
dev.off()

# Write out the phenotype in PLINK format.
pheno_plink<-data.frame(FID=pheno_no_outlier$IID,
                        IID=pheno_no_outlier$IID,
                        BMI=pheno_no_outlier$BMI)
                        
fwrite(pheno_plink, paste0(UKBB_output,'/Phenotype/BMI/UKBB_BMI.pheno'), sep='\t')

q()
n
```

</details>

```{bash, echo=F, eval=T}
# Set variables
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config

mkdir -p /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/BMI

cp ${UKBB_output}/Phenotype/BMI/Histogram.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/BMI/
cp ${UKBB_output}/Phenotype/BMI/Histogram_noOutlier.png /users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/BMI/
```

<details><summary>Descriptives of BMI phenotype</summary>

<center>

![Figure 4: Distribution of BMI in UK Biobank](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/BMI/Histogram.png)

![Figure 5: Distribution of BMI in UK Biobank after removal of outliers](/users/k1806347/brc_scratch/Software/MyGit/GenoPred/Images/Target_sample_phenotype_prep/BMI/Histogram_noOutlier.png)

\center

<br/>

***

```{r, echo=F, eval=T, results='asis'}
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
desc<-read.table(paste0(UKBB_output,'/Phenotype/BMI/Descriptives.txt'), header=T)
desc$Mean<-round(desc$Mean,3)
desc$SD<-round(desc$SD,3)
desc$Min<-round(desc$Min,3)
desc$Max<-round(desc$Max,3)
desc$Skewness<-round(desc$Skewness,3)
desc$N<-round(desc$N,3)

library(knitr)
kable(desc, row.names = FALSE, caption='Table 1: Descriptives of BMI in UK Biobank')
```

</details>

<br/>

## CAD
CAD phenotype was shared by Saskia Hagenaars

<details><summary>Preparation of CAD phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Modify CAD phenotype file so IDs match genotypic data
#####

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)
pheno<-fread(paste0(UKBB_output, '/Phenotype/CAD/cad_only_111119.txt'))
names(pheno)[1]<-'IID'
fam<-fread(paste0(UKBB_output,'/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam'))
names(fam)[1:2]<-c('FID','IID')

pheno_fam<-merge(pheno, fam, by=c('IID'))
pheno_new<-pheno_fam[,c('FID','IID','CAD_all'), with=F]

fwrite(pheno_new, paste0(UKBB_output,'/Phenotype/CAD/cad_only_111119.UpdateIDs.txt'), sep=' ', na='NA')

q()
n
```
</details>

## Type-2 Diabetes (T2D)
T2D phenotypes were shared by Saskia Hagenaars

<details><summary>Preparation of T2D phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Modify T2D phenotype file so IDs match genotypic data
#####

module add apps/R/3.6.0
R
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)
pheno<-fread(paste0(UKBB_output, '/Phenotype/T2D/t2d_only_111119.txt'))
names(pheno)[1]<-'IID'
fam<-fread(paste0(UKBB_output,'/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam'))
names(fam)[1:2]<-c('FID','IID')

pheno_fam<-merge(pheno, fam, by=c('IID'))
pheno_new<-pheno_fam[,c('FID','IID','t2d_all'), with=F]

fwrite(pheno_new, paste0(UKBB_output,'/Phenotype/T2D/t2d_only_111119.UpdateIDs.txt'), sep=' ', na='NA')

q()
n
```
</details>

<br/>

## Autoimmune disorders
Phenotypes were shared by Kylie Glanville

<details><summary>Preparation of Autoimmune disorder phenotype</summary>
```{bash, echo=T, eval=F}
#####
# Modify Autoimmune Disorder phenotype file so IDs match genotypic data
#####

module add apps/R/3.6.0
R
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)
pheno<-fread(paste0(UKBB_output, '/Phenotype/AutoImmune/new_autoimmune_all'))
pheno$FID<-NULL
fam<-fread(paste0(UKBB_output,'/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr22.fam'))
names(fam)[1:2]<-c('FID','IID')

pheno_fam<-merge(pheno, fam, by=c('IID'))

# Extract probable phnotypes only
pheno_new<-pheno_fam[,c('FID','IID',names(pheno_fam)[grepl('_prob', names(pheno_fam))]), with=F]

# Identify individuals that don't have any autoimmune disorders
pheno_new_control<-pheno_new[which(rowSums(pheno_new[,-1:-2]) == 0),]

# Create a pheno file for Crohns MultiScler RheuArth
pheno_name<-c('IBD','MultiScler','RheuArth')
pheno_col<-c('ibd_prob',"ms_prob","arthritis_prob")
for(i in 1:3){
  pheno_new_i_case<-pheno_new[pheno_new[[pheno_col[i]]] == 1,]
  pheno_clean_i<-rbind(pheno_new_i_case, pheno_new_control)
  pheno_clean_i<-pheno_clean_i[,c('FID','IID',pheno_col[i]), with=F]
  system(paste0('mkdir ',UKBB_output,'/Phenotype/',pheno_name[i]))
  fwrite(pheno_clean_i, paste0(UKBB_output,'/Phenotype/',pheno_name[i],'/UKBB.',pheno_name[i],'.txt'), sep=' ', na='NA', quote=F)

}

q()
n
```
</details>

<br/>

## Create subset of UK Biobank individuals
The UK Biobank sample is very large making the computation of the various scores and derivation of the models computationally intensive. We do not need such large samples to compare PRS methods, and it would also be useful to have a non-overlapping subset of UKBiobank which can be used as a reference sample.

<details><summary>Subsetting phenotypic data</summary>
```{bash, echo=T, eval=F}
#####
# Idenitfy subset of UK biobank with phenotypic data and European ancestry
#####

module add apps/R/3.6.0
R
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')
library(data.table)

pheno<-c('Depression','Intelligence','BMI','Height','T2D','CAD','IBD','MultiScler','RheuArth')
pheno_file<-c('ever_depressed_pheno_final.UpdateIDs.txt','UKBB_Fluid.intelligence.score.UpdateIDs.pheno','UKBB_BMI.score.UpdateIDs.pheno','UKBB_Height.score.UpdateIDs.pheno','t2d_only_111119.UpdateIDs.txt','cad_only_111119.UpdateIDs.txt','UKBB.IBD.txt','UKBB.MultiScler.txt','UKBB.RheuArth.txt')

pheno_list<-list()
for(i in 1:length(pheno)){
  pheno_list[[pheno[i]]]<-fread(paste0(UKBB_output,'/Phenotype/',pheno[i],'/',pheno_file[i]))
}

pheno_all<-Reduce(function(...) merge(..., all=T, by=c('FID','IID')), pheno_list)

# We want to retain all individuals that are cases, and those that have complete data.
pheno_all_2<-pheno_all[complete.cases(pheno_all) | pheno_all$Depressed.Ever.Clean == 1 | pheno_all$t2d_all == 1 | pheno_all$CAD_all == 1 | pheno_all$ibd_prob == 1 | pheno_all$ms_prob == 1 | pheno_all$arthritis_prob == 1,]

# Extract individuals with EUR ancestry
EUR<-fread(paste0(UKBB_output,'/Projected_PCs/AllAncestry/UKBB.w_hm3.AllAncestry.EUR.keep'))
pheno_all_2_EUR<-pheno_all_2[(pheno_all_2$IID %in% EUR$V2),]

# There are still >100k individuals. Write out files for each phenotype, retaining individuals with complete data, with a maximum of 50000 individuals for quantitative outcomes.
system(paste0('mkdir ',UKBB_output,'/Phenotype/PRS_comp_subset'))
pheno_N<-NULL
for(i in 1:length(pheno)){
  pheno_all_2_EUR_i<-pheno_all_2_EUR[,c('FID','IID',names(pheno_all_2_EUR)[2+i]),with=F]
  pheno_all_2_EUR_i<-pheno_all_2_EUR_i[complete.cases(pheno_all_2_EUR_i),]
  if(sum(pheno_all_2_EUR_i[[3]] == 0 | pheno_all_2_EUR_i[[3]] == 1) == length(pheno_all_2_EUR_i[[3]])){
    pheno_all_2_EUR_i_case<-pheno_all_2_EUR_i[pheno_all_2_EUR_i[[3]] == 1,]
    pheno_all_2_EUR_i_case<-pheno_all_2_EUR_i_case[1:25000,]
    pheno_all_2_EUR_i_case<-pheno_all_2_EUR_i_case[complete.cases(pheno_all_2_EUR_i_case),]
    pheno_all_2_EUR_i_con<-pheno_all_2_EUR_i[pheno_all_2_EUR_i[[3]] == 0,]
    pheno_all_2_EUR_i_con<-pheno_all_2_EUR_i_con[1:(50000 - dim(pheno_all_2_EUR_i_case)[1]),]
    pheno_all_2_EUR_i<-rbind(pheno_all_2_EUR_i_case, pheno_all_2_EUR_i_con)
    
    pheno_N<-rbind(pheno_N,data.frame(Pheno=pheno[i], 
                                      N=length(pheno_all_2_EUR_i[[3]]),
                                      N_case=sum(pheno_all_2_EUR_i[[3]] == 1),
                                      N_con=sum(pheno_all_2_EUR_i[[3]] == 0)))
  } else {
    pheno_all_2_EUR_i<-pheno_all_2_EUR_i[1:50000,]
    pheno_N<-rbind(pheno_N,data.frame(Pheno=pheno[i], 
                                      N=length(pheno_all_2_EUR_i[[3]]),
                                      N_case=NA,
                                      N_con=NA))
  }
  write.table(pheno_all_2_EUR_i, paste0(UKBB_output,'/Phenotype/PRS_comp_subset/UKBB.',pheno[i],'.txt'), col.names=T, row.names=F, quote=F)
}

write.csv(pheno_N, paste0(UKBB_output,'/Phenotype/PRS_comp_subset/Pheno_N.csv'), row.names=F, quote=F)
  
```

## Estimate SNP-h2 of UK Biobank phenotypes
We will estimate the SNP-heritability of each phenotype based on the 50k subsets we have just created. We will retain only indiviudals that pass Kylie's QC, as we have done in the prediction modelling stage. Given the sample size of each phenotype, LD-score regression seems the most suitable strategy for estimating heritability. GREML in GCTA will be computationally intensive.

<details><summary>Estimating SNP-heritability</summary>
```{bash, echo=T, eval=F}
. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config
####
# Perform GWAS for each phenotype
####

pheno=$(echo Depression Intelligence BMI Height T2D CAD IBD MultiScler RheuArth)
binary=$(echo T F F F T T T T T)

for i in $(seq 1 9); do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  binary_i=$(echo ${binary} | cut -f ${i} -d ' ')
  
  mkdir ${UKBB_output}/Phenotype/${pheno_i}/GWAS
  
  if [[ ${binary_i} == "T" ]]
  then
    for chr in $(seq 1 22);do
      sbatch -p brc,shared ${plink1_9} \
        --bfile ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr${chr} \
        --keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
        --pheno ${UKBB_output}/Phenotype/PRS_comp_subset/UKBB.${pheno_i}.txt \
        --1 \
        --assoc \
        --allow-no-sex \
        --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.chr${chr}
        
      sbatch -p brc,shared ${plink1_9} \
        --bfile ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr${chr} \
        --keep ${UKBB_output}/Phenotype/PRS_comp_subset/UKBB.${pheno_i}.txt \
        --freq \
        --allow-no-sex \
        --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.chr${chr}

    done
  else
      for chr in $(seq 1 22);do
      sbatch -p brc,shared ${plink1_9} \
        --bfile ${UKBB_output}/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr${chr} \
        --keep /users/k1806347/brc_scratch/Analyses/UKBB_outcomes_for_prediction/ukb18177_glanville_post_qc_id_list.UpdateIDs.fam \
        --pheno ${UKBB_output}/Phenotype/PRS_comp_subset/UKBB.${pheno_i}.txt \
        --assoc \
        --allow-no-sex \
        --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.chr${chr}
    done
  fi
    
done

####
# Format for munge_sumstats.py
####

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

library(data.table)

# Read in bim file
bim<-NULL
for(chr in 1:22){
  bim_chr<-fread(paste0(UKBB_output,'/Genotype/Harmonised/UKBB.w_hm3.QCd.AllSNP.chr',chr,'.bim'))
  bim<-rbind(bim,bim_chr)
}

pheno<-c('Depression','Intelligence','BMI','Height','T2D','CAD','IBD','MultiScler','RheuArth')
binary<-c(T,F,F,F,T,T,T,T,T)

for(i in 1:9){
gwas_i<-NULL
  for(chr in 1:22){
    if(binary[i] == T){
        gwas_i_chr<-fread(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.chr',chr,'.assoc'))
        frq_i_chr<-fread(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.chr',chr,'.frq'))
        frq_i_chr$NCHROBS<-frq_i_chr$NCHROBS/2
        gwas_i_chr<-merge(gwas_i_chr,frq_i_chr[,c('SNP','NCHROBS')],by='SNP')
        gwas_i_chr<-gwas_i_chr[,c('SNP','A1','A2','OR','P','NCHROBS')]
        names(gwas_i_chr)<-c('SNP','A1','A2','OR','P','N')
        gwas_i_chr<-gwas_i_chr[complete.cases(gwas_i_chr),]
    } else {
        gwas_i_chr<-fread(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.chr',chr,'.qassoc'))
        gwas_i_chr<-merge(gwas_i_chr, bim, by.x='SNP', by.y='V2')
        gwas_i_chr<-gwas_i_chr[,c('SNP','V5','V6','BETA','P','NMISS')]
        names(gwas_i_chr)<-c('SNP','A1','A2','BETA','P','N')
        gwas_i_chr<-gwas_i_chr[complete.cases(gwas_i_chr),]
    }
    gwas_i<-rbind(gwas_i, gwas_i_chr)
  }
  
  write.table(gwas_i, paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.GW.assoc_clean'), col.names=T, row.names=F, quote=F)
}

q()
n

####
# Munge sumstats
####

. /users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config

pheno=$(echo Depression Intelligence BMI Height T2D CAD IBD MultiScler RheuArth)
binary=$(echo T F F F T T T T T)

for i in $(seq 1 9); do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  binary_i=$(echo ${binary} | cut -f ${i} -d ' ')
  
  sbatch -p brc,shared ~/brc_scratch/Software/munge_sumstats.sh \
   --sumstats ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW.assoc_clean \
   --merge-alleles ${HapMap3_snplist_dir}/w_hm3.snplist \
   --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW
done

####
# Estimate heritability
####

pheno=$(echo Depression Intelligence BMI Height T2D CAD IBD MultiScler RheuArth)
binary=$(echo T F F F T T T T T)
pop_prev=$(echo 0.15 NA NA NA 0.05 0.03 0.013 0.00164 0.005)
samp_prev=$(echo 0.5 NA NA NA 0.298 0.5 0.069 0.023 0.068)

for i in $(seq 1 9); do
  pheno_i=$(echo ${pheno} | cut -f ${i} -d ' ')
  binary_i=$(echo ${binary} | cut -f ${i} -d ' ')
  pop_prev_i=$(echo ${pop_prev} | cut -f ${i} -d ' ')
  samp_prev_i=$(echo ${samp_prev} | cut -f ${i} -d ' ')
  
  if [[ ${binary_i} == "T" ]]
    then
    sbatch -p brc,shared ~/brc_scratch/Software/ldsc.sh \
     --h2 ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW.sumstats.gz \
     --ref-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --w-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW
     
    sbatch -p brc,shared ~/brc_scratch/Software/ldsc.sh \
     --h2 ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW.sumstats.gz \
     --ref-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --w-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --samp-prev ${pop_prev_i} \
     --pop-prev ${pop_prev_i} \
     --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW.liab
     
  else
  
   sbatch -p brc,shared ~/brc_scratch/Software/ldsc.sh \
     --h2 ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW.sumstats.gz \
     --ref-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --w-ld-chr ${HapMap3_snplist_dir}/eur_w_ld_chr/ \
     --out ${UKBB_output}/Phenotype/${pheno_i}/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.${pheno_i}.GW

  fi
  
done

#####
# Tabulate results
#####

module add apps/R/3.6.0
R

source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

library(data.table)

pheno<-c('Depression','Intelligence','BMI','Height','T2D','CAD','IBD','MultiScler','RheuArth')
binary<-c(T,F,F,F,T,T,T,T,T)

N_tab<-fread(paste0(UKBB_output,'/Phenotype/PRS_comp_subset/Pheno_N.csv'))
for(i in 1:9){
  if(binary[i] == T){
    h2_i<-read.table(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.GW.log'), header=F, sep='&')
    h2_liab_i<-read.table(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.GW.liab.log'), header=F, sep='&')
    N_tab$h2_obs[N_tab$Pheno == pheno[i]]<-gsub('Total Observed scale h2: ','',h2_i$V1[grepl('Total Observed scale h2: ',h2_i$V1)])
    N_tab$h2_liab[N_tab$Pheno == pheno[i]]<-gsub('Total Liability scale h2: ','',h2_liab_i$V1[grepl('Total Liability scale h2: ',h2_liab_i$V1)])
    N_tab$intercept[N_tab$Pheno == pheno[i]]<-gsub('Intercept: ','',h2_i$V1[grepl('Intercept: ',h2_i$V1)])
    N_tab$lambda[N_tab$Pheno == pheno[i]]<-gsub('Lambda GC: ','',h2_i$V1[grepl('Lambda GC: ',h2_i$V1)])
  } else {
      h2_i<-read.table(paste0(UKBB_output,'/Phenotype/',pheno[i],'/GWAS/UKBB.w_hm3.QCd.AllSNP.Subset.',pheno[i],'.GW.log'), header=F, sep='&')
    N_tab$h2_obs[N_tab$Pheno == pheno[i]]<-gsub('Total Observed scale h2: ','',h2_i$V1[grepl('Total Observed scale h2: ',h2_i$V1)])
    N_tab$intercept[N_tab$Pheno == pheno[i]]<-gsub('Intercept: ','',h2_i$V1[grepl('Intercept: ',h2_i$V1)])
    N_tab$lambda[N_tab$Pheno == pheno[i]]<-gsub('Lambda GC: ','',h2_i$V1[grepl('Lambda GC: ',h2_i$V1)])
  }
}

write.csv(N_tab, paste0(UKBB_output,'/Phenotype/PRS_comp_subset/Pheno_N_h2.csv'), row.names=F, quote=F)
q()
n

####
# Delete intermediate files
####

rm ${UKBB_output}/Phenotype/*/GWAS/*.assoc
rm ${UKBB_output}/Phenotype/*/GWAS/*chr*.frq
rm ${UKBB_output}/Phenotype/*/GWAS/*chr*.log
rm ${UKBB_output}/Phenotype/*/GWAS/*chr*.nosex

```

# TEDS

Phenotypes:

* BMI (age 21)
* Height (age 21)
* GCSE scores
* ADHD traits (age 12)

<details><summary>Preparation of TEDS phenotype</summary>

```{bash, eval=F}
################################################################
# Prepare phenotypes in TEDS for prediction
################################################################

module add general/R/3.5.0
R
source('/users/k1806347/brc_scratch/Software/MyGit/GenoPred/config_used/Phenotype_prep.config')

#########
# Read in the TEDS data
#########

library(foreign)

pheno<-read.spss(paste0(TEDS_output_dir,'/Phenotypic/data_recieved/399\ Oliver\ Pain\ GERS\ 060319.sav'),to.data.frame=T)

#########
# Rename required variables to match Saskia's script
#########

x<-data.frame(	FID=pheno$id_fam,
				IID=pheno$id_twin,
				twin=pheno$twin,
				sex1=pheno$sex1,
				sex2=pheno$sex2,
				age12par=pheno$lpqage,
				SDQhyp_par12_1=pheno$lpshypt1,
				SDQhyp_par12_2=pheno$lpshypt2,
				Con_ADHD12_par_1=pheno$lpconnt1,
				Con_ADHD12_par_2=pheno$lpconnt2,
				age16par=pheno$ppbhage,
				SDQhyp_par16_1=pheno$ppbhsdqhypt1,
				SDQhyp_par16_2=pheno$ppbhsdqhypt2,
				Con_ADHD16_par16_1=pheno$ppbhconnt1,
				Con_ADHD16_par16_2=pheno$ppbhconnt2,
				GCSEage=pheno$pcexgcseage1,
				GCSE_1=pheno$pcexgcsecorem1,
				GCSE_2=pheno$pcexgcsecorem2,
				Weight21_1=pheno$u1cwtkg1,
				Weight21_2=pheno$u1cwtkg2,
				Height21_1=pheno$u1chtcm1,
				Height21_2=pheno$u1chtcm2,
				age21_1=pheno$u1cage1,
				age21_2=pheno$u1cage2)

#########
# Create BMI variable
#########

x$BMI21_1 		<- x$Weight21_1/((x$Height21_1/100)^2)
x$BMI21_2 		<- x$Weight21_2/((x$Height21_2/100)^2)

#########
# Create ADHD symptom score variable
#########

#create age regressed mean scores of SDZ & ADHD parent ratings at age 12 & 16
#ADHD as the mean of parent reported SDZ & Connors scores
#set function to create cleaned age specific ADHD variable
zresid <- 	function(x,y,age) {
					scale(
						resid(
							lm(
								((scale(x)+scale(y))/2)~age,na.action=na.exclude
							)
						)
					)
				}		


#create age regressed mean scores of SDZ & ADHD parent ratings at age 12 & 16
x$ADHD12_1 		<- zresid(x=x$SDQhyp_par12_1,y=x$Con_ADHD12_par_1,age=x$age12par)
x$ADHD12_2 		<- zresid(x=x$SDQhyp_par12_2,y=x$Con_ADHD12_par_2,age=x$age12par)

x$ADHD16_1 		<- zresid(x=x$SDQhyp_par16_1,y=x$Con_ADHD16_par16_1,age=x$age16par)
x$ADHD16_2 		<- zresid(x=x$SDQhyp_par16_2,y=x$Con_ADHD16_par16_2,age=x$age16par)

#create ADHD mean score. If both age 12 & 16 available, take mean, if 16 is missing use 12 only and vice versa
x$ADHD_1 			<- ifelse((!is.na(x$ADHD12_1) & !is.na(x$ADHD16_1)),  
								(x$ADHD12_1+x$ADHD16_1)/2,
						ifelse(is.na(x$ADHD12_1),x$ADHD16_1,
							ifelse(is.na(x$ADHD16_1), x$ADHD12_1, NA
							) 
						)
					)	

x$ADHD_2 			<- ifelse((!is.na(x$ADHD12_2) & !is.na(x$ADHD16_2)),  
								(x$ADHD12_2+x$ADHD16_2)/2,
						ifelse(is.na(x$ADHD12_2),x$ADHD16_2,
							ifelse(is.na(x$ADHD16_2), x$ADHD12_2, NA
							) 
						)
					)	

#create ADHD age variables for longitudinal measures, which is age 16 centered
x$ADHD_AGE_1 			<- ifelse((!is.na(x$ADHD12_1) & !is.na(x$ADHD16_1)),  
									x$age16par-16,
								ifelse(is.na(x$ADHD12_1),x$age16par-16,
									ifelse(is.na(x$ADHD16_1), x$age12par-16, NA
								) 
							)
						)	


x$ADHD_AGE_2 			<- ifelse((!is.na(x$ADHD12_2) & !is.na(x$ADHD16_2)),  
									x$age16par-16,
								ifelse(is.na(x$ADHD12_2),x$age16par-16,
									ifelse(is.na(x$ADHD16_2), x$age12par-16, NA
								) 
							)
						)	

#########
# Now regress out sex (and age if necessary)
#########

x$clHeight21_1 <- rstandard(lm(Height21_1~sex1+age21_1,data=x,na.action=na.exclude))
x$clHeight21_2 <- rstandard(lm(Height21_2~sex2+age21_2,data=x,na.action=na.exclude))
x$clBMI21_1 <- rstandard(lm(BMI21_1~sex1+age21_1,data=x,na.action=na.exclude))
x$clBMI21_2 <- rstandard(lm(BMI21_2~sex2+age21_2,data=x,na.action=na.exclude))
x$clGCSE_1 <- rstandard(lm(GCSE_1~sex1+GCSEage,data=x,na.action=na.exclude))
x$clGCSE_2 <- rstandard(lm(GCSE_2~sex2+GCSEage,data=x,na.action=na.exclude))
x$clADHD_1 <- rstandard(lm(ADHD_1~sex1+ADHD_AGE_1,data=x,na.action=na.exclude))
x$clADHD_2 <- rstandard(lm(ADHD_2~sex2+ADHD_AGE_2,data=x,na.action=na.exclude))

#########
# Create files containing each phenotype with unrelated individuals only
#########

# Make function that randomly removes duplicates
duplicated.random = function(x, incomparables = FALSE, ...) 
{ 
     if ( is.vector(x) ) 
     { 
         permutation = sample(length(x)) 
         x.perm      = x[permutation] 
         result.perm = duplicated(x.perm, incomparables, ...) 
         result      = result.perm[order(permutation)] 
         return(result) 
     } 
     else if ( is.matrix(x) ) 
     { 
         permutation = sample(nrow(x)) 
         x.perm      = x[permutation,] 
         result.perm = duplicated(x.perm, incomparables, ...) 
         result      = result.perm[order(permutation)] 
         return(result) 
     } 
     else 
     { 
         stop(paste("duplicated.random() only supports vectors", 
                "matrices for now.")) 
     } 
} 

# Set FID to IID to match genetic data
x$FID<-x$IID

for(outcome in c('Height21','BMI21','GCSE','ADHD')){
	tmp<-x[c('FID','IID',paste0('cl',outcome,'_1'))]
	tmp<-tmp[complete.cases(tmp),]
	set.seed(1)
	tmp<-tmp[!duplicated.random(tmp$FID),]
	names(tmp)<-c('FID','IID',outcome)
	
	write.table(tmp, paste0(TEDS_output_dir,'/Phenotypic/Derived_outcomes/TEDS_',outcome,'.txt'), col.names=T, row.names=F, quote=F)
}

q()
n
```

</details>

<br/>
