<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Calculating Polytranscriptomic Scores</title>

<script src="site_libs/header-attrs-2.23/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-6.4.0/css/all.min.css" rel="stylesheet" />
<link href="site_libs/font-awesome-6.4.0/css/v4-shims.min.css" rel="stylesheet" />
<link rel="stylesheet" href="styles/night-mode.css" id="nightModeStylesheet">

<script>
function toggleNightMode() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (stylesheet.disabled) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
}
</script>

<label class="switch">
  <input type="checkbox" id="toggleNightMode" checked>
  <span class="slider round"></span>
</label>

<script>
document.getElementById('toggleNightMode').addEventListener('change', function() {
    var stylesheet = document.getElementById('nightModeStylesheet');
    if (this.checked) {
        stylesheet.disabled = false;
    } else {
        stylesheet.disabled = true;
    }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>






<link rel="stylesheet" href="styles/styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img class="logo-img" src="Images/logo/Horizontal_white.png" style="height: 42px;" /></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Pipeline
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="pipeline_overview.html">Overview</a>
    </li>
    <li>
      <a href="pipeline_readme.html">Instructions</a>
    </li>
    <li>
      <a href="pipeline_technical.html">Technical documentation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Research
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="research_index.html">Overview</a>
    </li>
    <li>
      <a href="comparison_of_methods_summary.html">Polygenic Scoring Methods Comparison</a>
    </li>
    <li>
      <a href="Functionally_informed_prediction.html">Quantifying Polygenic Signal Mediated by Altered Gene Expression</a>
    </li>
    <li>
      <a href="Absolute_Conversion.html">Translating Polygenic Scores onto the Absolute Scale</a>
    </li>
  </ul>
</li>
<li>
  <a href="more_index.html">More</a>
</li>
<li>
  <a href="https://github.com/opain/GenoPred">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Calculating Polytranscriptomic Scores</h1>

</div>


<p>This document describes the process of calculating polytranscriptomic
scores (PTS).</p>
<p>PTS are scores derived using differential gene expression summary
statistics, a gene expression correlation matrix, and individual-level
gene expression data in a target sample. All these data can come from
either observed gene expression or genetically-imputed gene expression
Genetically-imputed gene expression data refers to expression data that
has been imputed using genetic variation, typically genetic variation
surrounding and within each gene. Imputed differential gene expression
studies are referred to as transcriptome-wide association studies
(TWAS).</p>
<p>I will describe the two most common scenarios:</p>
<ol style="list-style-type: decimal">
<li>Imputed PTS: Derived using TWAS summary statistics and imputed
individual-level gene expression data in the target sample</li>
<li>Observed PTS: Derived using TWAS summary statistics and observed
individual-level gene expression data in the target sample</li>
</ol>
<p>The same procedure for calculating both imputed and observed PTS is
the same, it is just the origin of the target gene expression data.</p>
<p>To demonstrate the process, I will use some example TWAS results,
available <a
href="https://github.com/opain/Inferred-functional-risk-scoring/blob/master/ukbiobank-2017-1160-prePRS-fusion.tsv.GW">here</a>.</p>
<hr />
<div id="preparation-of-target-gene-expression-data"
class="section level1">
<h1>Preparation of target gene expression data</h1>
<hr />
<div id="imputed" class="section level2">
<h2>Imputed</h2>
<p>Imputing gene expression data requires target sample genotype data,
and gene expression SNP-weights (model predicting gene expression using
genetic variation). SNP-weights are available for a range contexts, and
can be downloaded from various sources, such as the <a
href="http://gusevlab.org/projects/fusion/">FUSION website</a>. Then you
can use tools such as PLINK to calculate the gene expression level in
the target sample, using the linear scoring function
(<code>--score</code>). Check out <a
href="https://github.com/opain/Predicting-TWAS-features">FeaturePred</a>
as well, which automates this process, using standard input formats.</p>
<hr />
</div>
<div id="observed" class="section level2">
<h2>Observed</h2>
<p>Observed gene expression data typically comes from array or RNA-seq
based assays. Quality control of the data typically involves
normalisation and covariate adjustment.</p>
<hr />
</div>
</div>
<div id="adjusting-twas-summary-statistics" class="section level1">
<h1>Adjusting TWAS summary statistics</h1>
<p>TWAS summary statistics must be adjusted to improve their utility for
polytranscriptomic scoring, mainly accounting for non-independent TWAS
effects, and some kind of penalisation of effect sizes to account for
overfitting. This is analagous to the adjustment of GWAS effect sizes
prior to the calculation of polygenic scores (PGS), and similar
approaches can be used.</p>
<p>As with PGS, PTS should be calculated in a reference-standardised
manner to enhance generalisability of the PTS to other samples.
Therefore, a reference genetic dataset, such as the 1000 Genomes sample
(1KG), should be used to model the relationship between TWAS feature
(genes). To do this, we first impute gene expression into the reference
dataset, which can be done using <a
href="https://github.com/opain/Predicting-TWAS-features">FeaturePred</a>.</p>
<hr />
<div id="ptclump" class="section level2">
<h2>pT+clump</h2>
<p>p-value thresholding and clumping is a simple approach. First lead
associations within each locus are identified, and correlated
associations are removed (this is referred to as clumping), and then a
range of p-value thresholds are used to select the genes that are
included in the PTS (which is a kind of penalisation).</p>
<p>The <a
href="https://github.com/opain/Inferred-functional-risk-scoring">IFRisk
software</a> can be used to perform the pT+clump procedure with TWAS
summary statistics and gene expression data. This should be applied
using the reference imputed expression data, as you want to model the
correlation beteen the SNP-weights used to generate the TWAS effect
sizes.</p>
<pre class="bash"><code># Impute gene expression into reference
sbatch -p shared -n 6 --mem 50G Rscript FeaturePred.V2.0.R \
    --PLINK_prefix_chr FUSION/LDREF/1000G.EUR. \
    --weights test_data/CMC.BRAIN.RNASEQ/CMC.BRAIN.RNASEQ.pos \
    --weights_dir test_data/CMC.BRAIN.RNASEQ \
    --ref_ld_chr FUSION/LDREF/1000G.EUR. \
    --plink plink \
    --ref_maf FUSION/LDREF/1000G.EUR. \
    --pigz pigz \
    --chr 1 \
    --output demo \
    --n_cores 6 \
    --memory 50000

# Perform pT+clump
Rscript IFRisk.V1.0.R \
    --twas_results ukbiobank-2017-1160-prePRS-fusion.tsv.GW \
    --target_gene_exp demo/FeaturePredictions_all_panels.txt.gz \
    --output demo
</code></pre>
<hr />
</div>
<div id="other-approaches" class="section level2">
<h2>Other approaches</h2>
<p>Other PGS-like approaches have not yet been explored for PTS, but
some approaches could be easily extended, such as lassosum2 or ldpred2,
replacing the SNP-correlation matrix with an imputed gene expression
correlation matrix, and the GWAS summary statistics, with the TWAS
summary statistics.</p>
<hr />
</div>
</div>
<div id="target-scoring" class="section level1">
<h1>Target scoring</h1>
<p>With target gene expression data (imputed or observed), and the
adjusted TWAS summary statistics, the PTS can now be calculated in the
target sample, simply as the TWAS effect size weighted sum of gene
expression levels for each individual. Since TWAS scores will be used as
the weights, the expression level of each gene must be standardised
(mean = 0 and sd = 1). FeaturePred automatically reference</p>
<p>Here is some code for doing this.</p>
<pre class="r"><code>library(data.table)
set.seed(1)

# Simulate target expression data
n_gene &lt;- 5
n_indiv &lt;- 5

exp &lt;- matrix(rnorm(n_gene*n_indiv), nrow = n_indiv, ncol = n_gene)
exp &lt;- data.table(exp)
names(exp)&lt;-paste0(&#39;gene_&#39;, 1:n_gene)

exp</code></pre>
<pre><code>##        gene_1     gene_2     gene_3      gene_4      gene_5
## 1: -0.6264538 -0.8204684  1.5117812 -0.04493361  0.91897737
## 2:  0.1836433  0.4874291  0.3898432 -0.01619026  0.78213630
## 3: -0.8356286  0.7383247 -0.6212406  0.94383621  0.07456498
## 4:  1.5952808  0.5757814 -2.2146999  0.82122120 -1.98935170
## 5:  0.3295078 -0.3053884  1.1249309  0.59390132  0.61982575</code></pre>
<pre class="r"><code># Simulate adjusted TWAS summary statistics
twas_effect &lt;- rnorm(n_gene)
pt &lt;- c(0.1, 0.5, 1) # p-value thresholds
score &lt;- data.table(
  ID = paste0(&#39;gene_&#39;, n_gene:1)
)
for(i in pt){
  score$tmp &lt;- twas_effect
  score$tmp[abs(twas_effect) &lt; qnorm(1-(i/2))] &lt;- 0
  names(score)[names(score) == &#39;tmp&#39;] &lt;- paste0(&#39;pt_&#39;,i)
}

score</code></pre>
<pre><code>##        ID pt_0.1    pt_0.5        pt_1
## 1: gene_5      0  0.000000 -0.05612874
## 2: gene_4      0  0.000000 -0.15579551
## 3: gene_3      0 -1.470752 -1.47075238
## 4: gene_2      0  0.000000 -0.47815006
## 5: gene_1      0  0.000000  0.41794156</code></pre>
<pre class="r"><code># Scale target expression data
exp&lt;-data.table(scale(exp))

# Find intersecting genes
exp &lt;- exp[, names(exp) %in% score$ID, with=F]
score &lt;- score[score$ID %in% names(exp),]

# Match order of genes
score &lt;- score[match(names(exp), score$ID), ]

# Weight and sum target expression by each score
pts &lt;- data.table()
for(i in pt){
  weighted_exp &lt;- cbind(pts, t(t(exp) * score[[paste0(&#39;pt_&#39;,i)]]))
  pts$tmp &lt;- rowSums(weighted_exp)
  names(pts)[names(pts) == &#39;tmp&#39;] &lt;- paste0(&#39;pt_&#39;,i)
}

pts</code></pre>
<pre><code>##    pt_0.1     pt_0.5       pt_1
## 1:      0 -1.4460093 -2.4075621
## 2:      0 -0.3451213 -0.7917431
## 3:      0  0.6469925  0.2811412
## 4:      0  2.2105551  4.7191817
## 5:      0 -1.0664171 -1.8010177</code></pre>
</div>

<!-- footer.html -->
<hr/>

<div class="centered-container">
<div class="rounded-image-container" style="width: 500px;">
<img src="Images/logo/sponsors.png">
</div>
</div>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
